---
output: github_document
always_allow_html: yes
bibliography: paper/bibliography.bib
---

![R](https://img.shields.io/badge/r-%23276DC3.svg?style=for-the-badge&logo=r&logoColor=white)
<a href="https://zenodo.org/badge/latestdoi/268765075"><img src="https://zenodo.org/badge/268765075.svg" alt="DOI"></a>
[![codecov](https://codecov.io/gh/adrientaudiere/MiscMetabar/graph/badge.svg?token=NXFRSIKYC0)](https://codecov.io/gh/adrientaudiere/MiscMetabar)
[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)](code_of_conduct.md) 
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- devtools::build_readme() -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# MiscMetabar

See the pkdown site [here](https://adrientaudiere.github.io/MiscMetabar/).

Biological studies, especially in ecology, health sciences and taxonomy, need to describe the biological composition of samples. During the last twenty years, (i) the development of DNA sequencing, (ii) reference databases, (iii) high-throughput sequencing (HTS), and (iv) bioinformatics resources have allowed the description of biological communities through metabarcoding. Metabarcoding involves the sequencing of millions (*meta*-) of short regions of specific DNA (*-barcoding*, @valentini2009) often from environmental samples (eDNA, @taberlet2012) such as human stomach contents, lake water, soil and air.

`MiscMetabar` aims to facilitate the **description**, **transformation**, **exploration** and **reproducibility** of metabarcoding analysis using R. The development of `MiscMetabar` relies heavily on the R packages [`dada2`](https://benjjneb.github.io/dada2/index.html), [`phyloseq`](https://joey711.github.io/phyloseq/) and [`targets`](https://books.ropensci.org/targets/). 


## Installation

There is no CRAN or bioconductor version of MiscMetabar for now (work in progress).

You can install the stable version from [GitHub](https://github.com/) with:

```{r, message=FALSE, eval=FALSE}
install.packages("devtools")
devtools::install_github("adrientaudiere/MiscMetabar")
```

You can install the developement version from [GitHub](https://github.com/) with:

```{r, message=FALSE, eval=FALSE}
install.packages("devtools")
devtools::install_github("adrientaudiere/MiscMetabar", ref = "dev")
```

## Some use of MiscMetabar

See vignettes in the [MiscMetabar](https://adrientaudiere.github.io/MiscMetabar/) website for more examples.


### Summarize a physeq object

```{r example, message=FALSE}
library("MiscMetabar")
library("phyloseq")
library("magrittr")
data("data_fungi")
summary_plot_pq(data_fungi)
```

### Alpha-diversity analysis

```{r, fig.cap="Hill number 1"}
p <- MiscMetabar::hill_pq(data_fungi, variable = "Height")
p$plot_Hill_0
```

```{r, fig.cap="Result of the Tuckey post-hoc test"}
p$plot_tuckey
```

### Beta-diversity analysis

```{r}
ggvenn_pq(data_fungi, fact = "Height") +
  ggplot2::scale_fill_distiller(palette = "BuPu", direction = 1) +
  labs(title = "Share number of ASV among Height in tree")
```

### Installation of other softwares for debian Linux distributions

#### blastn

```sh
sudo apt-get install ncbi-blast+
```

#### vsearch

```sh
sudo apt-get install vsearch
```
