---
output: github_document
always_allow_html: yes
bibliography: paper/bibliography.bib
---

![R](https://img.shields.io/badge/r-%23276DC3.svg?style=for-the-badge&logo=r&logoColor=white)
<a href="https://zenodo.org/badge/latestdoi/268765075"><img src="https://zenodo.org/badge/268765075.svg" alt="DOI"></a>
[![codecov](https://codecov.io/gh/adrientaudiere/MiscMetabar/graph/badge.svg?token=NXFRSIKYC0)](https://app.codecov.io/gh/adrientaudiere/MiscMetabar)
[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)](code_of_conduct.md) 
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![CodeFactor](https://www.codefactor.io/repository/github/adrientaudiere/miscmetabar/badge/master)](https://www.codefactor.io/repository/github/adrientaudiere/miscmetabar/overview/master)
[![R-CMD-check](https://github.com/adrientaudiere/MiscMetabar/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/adrientaudiere/MiscMetabar/actions/workflows/R-CMD-check.yaml)
[![DOI](https://joss.theoj.org/papers/10.21105/joss.06038/status.svg)](https://doi.org/10.21105/joss.06038)

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- devtools::build_readme() -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  message = FALSE
)
```

# MiscMetabar <a href="https://adrientaudiere.github.io/MiscMetabar/"><img src="https://adrientaudiere.github.io/MiscMetabar/reference/figures/logo.png" align="right" height="138" alt="MiscMetabar website" /></a>

See the pkgdown documentation site [here](https://adrientaudiere.github.io/MiscMetabar/) and the [package paper](https://doi.org/10.21105/joss.06038) in the Journal Of Open Softwares. 

Biological studies, especially in ecology, health sciences and taxonomy, need to describe the biological composition of samples. During the last twenty years, (i) the development of DNA sequencing, (ii) reference databases, (iii) high-throughput sequencing (HTS), and (iv) bioinformatics resources have allowed the description of biological communities through metabarcoding. Metabarcoding involves the sequencing of millions (*meta*-) of short regions of specific DNA (*-barcoding*, @valentini2009) often from environmental samples (eDNA, @taberlet2012) such as human stomach contents, lake water, soil and air.

`MiscMetabar` aims to facilitate the **description**, **transformation**, **exploration** and **reproducibility** of metabarcoding analysis using R. The development of `MiscMetabar` relies heavily on the R packages [`dada2`](https://benjjneb.github.io/dada2/index.html) [@callahan2016], [`phyloseq`](https://joey711.github.io/phyloseq/) [@mcmurdie2013] and [`targets`](https://books.ropensci.org/targets/) [@landau2021]. 



## Installation

There is no CRAN version of MiscMetabar for now (work in
progress). As MiscMetabar heavily relies on two bioconductor packages
(dada and phyloseq), we need to first install those 2 packages using
BiocManager.

You can install the stable version from [GitHub](https://github.com/)
with:

```{r, results = 'hide', eval=FALSE}
if (!require("devtools", quietly = TRUE)) {
  install.packages("devtools")
}
devtools::install_github("adrientaudiere/MiscMetabar")
```

You can install the development version from
[GitHub](https://github.com/) with:

```{r, results = 'hide', eval=FALSE}
if (!require("devtools", quietly = TRUE)) {
  install.packages("devtools")
}
devtools::install_github("adrientaudiere/MiscMetabar", ref = "dev")
```


## Some use of MiscMetabar

See articles in the [MiscMetabar](https://adrientaudiere.github.io/MiscMetabar/) website for more examples.

For an introduction to metabarcoding in R, Please visite the [state of the field](https://adrientaudiere.github.io/MiscMetabar/articles/states_of_fields_in_R.html) articles. The [import, export and track](https://adrientaudiere.github.io/MiscMetabar/articles/import_export_track.html) article explains how import and export `phyloseq` object. Its also show how to summarize useful information (number of sequences, samples and clusters) accross bioinformatic pipelines. The article [explore data](https://adrientaudiere.github.io/MiscMetabar/articles/explore_data.html) takes a closer look to different way of explore samples and taxonomical data from `phyloseq` object. 

 
If you are interested in ecological metrics, see the articles describing [alpha-diversity](https://adrientaudiere.github.io/MiscMetabar/articles/alpha-div.html) and [beta-diversity](https://adrientaudiere.github.io/MiscMetabar/articles/beta-div.html) analysis. 
The article [filter taxa and samples](https://adrientaudiere.github.io/MiscMetabar/articles/filter.html) describes some data-filtering processes using MiscMetabar and the [reclustering](https://adrientaudiere.github.io/MiscMetabar/articles/Reclustering.html) tutorial introduces the different way of clustering already-clustered OTU/ASV. The article [tengeler](https://adrientaudiere.github.io/MiscMetabar/articles/tengeler.html) explore the dataset from Tengeler et al. (2020) using some MiscMetabar functions.

For developers, I also wrote a article describing som [rules of codes](https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html). 

### Summarize a physeq object

```{r example}
library("MiscMetabar")
library("phyloseq")
library("magrittr")
data("data_fungi")
summary_plot_pq(data_fungi)
```

### Alpha-diversity analysis

```{r, fig.cap="Hill number 1"}
p <- MiscMetabar::hill_pq(data_fungi, variable = "Height")
p$plot_Hill_0
```

```{r, fig.cap="Result of the Tuckey post-hoc test"}
p$plot_tuckey
```

### Beta-diversity analysis

```{r}
if (!require("ggVennDiagram", quietly = TRUE)) {
  install.packages("ggVennDiagramÃ ")
}
ggvenn_pq(data_fungi, fact = "Height") +
  ggplot2::scale_fill_distiller(palette = "BuPu", direction = 1) +
  labs(title = "Share number of ASV among Height in tree")
```

### Note for non-linux users

Some functions may not work on windows (*e.g.* [track_wflow()], [cutadapt_remove_primers()], [krona()], [vsearch_clustering()], ...). A solution is to exploit docker container, for example the using the great [rocker project](https://rocker-project.org/).

MiscMetabar is developed under Linux and the vast majority of functions may works on Unix system, but its functionning is not test under iOS.

### Installation of other softwares for debian Linux distributions

If you encounter any errors or have any questions about the installation of these softwares, please visit their dedicated websites. 

#### [blast+](https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html#downloadblastdata)

```sh
sudo apt-get install ncbi-blast+
```

#### [vsearch](https://github.com/torognes/vsearch)

```sh
sudo apt-get install vsearch
```

An other possibilities is to [install vsearch](https://bioconda.github.io/recipes/vsearch/README.html?highlight=vsearch#package-package%20&#x27;vsearch&#x27;) with `conda`.

#### [swarm](https://github.com/torognes/swarm)

```sh
git clone https://github.com/torognes/swarm.git
cd swarm/
make
```

An other possibilities is to [install swarm](https://bioconda.github.io/recipes/swarm/README.html?highlight=swarm#package-package%20&#x27;swarm&#x27;) with `conda`.

#### [Mumu](https://github.com/frederic-mahe/mumu)

```sh
git clone https://github.com/frederic-mahe/mumu.git
cd ./mumu/
make
make check
make install  # as root or sudo
```

#### [cutadapt](https://cutadapt.readthedocs.io/en/stable/)

```sh
conda create -n cutadaptenv cutadapt
```
