---
title: "import_export_track"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{import_export_track}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE}
library(MiscMetabar)
library(formattable)
```

## Export phyloseq object

You can export a phyloseq object to csv (and txt for phylogenetic tree) files in a folder. It is possible to export each table into one file or to merge all slot (exept phytree) in one file (args `one_file = TRUE`). Finally, if `rdata` is set to TRUE, a `rdata` file containing the phyloseq object is also writed. 

```{r, eval = FALSE}
write_pq(data_fungi, path = "fungi_phyloseq")
write_pq(data_fungi, path = "fungi_phyloseq", one_file = TRUE)
write_pq(data_fungi, path = "fungi_phyloseq", rdata = TRUE)
```

Finally, you can use the function `save_pq()` to write the phyloseq object in all 3 versions (one table for each slot, a file merging each slot and an Rdata file).

```{r, eval = FALSE}
save_pq(data_fungi)
```

## Import

To import a Rdata file, juste use `load()` base function. In order to import phyloseq object from a folder create using `write_pq()` or `save_pq()`, please use the function `read_pq()`. 

```{r, eval = FALSE}
d <- read_pq(path = "fungi_phyloseq")
```

## Tracking sequences, clusters and samples

In bioinformatic pipeline, we often need to track the number of samples, sequences and clusters across step in the pipeline. MiscMetabar propose two utilities to achieve this goal : `track_wkflow()` and a version to compute value per samples : `track_wkflow_samples()`. The function `track_wkflow()` can deal with (i) fastq and fastg.gz files, dada-class object, derep-class object, matrix of samples x clusters (e.g. `otu_table`) and phyloseq-class object.

```{r}
track_wkflow(list(data_fungi, data_fungi_sp_known))
```