<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.11.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Changelog</h1>
      <small>Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.11" id="miscmetabar-011-in-development">MiscMetabar 0.11 (in development)<a class="anchor" aria-label="anchor" href="#miscmetabar-011-in-development"></a></h2>
<ul><li>Add function <code><a href="../reference/filt_taxa_pq.html">filt_taxa_pq()</a></code> to filter taxa based on the number of sequences/occurences</li>
<li>Add functions <code><a href="../reference/no_legend.html">no_legend()</a></code> and <code><a href="../reference/hill_curves_pq.html">hill_curves_pq()</a></code> to plot hill diversity accumulation curves for phyloseq</li>
<li>Add function <code><a href="../reference/umap_pq.html">umap_pq()</a></code> to compute Dimensionality Reduction with UMAP</li>
<li>Add function <code><a href="../reference/plot_complexity_pq.html">plot_complexity_pq()</a></code> to plot kmer complexity of references sequences of a phyloseq object</li>
<li>Add param <code>type</code> to <code>ridge_pq()</code> to plot a cumulative version (type=“ecdf”) version of ridge</li>
<li>Introduce the idea of a pq-verse: some other packages will complete the MiscMetabar packages to make package maintenance easier. The <a href="https://github.com/adrientaudiere/comparpq" class="external-link">comparpq</a> package will facilitate the comparison of phyloseq object with different taxonomy, different clustering method, different samples with same modality or different primers.</li>
<li>Add functions [assign_vsearch_lca()], [assign_sintax()] and internal function [write_temp_fasta()]</li>
<li>Add param <code>method</code> to <code><a href="../reference/add_new_taxonomy_pq.html">add_new_taxonomy_pq()</a></code> to allow the use of [dada2::assign_taxonomy()] (default, precedent only method available), [assign_sintax()] or [assign_vsearch_lca()]</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.10.4" id="miscmetabar-0104">MiscMetabar 0.10.4<a class="anchor" aria-label="anchor" href="#miscmetabar-0104"></a></h2>
<ul><li>Add functions <code><a href="../reference/plot_refseq_pq.html">plot_refseq_pq()</a></code> and <code><a href="../reference/plot_refseq_extremity_pq.html">plot_refseq_extremity_pq()</a></code> to plot the proportion of each nucleotide and the diversity of nucleotides from <code>@refseq</code> of a phyloseq object.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.10.3" id="miscmetabar-0103">MiscMetabar 0.10.3<a class="anchor" aria-label="anchor" href="#miscmetabar-0103"></a></h2>
<ul><li>Add params <code>type</code>, <code>na_remove</code> and <code>verbose</code> to <code><a href="../reference/ggvenn_pq.html">ggvenn_pq()</a></code>. The type = “nb_seq” allow to plot Venn diagram with the number of shared sequences instead of shared ASV.</li>
<li>Add automatic report in json for the function <code><a href="../reference/cutadapt_remove_primers.html">cutadapt_remove_primers()</a></code>.</li>
<li>Add param <code>verbose</code> to <code><a href="../reference/track_wkflow.html">track_wkflow()</a></code> and improve examples for <code><a href="../reference/track_wkflow.html">track_wkflow()</a></code> and <code>list_fastq_files</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.10.2" id="miscmetabar-0102">MiscMetabar 0.10.2<a class="anchor" aria-label="anchor" href="#miscmetabar-0102"></a></h2>
<ul><li>Improve code thanks to {lintr} package</li>
<li>Add option <code>return_file_path</code> to <code><a href="../reference/cutadapt_remove_primers.html">cutadapt_remove_primers()</a></code> in order to facilitate targets pipeline</li>
<li>Add function <code><a href="../reference/sam_data_matching_names.html">sam_data_matching_names()</a></code> to match and verify congruence between fastq files names and sample metadata (sam_data)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.10.1" id="miscmetabar-0101">MiscMetabar 0.10.1<a class="anchor" aria-label="anchor" href="#miscmetabar-0101"></a></h2><p class="text-muted">CRAN release: 2024-10-07</p>
<blockquote>
<p>CRAN 2024-09-10</p>
</blockquote>
<ul><li>Delete function <code>heat_tree_pq()</code> because {metacoder} package is archived from CRAN.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.4" id="miscmetabar-094">MiscMetabar 0.9.4<a class="anchor" aria-label="anchor" href="#miscmetabar-094"></a></h2>
<ul><li>Set a seed in the example of <code>build_tree_pq</code> to resubmit to CRAN Add a param <code>return_a_vector</code> in function <code><a href="../reference/filter_trim.html">filter_trim()</a></code> to make possible to return a vector of path as it is useful when used with <code>targets::tar_targets(..., format="file")</code>)</li>
<li>Make some storage amelioration by replacing <code><a href="https://rdrr.io/r/base/list.html" class="external-link">list()</a></code> by <code>vector(list, ...)</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.3" id="miscmetabar-093">MiscMetabar 0.9.3<a class="anchor" aria-label="anchor" href="#miscmetabar-093"></a></h2><p class="text-muted">CRAN release: 2024-09-09</p>
<blockquote>
<p>CRAN 2024-09-09</p>
</blockquote>
<ul><li>Homogenize terminology replacing ASV by taxa/taxon in documentation and code</li>
<li>Build an alias function <code><a href="../reference/filter_asv_blast.html">filter_taxa_blast()</a></code> for <code><a href="../reference/filter_asv_blast.html">filter_asv_blast()</a></code>
</li>
<li>Build an alias function <code><a href="../reference/postcluster_pq.html">postcluster_pq()</a></code> for <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code>
</li>
<li>Add param <code>return_data_for_venn</code> in function <code>ggvenn_pq</code> in order to make more customizable plot following <a href="https://gaospecial.github.io/ggVennDiagram/articles/fully-customed.html" class="external-link">ggVennDiagram tutorial</a>
</li>
</ul><div class="section level3">
<h3 id="breaking-changes-0-9-3">BREAKING CHANGES<a class="anchor" aria-label="anchor" href="#breaking-changes-0-9-3"></a></h3>
<ul><li>Replacing misnamed param <code>rename_asv</code> by <code>rename_taxons</code> in <code><a href="../reference/clean_pq.html">clean_pq()</a></code>
</li>
<li>Replacing misnamed param <code>reorder_asv</code> by <code>reorder_taxons</code> in <code><a href="../reference/clean_pq.html">clean_pq()</a></code>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.2" id="miscmetabar-092">MiscMetabar 0.9.2<a class="anchor" aria-label="anchor" href="#miscmetabar-092"></a></h2>
<ul><li>Add param <code>default_fun</code> in function <code><a href="../reference/merge_samples2.html">merge_samples2()</a></code> in order to replace the default function that change the sample data in case of merging. A useful parameter is <code>default_fun=diff_fct_diff_class</code>.</li>
<li>Add param <code>kruskal_test</code> to <code><a href="../reference/hill_pq.html">hill_pq()</a></code> function to prevent user to mis-interpret Tuckey HSD result (and letters) if the global effect of the tested factor on Hill diversity is non significant.</li>
<li>Add param <code>vioplot</code> to hill_pq() function to allow violin plot instead of boxplot.</li>
<li>Modify <code>rarefy_sample_count_by_modality</code> to debug the case of modality with level of length one.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.1" id="miscmetabar-091">MiscMetabar 0.9.1<a class="anchor" aria-label="anchor" href="#miscmetabar-091"></a></h2><p class="text-muted">CRAN release: 2024-04-28</p>
<blockquote>
<p>CRAN 2024-04-28</p>
</blockquote>
<div class="section level3">
<h3 id="new-functions-0-9-1">New functions<a class="anchor" aria-label="anchor" href="#new-functions-0-9-1"></a></h3>
<ul><li>Add functions <code><a href="../reference/taxa_as_rows.html">taxa_as_rows()</a></code> and <code><a href="../reference/taxa_as_columns.html">taxa_as_columns()</a></code> to replace verbose called to <code><a href="../reference/clean_pq.html">clean_pq()</a></code>
</li>
<li>Add function <code><a href="../reference/ggscatt_pq.html">ggscatt_pq()</a></code> to plot and test for effect of a numerical columns in sam_data on Hill number. Its the equivalent for numerical variables of <code><a href="../reference/ggbetween_pq.html">ggbetween_pq()</a></code> which focus on the effect of a factor.</li>
<li>Add functions <code><a href="../reference/var_par_pq.html">var_par_pq()</a></code> , <code><a href="../reference/var_par_rarperm_pq.html">var_par_rarperm_pq()</a></code> and <code><a href="../reference/plot_var_part_pq.html">plot_var_part_pq()</a></code> to compute the partition of the variation of community and plot it. It introduce the notion of <code>rarperm</code> part in the function name. It refers to the fact that this function compute permutation of samples depth rarefaction to measure the variation due to the random process in rarefaction.</li>
<li>Add function <code><a href="../reference/hill_test_rarperm_pq.html">hill_test_rarperm_pq()</a></code> to test the effect of a factor on hill diversity accounting for the variation due to random nature of the rarefaction by sample depth.</li>
<li>Add function <code><a href="../reference/rarefy_sample_count_by_modality.html">rarefy_sample_count_by_modality()</a></code> to equalize the number of samples for each levels of a modality (factor)</li>
<li>Add function <code><a href="../reference/accu_plot_balanced_modality.html">accu_plot_balanced_modality()</a></code> to plot accumulation curves with balanced modality (same number of samples per level) and depth rarefaction (same number of sequences per sample)</li>
<li>Add function <code><a href="../reference/adonis_rarperm_pq.html">adonis_rarperm_pq()</a></code> to compute multiple Permanova analyses on different sample depth rarefaction.</li>
<li>Add function <code><a href="../reference/ggaluv_pq.html">ggaluv_pq()</a></code> to plot taxonomic distribution in alluvial fashion with ggplot2 (using the [ggalluvial] package)</li>
<li>Add function <code><a href="../reference/glmutli_pq.html">glmutli_pq()</a></code> to use automated model selection and multimodel inference with (G)LMs for phyloseq object</li>
</ul></div>
<div class="section level3">
<h3 id="new-parameters-0-9-1">New parameters<a class="anchor" aria-label="anchor" href="#new-parameters-0-9-1"></a></h3>
<ul><li>Add param <code>taxa_ranks</code> in function <code><a href="../reference/psmelt_samples_pq.html">psmelt_samples_pq()</a></code> to group results by samples AND taxonomic ranks.</li>
<li>Add param <code>hill_scales</code> in functions <code><a href="../reference/hill_tuckey_pq.html">hill_tuckey_pq()</a></code> and <code>hill_p()</code> to choose the level of the hill number.</li>
<li>Add param <code>na_remove</code> in function <code><a href="../reference/hill_pq.html">hill_pq()</a></code> to remove samples with NA in the factor fact.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.1" id="miscmetabar-081">MiscMetabar 0.8.1<a class="anchor" aria-label="anchor" href="#miscmetabar-081"></a></h2>
<ul><li>Add param <code>plot_with_tuckey</code> to <code><a href="../reference/hill_pq.html">hill_pq()</a></code>.,</li>
<li>Add function <code><a href="../reference/formattable_pq.html">formattable_pq()</a></code> to make beautiful table of the distribution of taxa across a modality using visualization inside in the table.</li>
<li>Add functions <code><a href="../reference/fac2col.html">fac2col()</a></code> and <code><a href="../reference/transp.html">transp()</a></code> to facilitate manipulation of colors, especially in function <code><a href="../reference/formattable_pq.html">formattable_pq()</a></code>
</li>
<li>Add functions <code><a href="../reference/signif_ancombc.html">signif_ancombc()</a></code> and <code><a href="../reference/plot_ancombc_pq.html">plot_ancombc_pq()</a></code> to plot significant results from <code><a href="../reference/ancombc_pq.html">ancombc_pq()</a></code> function</li>
<li>Add function <code><a href="../reference/distri_1_taxa.html">distri_1_taxa()</a></code> to summarize the distribution of one given taxa across level of a modality</li>
<li>Add function <code><a href="../reference/normalize_prop_pq.html">normalize_prop_pq()</a></code> to implement the method proposed by <a href="https://doi.org/10.5061/dryad.tn8qs35" class="external-link">McKnight et al. 2018</a>
</li>
<li>Add function <code><a href="../reference/psmelt_samples_pq.html">psmelt_samples_pq()</a></code> to build data frame of samples information including the number of sequences (Abundance) and Hill diversity metrics. Useful to use with the <a href="https://indrajeetpatil.github.io/ggstatsplot/" class="external-link">ggstatsplot</a> packages (see examples).</li>
<li>Replace param <code>variable</code> by <code>fact</code> in function <code><a href="../reference/ggbetween_pq.html">ggbetween_pq()</a></code> and <code><a href="../reference/hill_pq.html">hill_pq()</a></code> (keeping the variable option in <code><a href="../reference/hill_pq.html">hill_pq()</a></code> for backward compatibility)</li>
<li>Fix a bug in the class of the return object of function <code><a href="../reference/chimera_removal_vs.html">chimera_removal_vs()</a></code>. Now it return a matrix to be able to be parsed on to [dada2::getUniques()]</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7" id="miscmetabar-07">MiscMetabar 0.7<a class="anchor" aria-label="anchor" href="#miscmetabar-07"></a></h2>
<blockquote>
<p>CRAN 2024-03-08</p>
</blockquote>
<ul><li><p>Add functions <code><a href="../reference/chimera_detection_vs.html">chimera_detection_vs()</a></code> and <code><a href="../reference/chimera_removal_vs.html">chimera_removal_vs()</a></code> to process chimera detection and removal using <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a> software</p></li>
<li><p>Add functions <code><a href="../reference/filter_trim.html">filter_trim()</a></code>, <code><a href="../reference/sample_data_with_new_names.html">sample_data_with_new_names()</a></code> and <code><a href="../reference/rename_samples.html">rename_samples()</a></code> to facilitate the use of <a href="https://books.ropensci.org/targets/" class="external-link">targets</a> for bioinformatic pipeline.</p></li>
<li><p>Add function <code><a href="../reference/add_info_to_sam_data.html">add_info_to_sam_data()</a></code> to expand sam_data slot using a data.frame and using nb_asv and nb_seq</p></li>
<li><p>Add functions <code><a href="../reference/swarm_clustering.html">swarm_clustering()</a></code> and <code><a href="../reference/vsearch_clustering.html">vsearch_clustering()</a></code> and add <code>swarm</code> method in the function <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code></p></li>
<li><p>Add function <code><a href="../reference/physeq_or_string_to_dna.html">physeq_or_string_to_dna()</a></code> mostly for internal use</p></li>
<li><p>Add function <code><a href="../reference/cutadapt_remove_primers.html">cutadapt_remove_primers()</a></code> to remove primers using <a href="https://github.com/marcelm/cutadapt/" class="external-link">cutadapt</a></p></li>
<li><p>Add internal functions <code><a href="../reference/is_swarm_installed.html">is_swarm_installed()</a></code>, <code><a href="../reference/is_cutadapt_installed.html">is_cutadapt_installed()</a></code>, <code><a href="../reference/is_vsearch_installed.html">is_vsearch_installed()</a></code> and <code><a href="../reference/is_falco_installed.html">is_falco_installed()</a></code> to test for the availability of external software in order to run examples and test from testthat.</p></li>
<li><p>Submit to CRAN and change code to comply with their rules (patch 0.7.1 to 0.7.9)</p></li>
<li><p>Numerous examples and tests are skipped on CRAN because it spends to much time to run. Rules vignettes is updated to details the strategy for this.</p></li>
</ul><div class="section level3">
<h3 id="breaking-changes-0-7">BREAKING CHANGES<a class="anchor" aria-label="anchor" href="#breaking-changes-0-7"></a></h3>
<ul><li>Harmonization of parameters names:
<ul><li>
<code>add_nb_sequences</code> -&gt; <code>add_nb_seq</code> in <code><a href="../reference/ggvenn_pq.html">ggvenn_pq()</a></code>
</li>
<li>
<code>db</code> -&gt; <code>db_url</code> in <code><a href="../reference/get_funguild_db.html">get_funguild_db()</a></code>
</li>
<li>
<code>db</code> -&gt; <code>db_funguild</code> in <code><a href="../reference/get_funguild_db.html">get_funguild_db()</a></code>
</li>
<li>
<code>file</code> -&gt; <code>file_path</code> in <code><a href="../reference/get_file_extension.html">get_file_extension()</a></code>
</li>
<li>
<code>n_seq</code> -&gt; <code>nb_seq</code> in <code><a href="../reference/subsample_fastq.html">subsample_fastq()</a></code>
</li>
<li>
<code>otutable</code> -&gt; <code>otu_table</code> in <code><a href="../reference/lulu.html">lulu()</a></code>
</li>
<li>
<code>alpha</code> -&gt; <code>pval</code> in <code><a href="../reference/plot_edgeR_pq.html">plot_edgeR_pq()</a></code> and <code><a href="../reference/plot_deseq2_pq.html">plot_deseq2_pq()</a></code> and change default value from 0.01 to more classical 0.05</li>
<li>
<code>sequences</code> -&gt; <code>seq2search</code> in function <code><a href="../reference/search_exact_seq_pq.html">search_exact_seq_pq()</a></code>
</li>
<li>
<code>seq_names</code> -&gt; <code>dna_seq</code> in function <code>asv2otu</code>
</li>
</ul></li>
<li>Removing the function <code>install_pkg_needed()</code> which do not comply with CRAN policies</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.0" id="miscmetabar-060">MiscMetabar 0.6.0<a class="anchor" aria-label="anchor" href="#miscmetabar-060"></a></h2>
<ul><li>Add function <code><a href="../reference/ancombc_pq.html">ancombc_pq()</a></code> to simplify the call to <code><a href="https://rdrr.io/pkg/ANCOMBC/man/ancombc2.html" class="external-link">ANCOMBC::ancombc2()</a></code> : ANalysis of COmpositions of Microbiomes with Bias Correction 2</li>
<li>Add param <code>taxa_names_from_physeq</code> (default FALSE) to <code><a href="../reference/subset_taxa_pq.html">subset_taxa_pq()</a></code>
</li>
<li>Add param <code>rarefy_by_sample</code> (default FALSE) to function <code><a href="../reference/ggbetween_pq.html">ggbetween_pq()</a></code>
</li>
<li>Add function <code><a href="../reference/are_modality_even_depth.html">are_modality_even_depth()</a></code> to test if samples depth significantly vary among the modalities of a factor</li>
<li>Add functions <code><a href="../reference/merge_taxa_vec.html">merge_taxa_vec()</a></code> and <code><a href="../reference/merge_samples2.html">merge_samples2()</a></code> from the <a href="https://github.com/mikemc/speedyseq/" class="external-link">speedyseq</a> package into MiscMetabar to decrease package dependencies (Thanks to Mike R. Mclaren)</li>
<li>Add function <code><a href="../reference/reorder_taxa_pq.html">reorder_taxa_pq()</a></code> in order to replace the unique call to package MicroViz to decrease package dependencies.</li>
<li>Add functions <code><a href="../reference/get_funguild_db.html">get_funguild_db()</a></code> and <code><a href="../reference/funguild_assign.html">funguild_assign()</a></code> from the <a href="https://github.com/brendanf/FUNGuildR/" class="external-link">FUNGuildR</a> package into MiscMetabar to decrease package dependencies</li>
<li>Remove all dependencies from packages not available on CRAN or Bioconductor. Improve code using <code><a href="https://docs.ropensci.org/goodpractice/reference/gp.html" class="external-link">goodpractice::gp</a></code>() and <code><a href="https://devtools.r-lib.org/reference/check.html" class="external-link">devtools::check()</a></code> function</li>
<li>Add messages in various cases (NA in samples data, low number of sequences in samples, low number of sequences by taxa) when using <code><a href="../reference/verify_pq.html">verify_pq()</a></code> with args <code>verbose=TRUE</code>
</li>
<li>Fix a bug in <code><a href="../reference/multitax_bar_pq.html">multitax_bar_pq()</a></code> when using <code>nb_seq = FALSE</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.52" id="miscmetabar-052">MiscMetabar 0.52<a class="anchor" aria-label="anchor" href="#miscmetabar-052"></a></h2>
<ul><li>Add function <code><a href="../reference/ggbetween_pq.html">ggbetween_pq()</a></code> to facilitate comparison of hill number using the power of <code><a href="https://indrajeetpatil.github.io/ggstatsplot/reference/ggbetweenstats.html" class="external-link">ggstatsplot::ggbetweenstats()</a></code>
</li>
<li>Add function <code><a href="../reference/plot_SCBD_pq.html">plot_SCBD_pq()</a></code> to plot species contributions to beta diversity (SCBD) of samples</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.51" id="miscmetabar-051">MiscMetabar 0.51<a class="anchor" aria-label="anchor" href="#miscmetabar-051"></a></h2>
<ul><li>Add function <code><a href="../reference/LCBD_pq.html">LCBD_pq()</a></code> and <code><a href="../reference/plot_LCBD_pq.html">plot_LCBD_pq()</a></code> to compute, test and plot local contributions to beta diversity (LCBD) of samples</li>
<li>Add function <code><a href="../reference/tbl_sum_samdata.html">tbl_sum_samdata()</a></code> to summarize information from sample data in a table</li>
<li>Add function <code><a href="../reference/mumu_pq.html">mumu_pq()</a></code> to use <a href="https://github.com/frederic-mahe/mumu" class="external-link">mumu</a>, a fast and robust C++ implementation of lulu.</li>
<li>Add (a mostly internal) function <code>install_pkg_needed()</code> to install pkg (mostly for package list in <em>Suggest</em> in DESCRIPTION) if needed by a function.</li>
<li>Add function <code><a href="../reference/add_funguild_info.html">add_funguild_info()</a></code> and <code><a href="../reference/plot_guild_pq.html">plot_guild_pq()</a></code> to add and plot fungal guild information from taxonomy using <code>FUNGuild</code> package</li>
<li>Add function <code><a href="../reference/build_phytree_pq.html">build_phytree_pq()</a></code> to build 3 phylogenetic trees (NJ, UPGMA and ML using <code>phangorn</code> R package) from the <code>refseq</code> slot of a <code>phyloseq</code> object, possibly with bootstrap values. See the vignettes <a href="https://adrientaudiere.github.io/MiscMetabar/articles/tree_visualization.html">Tree visualization</a> for an introduction to tree visualization using <code>ggtree</code> R package.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5" id="miscmetabar-05">MiscMetabar 0.5<a class="anchor" aria-label="anchor" href="#miscmetabar-05"></a></h2>
<ul><li>Phyloseq object are converted in taxa_are_columns in the <code><a href="../reference/ggvenn_pq.html">ggvenn_pq()</a></code> thanks to issue <a href="https://github.com/adrientaudiere/MiscMetabar/issues/31" class="external-link">#31</a></li>
</ul><div class="section level3">
<h3 id="breaking-changes-0-5">BREAKING CHANGES<a class="anchor" aria-label="anchor" href="#breaking-changes-0-5"></a></h3>
<ul><li>Rename param <code>log_10</code> in function <code><a href="../reference/biplot_pq.html">biplot_pq()</a></code> into <code>log10trans</code>
</li>
<li>Rename param <code>log10transform</code> in function <code><a href="../reference/circle_pq.html">circle_pq()</a></code> into <code>log10trans</code>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.42" id="miscmetabar-042">MiscMetabar 0.42<a class="anchor" aria-label="anchor" href="#miscmetabar-042"></a></h2>
<ul><li>Add argument <code>one_plot</code> (default FALSE, same behavior than before) to <code>hill_pq</code> function in order to return an unique ggplot2 object with the four plots inside.</li>
<li>Add argument <code>correction_for_sample_size</code> (default TRUE, same behavior than before) to <code>hill_pq</code> and <code>hill_tuckey_pq</code> function to allow removing any correction for uneven sampling depth.</li>
<li>Add function <code><a href="../reference/multitax_bar_pq.html">multitax_bar_pq()</a></code> to plot 3 levels of taxonomy in function of samples attributes</li>
<li>Add function <code><a href="../reference/ridges_pq.html">ridges_pq()</a></code> to plot ridges of one taxonomic level in function of samples attributes</li>
<li>Add function <code>treemap_pq</code> to plot treemap of two taxonomic levels</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.41" id="miscmetabar-041">MiscMetabar 0.41<a class="anchor" aria-label="anchor" href="#miscmetabar-041"></a></h2>
<ul><li>Add function <code><a href="../reference/iNEXT_pq.html">iNEXT_pq()</a></code> to calculate hill diversity using the <a href="https://github.com/AnneChao/iNEXT" class="external-link">iNEXT</a> package.</li>
<li>Add argument <code>pairs</code> to <code><a href="../reference/multi_biplot_pq.html">multi_biplot_pq()</a></code> in order to indicate all pairs of samples we want to print.</li>
<li>Improve <code><a href="../reference/compare_pairs_pq.html">compare_pairs_pq()</a></code> with information about the number of shared sequences among pairs.</li>
<li>Add function <code><a href="../reference/upset_pq.html">upset_pq()</a></code> to plot upset of phyloseq object using the <a href="https://krassowski.github.io/complex-upset/" class="external-link">ComplexUpset</a> package.</li>
<li>Add function <code>upset_test_pq</code> to test for differences between intersections (wrapper of <code><a href="https://krassowski.github.io/complex-upset/reference/upset_test.html" class="external-link">ComplexUpset::upset_test()</a></code> for <code>phyloseq-object</code>).</li>
<li>Add info (param <code>add_info</code>) in subtitle of the <code><a href="../reference/hill_pq.html">hill_pq()</a></code> function.</li>
<li>Add argument <code>remove_space</code> to <code><a href="../reference/simplify_taxo.html">simplify_taxo()</a></code> function.</li>
<li>Add argument <code>simplify_taxo</code> to <code><a href="../reference/clean_pq.html">clean_pq()</a></code> function.</li>
<li>Change argument <code>rarefy_nb_seq</code> by <code>rarefy_before_merging</code> and add arguments <code>rarefy_after_merging</code> and <code>add_nb_seq</code> to <code><a href="../reference/ggvenn_pq.html">ggvenn_pq()</a></code> function.</li>
<li>Add arguments <code>rarefy_after_merging</code> to <code><a href="../reference/biplot_pq.html">biplot_pq()</a></code> and <code><a href="../reference/upset_pq.html">upset_pq()</a></code> functions.</li>
<li>Add argument <code>taxa_fill</code> to <code><a href="../reference/upset_pq.html">upset_pq()</a></code> function in order to fill the bar with taxonomic rank.</li>
<li>Add a function <code><a href="../reference/subsample_fastq.html">subsample_fastq()</a></code> to make subset of fastq files in order to test your pipeline with all samples but with a low number of reads.</li>
<li>Add a function <code><a href="../reference/accu_samp_threshold.html">accu_samp_threshold()</a></code> to compute the number of sequence to obtain a given proportion of ASV in accumulation curves (`accu_plot).</li>
<li>Add a function <code><a href="../reference/tax_bar_pq.html">tax_bar_pq()</a></code> in order to plot taxonomic distribution across samples.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.40" id="miscmetabar-040">MiscMetabar 0.40<a class="anchor" aria-label="anchor" href="#miscmetabar-040"></a></h2>
<ul><li>Add function <code><a href="../reference/multi_biplot_pq.html">multi_biplot_pq()</a></code> to visualize a collection of couples of samples for comparison through a list of <code><a href="../reference/biplot_pq.html">biplot_pq()</a></code>.</li>
<li>Add options <code>add_info</code>, <code>na_remove</code>, and <code>clean_pq</code> to <code><a href="../reference/plot_tax_pq.html">plot_tax_pq()</a></code> function.</li>
<li>Add options <code>vsearch_cluster_method</code> and <code>vsearch_args</code> to <code>otu2asv()</code> for more detailed control of the vsearch software.</li>
<li>Suppression of buggy function <code>MM_idtaxa()</code>.</li>
<li>Add a wrapper of <code><a href="../reference/write_pq.html">write_pq()</a></code> called <code><a href="../reference/save_pq.html">save_pq()</a></code> to save a <em>phyloseq</em> object in the three possible formats () at the same time
<ul><li>4 separate tables</li>
<li>1 table version</li>
<li>1 RData file</li>
</ul></li>
<li>Add a function <code><a href="../reference/add_blast_info.html">add_blast_info()</a></code> to add information from <code><a href="../reference/blast_pq.html">blast_pq()</a></code> to the <code>tax_table</code> slot of a <em>phyloseq</em> object.</li>
<li>Add option <code>keep_temporary_files</code> in <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code> function.</li>
<li>Improve the documentation of <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code> and fix a little bug in the name of the conserved ASV after <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code>.</li>
<li>Test coverage largely improved leading to numerous minor bug fixes.</li>
<li>Add function <code><a href="../reference/search_exact_seq_pq.html">search_exact_seq_pq()</a></code> to search for exact matching of sequences using complement, reverse and reverse-complement against a phyloseq object.</li>
<li>Add function <code><a href="../reference/add_new_taxonomy_pq.html">add_new_taxonomy_pq()</a></code> to add new taxonomic rank to a phyloseq object. For example to add taxonomic assignment from a new database.</li>
<li>Add a battery of test using <code>test_that</code> package and improve code compatibility with cran recommendations.</li>
</ul><div class="section level3">
<h3 id="breaking-changes-0-40">BREAKING CHANGES<a class="anchor" aria-label="anchor" href="#breaking-changes-0-40"></a></h3>
<ul><li>
<code><a href="../reference/postcluster_pq.html">asv2otu()</a></code> with <code>method="vsearch"</code> change two default values (to repeat the precedent behavior, use <code>asv2otu(..., vsearch_cluster_method = "--cluster_fast", tax_adjust = 1)</code>):
<ul><li>vsearch_cluster_method = “–cluster_size”</li>
<li>tax_adjust = 0</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.34" id="miscmetabar-034">MiscMetabar 0.34<a class="anchor" aria-label="anchor" href="#miscmetabar-034"></a></h2>
<ul><li>Add option <code>add_nb_samples</code> to <code><a href="../reference/ggvenn_pq.html">ggvenn_pq()</a></code> which add the number of samples to level name in the plot. Useful to see disequilibrium in the number of samples among the factor’s levels.</li>
<li>Add option <code>args_makedb</code> and <code>args_blastn</code> to functions <code><a href="../reference/blast_pq.html">blast_pq()</a></code>, <code><a href="../reference/blast_to_phyloseq.html">blast_to_phyloseq()</a></code>, <code><a href="../reference/blast_to_derep.html">blast_to_derep()</a></code> and <code><a href="../reference/filter_asv_blast.html">filter_asv_blast()</a></code>.</li>
<li>Add option <code>rarefy_nb_seqs</code> to <code>ggven_pq()</code> in order to rarefy samples before plotting.</li>
<li>Add function <code><a href="../reference/SRS_curve_pq.html">SRS_curve_pq()</a></code> to plot scaling with ranked subsampling (SRS) curves using the <code>SRS::SRS_curve()</code> function (see citation(“SRS”) for reference).</li>
<li>Add option <code>nb_samples_info</code> to <code><a href="../reference/biplot_pq.html">biplot_pq()</a></code> in order to add the number of samples merged by level of factors.</li>
<li>Add a message when two modalities differ greatly (more than x2) in their number of sequences in <code><a href="../reference/biplot_pq.html">biplot_pq()</a></code> and <code><a href="../reference/ggvenn_pq.html">ggvenn_pq()</a></code>.</li>
<li>Add options <code>na_remove</code>, <code>dist_method</code> (including Aitchinson and robust-Aitchinson distance), <code>correction_for_sample_size</code> and <code>rarefy_nb_seqs</code> options to <code><a href="../reference/adonis_pq.html">adonis_pq()</a></code> function.</li>
<li>Add option <code>na_remove</code> to <code><a href="../reference/graph_test_pq.html">graph_test_pq()</a></code> function.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.33" id="miscmetabar-033">MiscMetabar 0.33<a class="anchor" aria-label="anchor" href="#miscmetabar-033"></a></h2>
<ul><li>Add function <code><a href="../reference/plot_tax_pq.html">plot_tax_pq()</a></code> to plot taxonomic distribution (nb of sequences or nb of ASV) across factor.</li>
<li>Add option <code>add_points</code> and make better axis of <code><a href="../reference/hill_pq.html">hill_pq()</a></code> function</li>
<li>Add function <code><a href="../reference/blast_to_derep.html">blast_to_derep()</a></code> in order to facilitate searching some fasta sequences in dereplicated sequences (obtained by <code><a href="https://rdrr.io/pkg/dada2/man/derepFastq.html" class="external-link">dada2::derepFastq</a></code>)</li>
</ul><table class="table"><colgroup><col width="21%"><col width="45%"><col width="33%"></colgroup><thead><tr><th></th>
<th>Database (makeblastdb)</th>
<th>Sequences to blast (blastn)</th>
</tr></thead><tbody><tr><td><code><a href="../reference/blast_to_phyloseq.html">blast_to_phyloseq()</a></code></td>
<td>Built from <code>ref_seq</code> slot(physeq-class)</td>
<td>Custom fasta file</td>
</tr><tr><td><code><a href="../reference/blast_to_derep.html">blast_to_derep()</a></code></td>
<td>Built from dereplicate sequences (derep-class)</td>
<td>Custom fasta file</td>
</tr><tr><td><code><a href="../reference/blast_pq.html">blast_pq()</a></code></td>
<td>Custom database or custom fasta file</td>
<td>
<code>ref_seq</code> slot of a physeq object</td>
</tr></tbody></table><ul><li>Add functions <code><a href="../reference/tsne_pq.html">tsne_pq()</a></code> and <code><a href="../reference/plot_tsne_pq.html">plot_tsne_pq()</a></code> to quickly visualize results of the t-SNE multidimensional analysis based on the <code><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne::Rtsne()</a></code> function.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.32" id="miscmetabar-032">MiscMetabar 0.32<a class="anchor" aria-label="anchor" href="#miscmetabar-032"></a></h2>
<ul><li>Add the possibility to select a folder in the function <code><a href="../reference/count_seq.html">count_seq()</a></code>
</li>
<li>Add functions <code><a href="../reference/track_wkflow_samples.html">track_wkflow_samples()</a></code> and <code><a href="../reference/select_one_sample.html">select_one_sample()</a></code>
</li>
<li>Add option <code>sam_data_first</code> in function <code><a href="../reference/write_pq.html">write_pq()</a></code>
</li>
<li>Add option <code>reorder_asv</code> and <code>rename_asv</code> to in function <code><a href="../reference/write_pq.html">write_pq()</a></code> and <code>clean_pq</code>
</li>
<li>Add a function <code><a href="../reference/rotl_pq.html">rotl_pq()</a></code> to build a phylogenetic tree using the ASV binomial names of a physeq object and the Open Tree of Life tree.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.31" id="miscmetabar-031">MiscMetabar 0.31<a class="anchor" aria-label="anchor" href="#miscmetabar-031"></a></h2>
<ul><li>Argument <code>split_by</code> to make multiple plot given a variable in <code>sam_data</code> slot (function <code><a href="../reference/ggvenn_pq.html">ggvenn_pq()</a></code>)<br></li>
<li>Argument <code>seq_names</code> in <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code> function allow to clusterize sequences from a character vector of DNA.</li>
<li>Add a <code><a href="../reference/blast_pq.html">blast_pq()</a></code> function to blast the sequences of the <code>@ref_seq</code> slot against a custom database</li>
<li>Add a <code><a href="../reference/filter_asv_blast.html">filter_asv_blast()</a></code> function to filter ASV in <em>phyloseq</em> dataset using blast against a custom database</li>
<li>Add a <code><a href="../reference/subset_taxa_pq.html">subset_taxa_pq()</a></code> function to filter ASV based on a named conditional vector. Used in <code><a href="../reference/filter_asv_blast.html">filter_asv_blast()</a></code>.</li>
<li>Add parameter <code>force_taxa_as_columns</code> (default FALSE) and <code>force_taxa_as_rows</code> (default FALSE) to <code><a href="../reference/clean_pq.html">clean_pq()</a></code>.</li>
<li>Add a first version of the function <code>count_fastq_seq()</code> to count sequences from fastq.gz files directly from R.</li>
<li>Add taxonomic info to <code><a href="../reference/track_wkflow.html">track_wkflow()</a></code> function (parameter <code>taxonomy_rank</code>)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3" id="miscmetabar-03">MiscMetabar 0.3<a class="anchor" aria-label="anchor" href="#miscmetabar-03"></a></h2>
<ul><li>Change some function names, mainly replacing <code>physeq</code> by <code>pk</code>. </li>
<li>Improve documentation using some rules documented in the Rules vignettes.</li>
<li>Add a option <code>sam_names()</code> to <code><a href="../reference/read_pq.html">read_pq()</a></code>
</li>
<li>Correction of <code>data_fungi</code> and <code>data_fungi_sp_known</code> metadata</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.24" id="miscmetabar-024">MiscMetabar 0.24<a class="anchor" aria-label="anchor" href="#miscmetabar-024"></a></h2>
<ul><li>Add supplementary info in summary_plot_physeq()`</li>
<li>Better arguments in biplot_physeq()`)</li>
<li>Add merge_sample_by argument in biplot_physeq()`</li>
<li>Better documentation with more example.</li>
<li>For other minors bugs fixes and addition, see the list of commits</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.23" id="miscmetabar-023">MiscMetabar 0.23<a class="anchor" aria-label="anchor" href="#miscmetabar-023"></a></h2>
<ul><li><p>Adapt the function <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code> to <em>IdClusters</em> change in the DECIPHER package (commit 254100922f2093cc789d018c18a26752a3cda1e3). Then change the <em>IdClusters</em> function that was removed from DECIPHER to <em>Clusterize</em> function.</p></li>
<li><p>Better functioning of <code><a href="../reference/blast_to_phyloseq.html">blast_to_phyloseq()</a></code> when none query sequences are founded.</p></li>
<li><p>Add <em>tax_adjust</em> argument to <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code>function</p></li>
<li><p>Add some functions useful for the targets package</p></li>
<li><p>Add a <code><a href="../reference/MiscMetabar-deprecated.html">biplot_physeq()</a></code> function to visualize of two samples for comparison of physeq object</p></li>
<li><p>Add an argument <em>modality</em> in the <code><a href="../reference/tax_datatable.html">tax_datatable()</a></code> function to split OTU abundancy by level of the sample modality</p></li>
<li><p>Add a function <code>multiple_share_bisamples()</code> to help compare samples by pairs</p></li>
<li><p>Add a new function (<code><a href="../reference/MiscMetabar-deprecated.html">ggVenn_phyloseq()</a></code>) for better venn diagram but without area calculation (use <code><a href="../reference/MiscMetabar-deprecated.html">venn_phyloseq()</a></code> in this case).</p></li>
<li><p>Add two functions helpful for beta-diversity analysis (<code><a href="../reference/MiscMetabar-deprecated.html">adonis_phyloseq()</a></code> and <code><a href="../reference/MiscMetabar-deprecated.html">physeq_graph_test()</a></code>)</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.22" id="miscmetabar-022">MiscMetabar 0.22<a class="anchor" aria-label="anchor" href="#miscmetabar-022"></a></h2>
<ul><li>Add badge to set the development lifecycle of each function</li>
<li>Add the lulu_phyloseq function to make easy the reclustering of phyloseq object using the lulu algorithm (<a href="https://www.nature.com/articles/s41467-017-01312-x" class="external-link uri">https://www.nature.com/articles/s41467-017-01312-x</a>) from the <a href="https://github.com/adrientaudiere/lulu" class="external-link">lulu package</a>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.21" id="miscmetabar-021">MiscMetabar 0.21<a class="anchor" aria-label="anchor" href="#miscmetabar-021"></a></h2>
<ul><li>This is the first release of pkgdown.</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

