<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="
This function build tree phylogenetic tree and if nb_bootstrap is
set, it build also the 3 corresponding bootstrapped tree.
Default parameters are based on https://doi.org/10.12688/f1000research.8986.2
and phangorn vignette Estimating phylogenetic trees with phangorn. You should understand your data, especially the markers,
before using this function.
Note that phylogenetic reconstruction with markers used for metabarcoding are
not robust. You must verify the robustness of your phylogenetic tree using
taxonomic classification (see vignette Tree visualization) and bootstrap or multi-tree visualization"><title>Build phylogenetic trees from refseq slot of a phyloseq object — build_phytree_pq • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Build phylogenetic trees from refseq slot of a phyloseq object — build_phytree_pq"><meta property="og:description" content="
This function build tree phylogenetic tree and if nb_bootstrap is
set, it build also the 3 corresponding bootstrapped tree.
Default parameters are based on https://doi.org/10.12688/f1000research.8986.2
and phangorn vignette Estimating phylogenetic trees with phangorn. You should understand your data, especially the markers,
before using this function.
Note that phylogenetic reconstruction with markers used for metabarcoding are
not robust. You must verify the robustness of your phylogenetic tree using
taxonomic classification (see vignette Tree visualization) and bootstrap or multi-tree visualization"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Build phylogenetic trees from refseq slot of a phyloseq object</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/dada_phyloseq.R" class="external-link"><code>R/dada_phyloseq.R</code></a></small>
      <div class="d-none name"><code>build_phytree_pq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a></p>
<p>This function build tree phylogenetic tree and if nb_bootstrap is
set, it build also the 3 corresponding bootstrapped tree.</p>
<p>Default parameters are based on https://doi.org/10.12688/f1000research.8986.2
and phangorn vignette <a href="https://klausvigo.github.io/phangorn/articles/Trees.html" class="external-link">Estimating phylogenetic trees with phangorn</a>. You should understand your data, especially the markers,
before using this function.</p>
<p>Note that phylogenetic reconstruction with markers used for metabarcoding are
not robust. You must verify the robustness of your phylogenetic tree using
taxonomic classification (see vignette <a href="https://adrientaudiere.github.io/MiscMetabar/articles/tree_visualization.html">Tree visualization</a>) and bootstrap or multi-tree visualization</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">build_phytree_pq</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  nb_bootstrap <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"GTR"</span>,</span>
<span>  optInv <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  optGamma <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rearrangement <span class="op">=</span> <span class="st">"NNI"</span>,</span>
<span>  control <span class="op">=</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml.control.html" class="external-link">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>  optNni <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  multicore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>physeq</dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt>nb_bootstrap</dt>
<dd><p>(default 0): If a positive number is set,
the function also build 3 bootstrapped trees using <code>nb_bootstrap</code>
bootstrap samples</p></dd>


<dt>model</dt>
<dd><p>allows to choose an amino acid models or nucleotide model,
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt>optInv</dt>
<dd><p>Logical value indicating whether topology gets optimized
(NNI). See <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt>optGamma</dt>
<dd><p>Logical value indicating whether gamma rate parameter gets
optimized. See <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt>rearrangement</dt>
<dd><p>type of tree tree rearrangements to perform, one of
"NNI", "stochastic" or "ratchet"
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt>control</dt>
<dd><p>A list of parameters for controlling the fitting process.
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt>optNni</dt>
<dd><p>Logical value indicating whether topology gets optimized (NNI).
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt>multicore</dt>
<dd><p>(logical) whether models should estimated in parallel.
see <code><a href="https://klausvigo.github.io/phangorn/reference/bootstrap.pml.html" class="external-link">phangorn::bootstrap.pml()</a></code> for more details</p></dd>


<dt>...</dt>
<dd><p>other params for be passed on to
<code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> function</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list of phylogenetic tree</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is mainly a wrapper of the work of others.
Please make a reference to <code>phangorn</code> package if you
use this function.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/KlausVigo/phangorn" class="external-link">"phangorn"</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: ape</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘ape’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:dplyr’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     where</span>
<span class="r-in"><span><span class="co"># data(data_fungi)</span></span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="subset_taxa_pq.html">subset_taxa_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">9000</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 15 sample(s) ( BE9-006-B_S27_MERGED.fastq.gz / C21-NV1-M_S64_MERGED.fastq.gz / DJ2-008-B_S87_MERGED.fastq.gz / DY5-004-H_S97_MERGED.fastq.gz / DY5-004-M_S98_MERGED.fastq.gz / E9-009-B_S100_MERGED.fastq.gz / E9-009-H_S101_MERGED.fastq.gz / N22-001-B_S129_MERGED.fastq.gz / O20-X-B_S139_MERGED.fastq.gz / O21-007-M_S144_MERGED.fastq.gz / R28-008-H_S159_MERGED.fastq.gz / R28-008-M_S160_MERGED.fastq.gz / W26-001-M_S167_MERGED.fastq.gz / Y29-007-H_S182_MERGED.fastq.gz / Y29-007-M_S183_MERGED.fastq.gz ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 1379</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 41</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 170</span>
<span class="r-in"><span><span class="va">df_tree</span> <span class="op">&lt;-</span> <span class="fu">build_phytree_pq</span><span class="op">(</span><span class="va">df</span>, nb_bootstrap <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on shared 8-mers:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.01 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.01 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Aligning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.27 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iteration 1 of 2:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reclustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.01 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Realigning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.2 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iteration 2 of 2:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reclustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.01 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Realigning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.16 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Refining the alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.05 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -6910.742 --&gt; -6772.959 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -6772.959 --&gt; -6772.703 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -6772.703 --&gt; -6768.532  NNI moves:  2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -6768.532 --&gt; -6768.518 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -6768.518 --&gt; -6768.518  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -6768.518 --&gt; -6768.518 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7505.127 --&gt; -7312.42 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7312.42 --&gt; -7312.413 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7312.413 --&gt; -7288.834  NNI moves:  10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7288.834 --&gt; -7288.829 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7288.829 --&gt; -7288.829  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7288.829 --&gt; -7288.829 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7469.217 --&gt; -7295.771 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7295.771 --&gt; -7295.763 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7295.763 --&gt; -7286.68  NNI moves:  4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7286.68 --&gt; -7286.68 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7286.68 --&gt; -7286.68  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7286.68 --&gt; -7286.68 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7490.809 --&gt; -7282.507 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7282.507 --&gt; -7282.506 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7282.506 --&gt; -7269.593  NNI moves:  5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7269.593 --&gt; -7269.593 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7269.593 --&gt; -7269.593  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7269.593 --&gt; -7269.593 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7644.058 --&gt; -7474.304 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7474.304 --&gt; -7474.299 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7474.299 --&gt; -7464.905  NNI moves:  4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7464.905 --&gt; -7464.903 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -7464.903 --&gt; -7464.903  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -7464.903 --&gt; -7464.903 </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">UPGMA</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">UPGMA</span>, <span class="va">df_tree</span><span class="op">$</span><span class="va">UPGMA_bs</span>, main <span class="op">=</span> <span class="st">"UPGMA"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">NJ</span>, <span class="st">"unrooted"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># plotBS(df_tree$ML_bs)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">$</span><span class="va">tree</span>, <span class="va">df_tree</span><span class="op">$</span><span class="va">ML_bs</span>, p <span class="op">=</span> <span class="fl">20</span>, frame <span class="op">=</span> <span class="st">"circle"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">$</span><span class="va">tree</span>,</span></span>
<span class="r-in"><span>  <span class="va">df_tree</span><span class="op">$</span><span class="va">ML_bs</span>,</span></span>
<span class="r-in"><span>  p <span class="op">=</span> <span class="fl">20</span>,</span></span>
<span class="r-in"><span>  frame <span class="op">=</span> <span class="st">"circle"</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"TBE"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/consensusNet.html" class="external-link">consensusNet</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">ML_bs</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-7.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/consensusNet.html" class="external-link">consensusNet</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">NJ_bs</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="build_phytree_pq-8.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">ps_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/merge_phyloseq.html" class="external-link">merge_phyloseq</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">$</span><span class="va">tree</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

