<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Facilitate the description, transformation, exploration, and reproducibility of metabarcoding analyses. MiscMetabar is mainly built on top of the phyloseq, dada2 and targets R packages. It helps to build reproducible and robust bioinformatics pipelines in R. MiscMetabar makes ecological analysis of alpha and beta-diversity easier, more reproducible and more powerful by integrating a large number of tools. Important features are described in Taudière A. (2023) &lt;doi:10.21105/joss.06038&gt;.">
<title>Miscellaneous Functions for Metabarcoding Analysis • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Miscellaneous Functions for Metabarcoding Analysis">
<meta property="og:description" content="Facilitate the description, transformation, exploration, and reproducibility of metabarcoding analyses. MiscMetabar is mainly built on top of the phyloseq, dada2 and targets R packages. It helps to build reproducible and robust bioinformatics pipelines in R. MiscMetabar makes ecological analysis of alpha and beta-diversity easier, more reproducible and more powerful by integrating a large number of tools. Important features are described in Taudière A. (2023) &lt;doi:10.21105/joss.06038&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.7.11</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/MiscMetabar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="articles/explore_data.html">Explore phyloseq data</a>
    <a class="dropdown-item" href="articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6>
    <a class="dropdown-item" href="articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="articles/beta-div.html">Beta-diversity</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6>
    <a class="dropdown-item" href="articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9">
<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- devtools::build_readme() -->
<div class="section level1">
<div class="page-header"><h1 id="miscmetabar-">MiscMetabar <a href="https://adrientaudiere.github.io/MiscMetabar/"><img src="https://adrientaudiere.github.io/MiscMetabar/reference/figures/logo.png" align="right" height="138" alt="MiscMetabar website"></a>
<a class="anchor" aria-label="anchor" href="#miscmetabar-"></a>
</h1></div>
<p>See the pkgdown documentation site <a href="https://adrientaudiere.github.io/MiscMetabar/">here</a> and the <a href="https://doi.org/10.21105/joss.06038" class="external-link">package paper</a> in the Journal Of Open Softwares.</p>
<p>Biological studies, especially in ecology, health sciences and taxonomy, need to describe the biological composition of samples. During the last twenty years, (i) the development of DNA sequencing, (ii) reference databases, (iii) high-throughput sequencing (HTS), and (iv) bioinformatics resources have allowed the description of biological communities through metabarcoding. Metabarcoding involves the sequencing of millions (<em>meta</em>-) of short regions of specific DNA (<em>-barcoding</em>, Valentini, Pompanon, and Taberlet (2009)) often from environmental samples (eDNA, Taberlet et al. (2012)) such as human stomach contents, lake water, soil and air.</p>
<p><code>MiscMetabar</code> aims to facilitate the <strong>description</strong>, <strong>transformation</strong>, <strong>exploration</strong> and <strong>reproducibility</strong> of metabarcoding analysis using R. The development of <code>MiscMetabar</code> relies heavily on the R packages <a href="https://benjjneb.github.io/dada2/index.html" class="external-link"><code>dada2</code></a> (Callahan et al. 2016), <a href="https://joey711.github.io/phyloseq/" class="external-link"><code>phyloseq</code></a> (McMurdie and Holmes 2013) and <a href="https://books.ropensci.org/targets/" class="external-link"><code>targets</code></a> (Landau 2021).</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>There is no CRAN version of MiscMetabar for now (work in progress). As MiscMetabar heavily relies on two bioconductor packages (dada and phyloseq), we need to first install those 2 packages using BiocManager.</p>
<p>You can install the stable version from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"adrientaudiere/MiscMetabar"</span><span class="op">)</span></span></code></pre></div>
<p>You can install the development version from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"adrientaudiere/MiscMetabar"</span>, ref <span class="op">=</span> <span class="st">"dev"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="some-use-of-miscmetabar">Some use of MiscMetabar<a class="anchor" aria-label="anchor" href="#some-use-of-miscmetabar"></a>
</h2>
<p>See articles in the <a href="https://adrientaudiere.github.io/MiscMetabar/">MiscMetabar</a> website for more examples.</p>
<p>For an introduction to metabarcoding in R, Please visite the <a href="https://adrientaudiere.github.io/MiscMetabar/articles/states_of_fields_in_R.html">state of the field</a> articles. The <a href="https://adrientaudiere.github.io/MiscMetabar/articles/import_export_track.html">import, export and track</a> article explains how import and export <code>phyloseq</code> object. Its also show how to summarize useful information (number of sequences, samples and clusters) accross bioinformatic pipelines. The article <a href="https://adrientaudiere.github.io/MiscMetabar/articles/explore_data.html">explore data</a> takes a closer look to different way of explore samples and taxonomical data from <code>phyloseq</code> object.</p>
<p>If you are interested in ecological metrics, see the articles describing <a href="https://adrientaudiere.github.io/MiscMetabar/articles/alpha-div.html">alpha-diversity</a> and <a href="https://adrientaudiere.github.io/MiscMetabar/articles/beta-div.html">beta-diversity</a> analysis. The article <a href="https://adrientaudiere.github.io/MiscMetabar/articles/filter.html">filter taxa and samples</a> describes some data-filtering processes using MiscMetabar and the <a href="https://adrientaudiere.github.io/MiscMetabar/articles/Reclustering.html">reclustering</a> tutorial introduces the different way of clustering already-clustered OTU/ASV. The article <a href="https://adrientaudiere.github.io/MiscMetabar/articles/tengeler.html">tengeler</a> explore the dataset from Tengeler et al. (2020) using some MiscMetabar functions.</p>
<p>For developers, I also wrote a article describing som <a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html">rules of codes</a>.</p>
<div class="section level3">
<h3 id="summarize-a-physeq-object">Summarize a physeq object<a class="anchor" aria-label="anchor" href="#summarize-a-physeq-object"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">"MiscMetabar"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">"phyloseq"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://magrittr.tidyverse.org" class="external-link">"magrittr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"data_fungi"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-example-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="alpha-diversity-analysis">Alpha-diversity analysis<a class="anchor" aria-label="anchor" href="#alpha-diversity-analysis"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">MiscMetabar</span><span class="fu">::</span><span class="fu"><a href="reference/hill_pq.html">hill_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, variable <span class="op">=</span> <span class="st">"Height"</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">plot_Hill_0</span></span></code></pre></div>
<div class="figure">
<img src="reference/figures/README-unnamed-chunk-4-1.png" alt="Hill number 1" width="100%"><p class="caption">
Hill number 1
</p>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">plot_tuckey</span></span></code></pre></div>
<div class="figure">
<img src="reference/figures/README-unnamed-chunk-5-1.png" alt="Result of the Tuckey post-hoc test" width="100%"><p class="caption">
Result of the Tuckey post-hoc test
</p>
</div>
</div>
<div class="section level3">
<h3 id="beta-diversity-analysis">Beta-diversity analysis<a class="anchor" aria-label="anchor" href="#beta-diversity-analysis"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/gaospecial/ggVennDiagram" class="external-link">"ggVennDiagram"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggVennDiagramà"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="reference/ggvenn_pq.html">ggvenn_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"BuPu"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Share number of ASV among Height in tree"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-6-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="note-for-non-linux-users">Note for non-linux users<a class="anchor" aria-label="anchor" href="#note-for-non-linux-users"></a>
</h3>
<p>Some functions may not work on windows (<em>e.g.</em> [track_wflow()], [cutadapt_remove_primers()], [krona()], [vsearch_clustering()], …). A solution is to exploit docker container, for example the using the great <a href="https://rocker-project.org/" class="external-link">rocker project</a>.</p>
<p>Here is a list of functions with some limitations or not working at all on windows OS:</p>
<ul>
<li>[build_phytree_pq()]</li>
<li>[count_seq()]</li>
<li>[cutadapt_remove_primers()]</li>
<li>[krona()]</li>
<li>[merge_krona()]</li>
<li>[multipatt_pq()]</li>
<li>[plot_tsne_pq()]</li>
<li>[rotl_pq()]</li>
<li>[save_pq()]</li>
<li>[tax_datatable()]</li>
<li>[track_wkflow()]</li>
<li>[track_wkflow_samples()]</li>
<li>[tsne_pq()]</li>
<li>[venn_pq()]</li>
</ul>
<p>MiscMetabar is developed under Linux and the vast majority of functions may works on Unix system, but its functionning is not test under iOS.</p>
</div>
<div class="section level3">
<h3 id="installation-of-other-softwares-for-debian-linux-distributions">Installation of other softwares for debian Linux distributions<a class="anchor" aria-label="anchor" href="#installation-of-other-softwares-for-debian-linux-distributions"></a>
</h3>
<p>If you encounter any errors or have any questions about the installation of these softwares, please visit their dedicated websites.</p>
<div class="section level4">
<h4 id="blast">
<a href="https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html#downloadblastdata" class="external-link">blast+</a><a class="anchor" aria-label="anchor" href="#blast"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">sudo</span> apt-get install ncbi-blast+</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="vsearch">
<a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a><a class="anchor" aria-label="anchor" href="#vsearch"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">sudo</span> apt-get install vsearch</span></code></pre></div>
<p>An other possibilities is to <a href="https://bioconda.github.io/recipes/vsearch/README.html?highlight=vsearch#package-package%20'vsearch'" class="external-link">install vsearch</a> with <code>conda</code>.</p>
</div>
<div class="section level4">
<h4 id="swarm">
<a href="https://github.com/torognes/swarm" class="external-link">swarm</a><a class="anchor" aria-label="anchor" href="#swarm"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/torognes/swarm.git</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="bu">cd</span> swarm/</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">make</span></span></code></pre></div>
<p>An other possibilities is to <a href="https://bioconda.github.io/recipes/swarm/README.html?highlight=swarm#package-package%20'swarm'" class="external-link">install swarm</a> with <code>conda</code>.</p>
</div>
<div class="section level4">
<h4 id="mumu">
<a href="https://github.com/frederic-mahe/mumu" class="external-link">Mumu</a><a class="anchor" aria-label="anchor" href="#mumu"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/frederic-mahe/mumu.git</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="bu">cd</span> ./mumu/</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="fu">make</span> check</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="fu">make</span> install  <span class="co"># as root or sudo</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cutadapt">
<a href="https://cutadapt.readthedocs.io/en/stable/" class="external-link">cutadapt</a><a class="anchor" aria-label="anchor" href="#cutadapt"></a>
</h4>
<div class="sourceCode" id="cb11"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> cutadaptenv cutadapt</span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-callahan2016" class="csl-entry">
Callahan, Benjamin J, Paul J McMurdie, Michael J Rosen, Andrew W Han, Amy Jo A Johnson, and Susan P Holmes. 2016. “DADA2: High-Resolution Sample Inference from Illumina Amplicon Data.” <em>Nature Methods</em> 13 (7): 581–83. <a href="https://doi.org/10.1038/nmeth.3869" class="external-link uri">https://doi.org/10.1038/nmeth.3869</a>.
</div>
<div id="ref-landau2021" class="csl-entry">
Landau, William Michael. 2021. “The Targets r Package: A Dynamic Make-Like Function-Oriented Pipeline Toolkit for Reproducibility and High-Performance Computing.” <em>Journal of Open Source Software</em> 6 (57): 2959. <a href="https://doi.org/10.21105/joss.02959" class="external-link uri">https://doi.org/10.21105/joss.02959</a>.
</div>
<div id="ref-mcmurdie2013" class="csl-entry">
McMurdie, Paul J., and Susan Holmes. 2013. “Phyloseq: An r Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PLoS ONE</em> 8 (4): e61217. <a href="https://doi.org/10.1371/journal.pone.0061217" class="external-link uri">https://doi.org/10.1371/journal.pone.0061217</a>.
</div>
<div id="ref-taberlet2012" class="csl-entry">
Taberlet, Pierre, Eric Coissac, Mehrdad Hajibabaei, and Loren H Rieseberg. 2012. “Environmental Dna.” <em>Molecular Ecology</em>. Wiley Online Library. <a href="https://doi.org/10.1002/(issn)2637-4943" class="external-link uri">https://doi.org/10.1002/(issn)2637-4943</a>.
</div>
<div id="ref-valentini2009" class="csl-entry">
Valentini, Alice, François Pompanon, and Pierre Taberlet. 2009. “DNA Barcoding for Ecologists.” <em>Trends in Ecology &amp; Evolution</em> 24 (2): 110–17. <a href="https://doi.org/10.1016/j.tree.2008.09.011" class="external-link uri">https://doi.org/10.1016/j.tree.2008.09.011</a>.
</div>
</div>
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=MiscMetabar" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/adrientaudiere/MiscMetabar/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/adrientaudiere/MiscMetabar/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/AGPL-3" class="external-link">AGPL-3</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MiscMetabar</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Adrien Taudière <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0003-1088-1182" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
