<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Filter taxa and samples • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Filter taxa and samples">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.9.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Filter taxa and samples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/articles/filter.Rmd" class="external-link"><code>vignettes/articles/filter.Rmd</code></a></small>
      <div class="d-none name"><code>filter.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">MiscMetabar</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://renkun-ken.github.io/formattable/" class="external-link">formattable</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="filter-samples">Filter samples<a class="anchor" aria-label="anchor" href="#filter-samples"></a>
</h3>
<p><a href="https://joey711.github.io/phyloseq/" class="external-link">Phyloseq</a> package
already propose a function to select samples
(<code><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples()</a></code>), but in some case, the subset internal
function is painful. <code>MiscMetabar</code> propose a complementary
function ([subset_samples_pq()]) which is more versatile but need to be
used with caution because the order of the condition must match the
orders of the samples.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span>
<span><span class="va">cond_samp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"A1"</span>, <span class="va">data_fungi</span><span class="op">@</span><span class="va">sam_data</span><span class="op">[[</span><span class="st">"Sample_names"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/subset_samples_pq.html">subset_samples_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="va">cond_samp</span><span class="op">)</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 1420 taxa and 9 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 9 samples by 7 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 1420 taxa by 12 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; refseq()      DNAStringSet:      [ 1420 reference sequences ]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-taxa">Filter Taxa<a class="anchor" aria-label="anchor" href="#filter-taxa"></a>
</h3>
<div class="section level4">
<h4 id="filter-taxa-using-conditions">Filter taxa using condition(s)<a class="anchor" aria-label="anchor" href="#filter-taxa-using-conditions"></a>
</h4>
<p><a href="https://joey711.github.io/phyloseq/" class="external-link">Phyloseq</a> package
already propose a function to select samples
(<code><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa()</a></code>), but in some case, the subset internal
function is painful. <code>MiscMetabar</code> propose a complementary
function ([subset_taxa_pq()]) which is more versatile and is based on
the names of the taxa to match the condition and the taxa in the
phyloseq object. In the example code above, we filter fungi from the
Phylum “Basidiomycota” using <code><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">phyloseq::subset_taxa()</a></code> and
then select only taxa with more than 1000 nb_sequences using
<code><a href="../reference/subset_taxa_pq.html">subset_taxa_pq()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_basidio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="va">Phylum</span> <span class="op">==</span> <span class="st">"Basidiomycota"</span><span class="op">)</span></span>
<span><span class="va">df_basidio_abundant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subset_taxa_pq.html">subset_taxa_pq</a></span><span class="op">(</span></span>
<span>  <span class="va">df_basidio</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">df_basidio</span><span class="op">@</span><span class="va">otu_table</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filter-taxa-using-blast-score-against-a-database">Filter taxa using blast score against a database<a class="anchor" aria-label="anchor" href="#filter-taxa-using-blast-score-against-a-database"></a>
</h4>
<p>In some cases, we want to select only a given clade of taxon. One
solution is to select only taxa with information given by taxonomic
assignment (e.g. using the function
<code><a href="https://rdrr.io/pkg/dada2/man/assignTaxonomy.html" class="external-link">dada2::assignTaxonomy()</a></code>). However, in some cases, this
information lead to false positive and false negative selection.
<code>MiscMetabar</code> implement an other method relying on <a href="https://blast.ncbi.nlm.nih.gov/" class="external-link">blast</a> software (Altschul
<em>et al.</em> 1990 &amp; 1997) and database. The idea is to set a
cutoff in four parameters to select only taxa which are close enough to
sequences in the database :</p>
<ul>
<li>
<strong>id_filter</strong> (default: 90) cut of in identity percent
to keep result.</li>
<li>
<strong>bit_score_filter</strong> (default: 50) cut of in bit score
to keep result. The higher the bit-score, the better the sequence
similarity. The bit-score is the requires size of a sequence database in
which the current match could be found just by chance. The bit-score is
a log2 scaled and normalized raw-score. Each increase by one doubles the
required database size (2bit-score).</li>
<li>
<strong>min_cover_filter</strong> (default: 50) cut of in query
cover (%) to keep result.</li>
<li>
<strong>e_value_filter</strong> (default: 1e-30) cut of in e-value
(%) to keep result. The BLAST E-value is the number of expected hits of
similar quality (score) that could be found just by chance.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>  <span class="st">"100_sp_UNITE_sh_general_release_dynamic.fasta"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"MiscMetabar"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="va">blast_error_or_not</span> <span class="op">&lt;-</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"blastn 2&gt;&amp;1"</span>, intern <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, silent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/methods/is.html" class="external-link">is</a></span><span class="op">(</span><span class="va">blast_error_or_not</span>, <span class="st">"try-error"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">df_blast_80</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_asv_blast.html">filter_asv_blast</a></span><span class="op">(</span><span class="va">df_basidio</span>, fasta_for_db <span class="op">=</span> <span class="va">path_db</span><span class="op">)</span></span>
<span>  <span class="va">df_blast_50</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_asv_blast.html">filter_asv_blast</a></span><span class="op">(</span><span class="va">df_basidio</span>,</span>
<span>    fasta_for_db <span class="op">=</span> <span class="va">path_db</span>,</span>
<span>    id_filter <span class="op">=</span> <span class="fl">50</span>, e_value_filter <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    bit_score_filter <span class="op">=</span> <span class="fl">20</span>, min_cover_filter <span class="op">=</span> <span class="fl">20</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">track_formattable</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"raw data"</span> <span class="op">=</span> <span class="va">df_basidio</span>,</span>
<span>        <span class="st">"id_filter = 80"</span> <span class="op">=</span> <span class="va">df_blast_80</span>,</span>
<span>        <span class="st">"id_filter = 50"</span> <span class="op">=</span> <span class="va">df_blast_50</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/methods/is.html" class="external-link">is</a></span><span class="op">(</span><span class="va">blast_error_or_not</span>, <span class="st">"try-error"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/formattable.html" class="external-link">formattable</a></span><span class="op">(</span></span>
<span>    <span class="va">track_formattable</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/area.html" class="external-link">area</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">nb_sequences</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/color_bar.html" class="external-link">color_bar</a></span><span class="op">(</span><span class="st">"cyan"</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/area.html" class="external-link">area</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">nb_clusters</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/normalize_bar.html" class="external-link">normalize_bar</a></span><span class="op">(</span><span class="st">"yellowgreen"</span>,</span>
<span>        na.rm <span class="op">=</span> <span class="cn">TRUE</span>, min <span class="op">=</span> <span class="fl">0.3</span></span>
<span>      <span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/area.html" class="external-link">area</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">nb_samples</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/normalize_bar.html" class="external-link">normalize_bar</a></span><span class="op">(</span><span class="st">"lightpink"</span>,</span>
<span>        na.rm <span class="op">=</span> <span class="cn">TRUE</span>, min <span class="op">=</span> <span class="fl">0.3</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filter-taxa-using-a-known-taxa-for-control">Filter taxa using a known taxa for control<a class="anchor" aria-label="anchor" href="#filter-taxa-using-a-known-taxa-for-control"></a>
</h4>
<p>To filter out contamination, one solution is to add a proportion of a
known taxa which is not present in the environment of the study. In that
case we can define some threshold for each sample to discard taxon based
on pseudo-abundance. In the example code above, we select taxon using
the ASV_50 as control through 6 different algorithms.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_seq</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/subset_taxa_tax_control.html">subset_taxa_tax_control</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">otu_table</span><span class="op">[</span>, <span class="fl">50</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"cutoff_seq"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">res_mixt</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/subset_taxa_tax_control.html">subset_taxa_tax_control</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">otu_table</span><span class="op">[</span>, <span class="fl">50</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"cutoff_mixt"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">res_diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/subset_taxa_tax_control.html">subset_taxa_tax_control</a></span><span class="op">(</span></span>
<span>    <span class="va">data_fungi</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">otu_table</span><span class="op">[</span>, <span class="fl">50</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"cutoff_diff"</span>,</span>
<span>    min_diff_for_cutoff <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">res_min</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/subset_taxa_tax_control.html">subset_taxa_tax_control</a></span><span class="op">(</span></span>
<span>      <span class="va">data_fungi</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">otu_table</span><span class="op">[</span>, <span class="fl">50</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"min"</span>,</span>
<span>      min_diff_for_cutoff <span class="op">=</span> <span class="fl">2</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">res_max</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/subset_taxa_tax_control.html">subset_taxa_tax_control</a></span><span class="op">(</span></span>
<span>      <span class="va">data_fungi</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">otu_table</span><span class="op">[</span>, <span class="fl">50</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"max"</span>,</span>
<span>      min_diff_for_cutoff <span class="op">=</span> <span class="fl">2</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">res_mean</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/subset_taxa_tax_control.html">subset_taxa_tax_control</a></span><span class="op">(</span></span>
<span>      <span class="va">data_fungi</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">otu_table</span><span class="op">[</span>, <span class="fl">50</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>      min_diff_for_cutoff <span class="op">=</span> <span class="fl">2</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">track_formattable</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"raw data"</span> <span class="op">=</span> <span class="va">data_fungi</span>,</span>
<span>  <span class="st">"cutoff_seq"</span> <span class="op">=</span> <span class="va">res_seq</span>,</span>
<span>  <span class="st">"cutoff_mixt"</span> <span class="op">=</span> <span class="va">res_mixt</span>,</span>
<span>  <span class="st">"cutoff_diff"</span> <span class="op">=</span> <span class="va">res_diff</span>,</span>
<span>  <span class="st">"min"</span> <span class="op">=</span> <span class="va">res_min</span>,</span>
<span>  <span class="st">"max"</span> <span class="op">=</span> <span class="va">res_max</span>,</span>
<span>  <span class="st">"mean"</span> <span class="op">=</span> <span class="va">res_mean</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/formattable.html" class="external-link">formattable</a></span><span class="op">(</span></span>
<span>  <span class="va">track_formattable</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/area.html" class="external-link">area</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">nb_sequences</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/color_bar.html" class="external-link">color_bar</a></span><span class="op">(</span><span class="st">"cyan"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/area.html" class="external-link">area</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">nb_clusters</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/normalize_bar.html" class="external-link">normalize_bar</a></span><span class="op">(</span><span class="st">"yellowgreen"</span>,</span>
<span>      na.rm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      min <span class="op">=</span> <span class="fl">0.3</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/area.html" class="external-link">area</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">nb_samples</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://renkun-ken.github.io/formattable/reference/normalize_bar.html" class="external-link">normalize_bar</a></span><span class="op">(</span><span class="st">"lightpink"</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<table class="table table table-condensed">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
nb_sequences
</th>
<th style="text-align:right;">
nb_clusters
</th>
<th style="text-align:right;">
nb_samples
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
raw data
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: cyan; width: 100.00%">1839124</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: yellowgreen; width: 100.00%">1420</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: lightpink; width: 100.00%">185</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
cutoff_seq
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: cyan; width: 98.83%">1817543</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: yellowgreen; width: 93.88%">1402</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: lightpink; width: 100.00%">185</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
cutoff_mixt
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: cyan; width: 96.47%">1774180</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: yellowgreen; width: 52.09%">1279</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: lightpink; width: 100.00%">185</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
cutoff_diff
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: cyan; width: 98.16%">1805274</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: yellowgreen; width: 85.73%">1378</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: lightpink; width: 100.00%">185</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
min
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: cyan; width: 99.84%">1836158</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: yellowgreen; width: 98.98%">1417</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: lightpink; width: 100.00%">185</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
max
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: cyan; width: 95.27%">1752217</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: yellowgreen; width: 30.00%">1214</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: lightpink; width: 100.00%">185</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
mean
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: cyan; width: 97.36%">1790569</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: yellowgreen; width: 67.72%">1325</span>
</td>
<td style="text-align:right;">
<span style="display: inline-block; direction: rtl; unicode-bidi: plaintext; border-radius: 4px; padding-right: 2px; background-color: lightpink; width: 100.00%">185</span>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.11.0 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=fr_FR.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Paris</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] formattable_0.2.1 MiscMetabar_0.9.4 purrr_1.0.2       dplyr_1.1.4      </span></span>
<span><span class="co">#&gt; [5] dada2_1.32.0      Rcpp_1.0.13       ggplot2_3.5.1     phyloseq_1.48.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-8                pbapply_1.7-2              </span></span>
<span><span class="co">#&gt;   [3] deldir_2.0-4                permute_0.9-7              </span></span>
<span><span class="co">#&gt;   [5] rlang_1.1.4                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [7] ade4_1.7-22                 matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt;   [9] compiler_4.4.1              mgcv_1.9-1                 </span></span>
<span><span class="co">#&gt;  [11] png_0.1-8                   systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt;  [13] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [15] stringr_1.5.1               pwalign_1.0.0              </span></span>
<span><span class="co">#&gt;  [17] pkgconfig_2.0.3             crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [19] fastmap_1.2.0               XVector_0.44.0             </span></span>
<span><span class="co">#&gt;  [21] utf8_1.2.4                  Rsamtools_2.20.0           </span></span>
<span><span class="co">#&gt;  [23] rmarkdown_2.28              UCSC.utils_1.0.0           </span></span>
<span><span class="co">#&gt;  [25] ragg_1.3.2                  xfun_0.47                  </span></span>
<span><span class="co">#&gt;  [27] zlibbioc_1.50.0             cachem_1.1.0               </span></span>
<span><span class="co">#&gt;  [29] GenomeInfoDb_1.40.1         jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;  [31] biomformat_1.32.0           rhdf5filters_1.16.0        </span></span>
<span><span class="co">#&gt;  [33] DelayedArray_0.30.1         Rhdf5lib_1.26.0            </span></span>
<span><span class="co">#&gt;  [35] BiocParallel_1.38.0         jpeg_0.1-10                </span></span>
<span><span class="co">#&gt;  [37] parallel_4.4.1              cluster_2.1.6              </span></span>
<span><span class="co">#&gt;  [39] R6_2.5.1                    RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [41] bslib_0.8.0                 stringi_1.8.4              </span></span>
<span><span class="co">#&gt;  [43] GenomicRanges_1.56.1        jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [45] SummarizedExperiment_1.34.0 iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [47] knitr_1.48                  mixtools_2.0.0             </span></span>
<span><span class="co">#&gt;  [49] IRanges_2.38.1              Matrix_1.7-0               </span></span>
<span><span class="co">#&gt;  [51] splines_4.4.1               igraph_2.0.3               </span></span>
<span><span class="co">#&gt;  [53] tidyselect_1.2.1            rstudioapi_0.16.0          </span></span>
<span><span class="co">#&gt;  [55] abind_1.4-5                 yaml_2.3.10                </span></span>
<span><span class="co">#&gt;  [57] vegan_2.6-8                 codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [59] hwriter_1.3.2.1             lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [61] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [63] Biobase_2.64.0              withr_3.0.1                </span></span>
<span><span class="co">#&gt;  [65] ShortRead_1.62.0            evaluate_0.24.0            </span></span>
<span><span class="co">#&gt;  [67] desc_1.4.3                  survival_3.7-0             </span></span>
<span><span class="co">#&gt;  [69] RcppParallel_5.1.9          kernlab_0.9-33             </span></span>
<span><span class="co">#&gt;  [71] Biostrings_2.72.1           pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [73] MatrixGenerics_1.16.0       foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [75] stats4_4.4.1                plotly_4.10.4              </span></span>
<span><span class="co">#&gt;  [77] generics_0.1.3              S4Vectors_0.42.1           </span></span>
<span><span class="co">#&gt;  [79] munsell_0.5.1               scales_1.3.0               </span></span>
<span><span class="co">#&gt;  [81] glue_1.7.0                  lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;  [83] tools_4.4.1                 interp_1.1-6               </span></span>
<span><span class="co">#&gt;  [85] data.table_1.16.0           GenomicAlignments_1.40.0   </span></span>
<span><span class="co">#&gt;  [87] fs_1.6.4                    rhdf5_2.48.0               </span></span>
<span><span class="co">#&gt;  [89] grid_4.4.1                  tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [91] ape_5.8                     latticeExtra_0.6-30        </span></span>
<span><span class="co">#&gt;  [93] colorspace_2.1-1            nlme_3.1-165               </span></span>
<span><span class="co">#&gt;  [95] GenomeInfoDbData_1.2.12     cli_3.6.3                  </span></span>
<span><span class="co">#&gt;  [97] textshaping_0.4.0           fansi_1.0.6                </span></span>
<span><span class="co">#&gt;  [99] segmented_2.1-1             viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [101] S4Arrays_1.4.1              gtable_0.3.5               </span></span>
<span><span class="co">#&gt; [103] sass_0.4.9                  digest_0.6.37              </span></span>
<span><span class="co">#&gt; [105] BiocGenerics_0.50.0         SparseArray_1.4.8          </span></span>
<span><span class="co">#&gt; [107] htmlwidgets_1.6.4           htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [109] pkgdown_2.1.0               multtest_2.60.0            </span></span>
<span><span class="co">#&gt; [111] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [113] MASS_7.3-61</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
