<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MiscMetabar">
<title>import_export_track â€¢ MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="import_export_track">
<meta property="og:description" content="MiscMetabar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.9.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>import_export_track</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/articles/import_export_track.Rmd" class="external-link"><code>vignettes/articles/import_export_track.Rmd</code></a></small>
      <div class="d-none name"><code>import_export_track.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">MiscMetabar</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://renkun-ken.github.io/formattable/" class="external-link">formattable</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="export-phyloseq-object">Export phyloseq object<a class="anchor" aria-label="anchor" href="#export-phyloseq-object"></a>
</h3>
<p>You can export a phyloseq object to csv (and txt for phylogenetic
tree) files in a folder. It is possible to export each table into one
file or to merge all slot (except phytree) in one file (args
<code>one_file = TRUE</code>). Finally, if <code>rdata</code> is set to
TRUE, a <code>rdata</code> file containing the phyloseq object is also
writed.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_pq.html">write_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/write_pq.html">write_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span>, one_file <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/write_pq.html">write_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span>, rdata <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Finally, you can use the function <code><a href="../reference/save_pq.html">save_pq()</a></code> to write the
phyloseq object in all 3 versions (one table for each slot, a file
merging each slot and an Rdata file).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/save_pq.html">save_pq</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="import">Import<a class="anchor" aria-label="anchor" href="#import"></a>
</h3>
<p>To import a Rdata file, just use <code><a href="https://rdrr.io/r/base/load.html" class="external-link">load()</a></code> base function.
In order to import phyloseq object from a folder create using
<code><a href="../reference/write_pq.html">write_pq()</a></code> or <code><a href="../reference/save_pq.html">save_pq()</a></code>, please use the
function <code><a href="../reference/read_pq.html">read_pq()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_pq.html">read_pq</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="tracking-sequences-clusters-and-samples">Tracking sequences, clusters and samples<a class="anchor" aria-label="anchor" href="#tracking-sequences-clusters-and-samples"></a>
</h3>
<p>In bioinformatic pipeline, we often need to track the number of
samples, sequences and clusters across step in the pipeline. MiscMetabar
propose two utilities to achieve this goal : <code><a href="../reference/track_wkflow.html">track_wkflow()</a></code>
and a version to compute value per samples :
<code><a href="../reference/track_wkflow_samples.html">track_wkflow_samples()</a></code>. The function
<code><a href="../reference/track_wkflow.html">track_wkflow()</a></code> can deal with (i) fastq and fastg.gz files,
dada-class object, derep-class object, matrix of samples x clusters
(e.g.Â <code>otu_table</code>) and phyloseq-class object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="va">data_fungi_sp_known</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   nb_sequences nb_clusters nb_samples</span></span>
<span><span class="co">#&gt; 1      1839124        1420        185</span></span>
<span><span class="co">#&gt; 2      1106581         651        185</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 11 (bullseye)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.13.so;  LAPACK version 3.9.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=fr_FR.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Paris</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] formattable_0.2.1 MiscMetabar_0.9.1 purrr_1.0.2       dplyr_1.1.4      </span></span>
<span><span class="co">#&gt; [5] dada2_1.30.0      Rcpp_1.0.12       ggplot2_3.5.0     phyloseq_1.46.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7                pbapply_1.7-2              </span></span>
<span><span class="co">#&gt;   [3] deldir_2.0-4                permute_0.9-7              </span></span>
<span><span class="co">#&gt;   [5] rlang_1.1.3                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [7] ade4_1.7-22                 matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt;   [9] compiler_4.3.3              mgcv_1.9-1                 </span></span>
<span><span class="co">#&gt;  [11] png_0.1-8                   systemfonts_1.0.6          </span></span>
<span><span class="co">#&gt;  [13] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [15] stringr_1.5.1               pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [17] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [19] XVector_0.42.0              utf8_1.2.4                 </span></span>
<span><span class="co">#&gt;  [21] Rsamtools_2.18.0            rmarkdown_2.26             </span></span>
<span><span class="co">#&gt;  [23] ragg_1.3.0                  xfun_0.43                  </span></span>
<span><span class="co">#&gt;  [25] zlibbioc_1.48.2             cachem_1.0.8               </span></span>
<span><span class="co">#&gt;  [27] GenomeInfoDb_1.38.8         jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;  [29] biomformat_1.30.0           rhdf5filters_1.14.1        </span></span>
<span><span class="co">#&gt;  [31] DelayedArray_0.28.0         Rhdf5lib_1.24.2            </span></span>
<span><span class="co">#&gt;  [33] BiocParallel_1.36.0         jpeg_0.1-10                </span></span>
<span><span class="co">#&gt;  [35] parallel_4.3.3              cluster_2.1.6              </span></span>
<span><span class="co">#&gt;  [37] R6_2.5.1                    bslib_0.7.0                </span></span>
<span><span class="co">#&gt;  [39] stringi_1.8.3               RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [41] GenomicRanges_1.54.1        jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [43] SummarizedExperiment_1.32.0 iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [45] knitr_1.46                  IRanges_2.36.0             </span></span>
<span><span class="co">#&gt;  [47] Matrix_1.6-5                splines_4.3.3              </span></span>
<span><span class="co">#&gt;  [49] igraph_2.0.3                tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [51] rstudioapi_0.16.0           abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [53] yaml_2.3.8                  vegan_2.6-4                </span></span>
<span><span class="co">#&gt;  [55] codetools_0.2-19            hwriter_1.3.2.1            </span></span>
<span><span class="co">#&gt;  [57] lattice_0.22-6              tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [59] plyr_1.8.9                  Biobase_2.62.0             </span></span>
<span><span class="co">#&gt;  [61] withr_3.0.0                 ShortRead_1.60.0           </span></span>
<span><span class="co">#&gt;  [63] evaluate_0.23               desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [65] survival_3.5-8              RcppParallel_5.1.7         </span></span>
<span><span class="co">#&gt;  [67] Biostrings_2.70.3           pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [69] MatrixGenerics_1.14.0       foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [71] stats4_4.3.3                generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [73] RCurl_1.98-1.14             S4Vectors_0.40.2           </span></span>
<span><span class="co">#&gt;  [75] munsell_0.5.1               scales_1.3.0               </span></span>
<span><span class="co">#&gt;  [77] glue_1.7.0                  tools_4.3.3                </span></span>
<span><span class="co">#&gt;  [79] interp_1.1-6                data.table_1.15.4          </span></span>
<span><span class="co">#&gt;  [81] GenomicAlignments_1.38.2    fs_1.6.3                   </span></span>
<span><span class="co">#&gt;  [83] rhdf5_2.46.1                grid_4.3.3                 </span></span>
<span><span class="co">#&gt;  [85] ape_5.8                     latticeExtra_0.6-30        </span></span>
<span><span class="co">#&gt;  [87] colorspace_2.1-0            nlme_3.1-164               </span></span>
<span><span class="co">#&gt;  [89] GenomeInfoDbData_1.2.11     cli_3.6.2                  </span></span>
<span><span class="co">#&gt;  [91] textshaping_0.3.7           fansi_1.0.6                </span></span>
<span><span class="co">#&gt;  [93] S4Arrays_1.2.1              gtable_0.3.4               </span></span>
<span><span class="co">#&gt;  [95] sass_0.4.9                  digest_0.6.35              </span></span>
<span><span class="co">#&gt;  [97] BiocGenerics_0.48.1         SparseArray_1.2.4          </span></span>
<span><span class="co">#&gt;  [99] htmlwidgets_1.6.4           memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [101] htmltools_0.5.8.1           pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [103] multtest_2.58.0             lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [105] MASS_7.3-60.0.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Adrien TaudiÃ¨re.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
