<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Import/export and track • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Import/export and track">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Import/export and track</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/articles/import_export_track.Rmd" class="external-link"><code>vignettes/articles/import_export_track.Rmd</code></a></small>
      <div class="d-none name"><code>import_export_track.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">MiscMetabar</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="export-phyloseq-object">Export phyloseq object<a class="anchor" aria-label="anchor" href="#export-phyloseq-object"></a>
</h3>
<p>You can export a phyloseq object to csv (and txt for phylogenetic
tree) files in a folder. It is possible to export each table into one
file or to merge all slot (except phytree) in one file (args
<code>one_file = TRUE</code>). Finally, if <code>rdata</code> is set to
TRUE, a <code>rdata</code> file containing the phyloseq object is also
writed.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_pq.html">write_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/write_pq.html">write_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span>, one_file <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/write_pq.html">write_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span>, rdata <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Finally, you can use the function <code><a href="../reference/save_pq.html">save_pq()</a></code> to write the
phyloseq object in all 3 versions (one table for each slot, a file
merging each slot and an Rdata file).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/save_pq.html">save_pq</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="import">Import<a class="anchor" aria-label="anchor" href="#import"></a>
</h3>
<p>To import a Rdata file, just use <code><a href="https://rdrr.io/r/base/load.html" class="external-link">load()</a></code> base function.
In order to import phyloseq object from a folder create using
<code><a href="../reference/write_pq.html">write_pq()</a></code> or <code><a href="../reference/save_pq.html">save_pq()</a></code>, please use the
function <code><a href="../reference/read_pq.html">read_pq()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_pq.html">read_pq</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"fungi_phyloseq"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="tracking-sequences-clusters-and-samples">Tracking sequences, clusters and samples<a class="anchor" aria-label="anchor" href="#tracking-sequences-clusters-and-samples"></a>
</h3>
<p>In bioinformatic pipeline, we often need to track the number of
samples, sequences and clusters across step in the pipeline. MiscMetabar
propose two utilities to achieve this goal : <code><a href="../reference/track_wkflow.html">track_wkflow()</a></code>
and a version to compute value per samples :
<code><a href="../reference/track_wkflow_samples.html">track_wkflow_samples()</a></code>. The function
<code><a href="../reference/track_wkflow.html">track_wkflow()</a></code> can deal with (i) fastq and fastg.gz files,
dada-class object, derep-class object, matrix of samples x clusters
(e.g. <code>otu_table</code>) and phyloseq-class object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="va">data_fungi_sp_known</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   nb_sequences nb_clusters nb_samples</span></span>
<span><span class="co">#&gt; 1      1839124        1420        185</span></span>
<span><span class="co">#&gt; 2      1106581         651        185</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=fr_FR.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Paris</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] MiscMetabar_0.14.3 purrr_1.0.4        dplyr_1.1.4        dada2_1.34.0      </span></span>
<span><span class="co">#&gt; [5] Rcpp_1.0.14        ggplot2_3.5.1      phyloseq_1.50.0   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-9                pbapply_1.7-2              </span></span>
<span><span class="co">#&gt;   [3] deldir_2.0-4                permute_0.9-7              </span></span>
<span><span class="co">#&gt;   [5] rlang_1.1.5                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [7] ade4_1.7-22                 matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt;   [9] compiler_4.4.3              mgcv_1.9-1                 </span></span>
<span><span class="co">#&gt;  [11] png_0.1-8                   systemfonts_1.2.1          </span></span>
<span><span class="co">#&gt;  [13] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [15] stringr_1.5.1               pwalign_1.0.0              </span></span>
<span><span class="co">#&gt;  [17] pkgconfig_2.0.3             crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [19] fastmap_1.2.0               XVector_0.44.0             </span></span>
<span><span class="co">#&gt;  [21] Rsamtools_2.20.0            rmarkdown_2.29             </span></span>
<span><span class="co">#&gt;  [23] UCSC.utils_1.0.0            ragg_1.3.3                 </span></span>
<span><span class="co">#&gt;  [25] xfun_0.51                   zlibbioc_1.50.0            </span></span>
<span><span class="co">#&gt;  [27] cachem_1.1.0                GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.9.1              biomformat_1.32.0          </span></span>
<span><span class="co">#&gt;  [31] rhdf5filters_1.16.0         DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt;  [33] Rhdf5lib_1.26.0             BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt;  [35] jpeg_0.1-10                 parallel_4.4.3             </span></span>
<span><span class="co">#&gt;  [37] cluster_2.1.8               R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [39] bslib_0.9.0                 stringi_1.8.4              </span></span>
<span><span class="co">#&gt;  [41] RColorBrewer_1.1-3          GenomicRanges_1.56.2       </span></span>
<span><span class="co">#&gt;  [43] jquerylib_0.1.4             SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt;  [45] iterators_1.0.14            knitr_1.49                 </span></span>
<span><span class="co">#&gt;  [47] IRanges_2.38.1              Matrix_1.7-2               </span></span>
<span><span class="co">#&gt;  [49] splines_4.4.3               igraph_2.1.4               </span></span>
<span><span class="co">#&gt;  [51] tidyselect_1.2.1            abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [53] yaml_2.3.10                 vegan_2.6-10               </span></span>
<span><span class="co">#&gt;  [55] codetools_0.2-20            hwriter_1.3.2.1            </span></span>
<span><span class="co">#&gt;  [57] lattice_0.22-6              tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [59] plyr_1.8.9                  Biobase_2.64.0             </span></span>
<span><span class="co">#&gt;  [61] withr_3.0.2                 ShortRead_1.62.0           </span></span>
<span><span class="co">#&gt;  [63] evaluate_1.0.3              desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [65] survival_3.8-3              RcppParallel_5.1.10        </span></span>
<span><span class="co">#&gt;  [67] Biostrings_2.72.1           pillar_1.10.1              </span></span>
<span><span class="co">#&gt;  [69] MatrixGenerics_1.16.0       foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [71] stats4_4.4.3                generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [73] S4Vectors_0.42.1            munsell_0.5.1              </span></span>
<span><span class="co">#&gt;  [75] scales_1.3.0                glue_1.8.0                 </span></span>
<span><span class="co">#&gt;  [77] tools_4.4.3                 interp_1.1-6               </span></span>
<span><span class="co">#&gt;  [79] data.table_1.17.0           GenomicAlignments_1.40.0   </span></span>
<span><span class="co">#&gt;  [81] fs_1.6.5                    rhdf5_2.48.0               </span></span>
<span><span class="co">#&gt;  [83] grid_4.4.3                  ape_5.8-1                  </span></span>
<span><span class="co">#&gt;  [85] latticeExtra_0.6-30         colorspace_2.1-1           </span></span>
<span><span class="co">#&gt;  [87] nlme_3.1-167                GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt;  [89] cli_3.6.4                   textshaping_1.0.0          </span></span>
<span><span class="co">#&gt;  [91] S4Arrays_1.4.1              gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [93] sass_0.4.9                  digest_0.6.37              </span></span>
<span><span class="co">#&gt;  [95] BiocGenerics_0.50.0         SparseArray_1.4.8          </span></span>
<span><span class="co">#&gt;  [97] htmlwidgets_1.6.4           htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt;  [99] pkgdown_2.1.1               multtest_2.60.0            </span></span>
<span><span class="co">#&gt; [101] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [103] MASS_7.3-65</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
