<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="MiscMetabar_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="MiscMetabar_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="MiscMetabar_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="MiscMetabar_files/datatables-binding-0.33/datatables.js"></script><link href="MiscMetabar_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="MiscMetabar_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="MiscMetabar_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="MiscMetabar_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="MiscMetabar_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><script src="MiscMetabar_files/d3-4.5.0/d3.min.js"></script><script src="MiscMetabar_files/sankey-1/sankey.js"></script><script src="MiscMetabar_files/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/MiscMetabar.Rmd" class="external-link"><code>vignettes/MiscMetabar.Rmd</code></a></small>
      <div class="d-none name"><code>MiscMetabar.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction-to-miscmetabar-an-r-package-to-facilitate-visualization-and-reproducibility-in-metabarcoding-analysis">Introduction to MiscMetabar: an R package to facilitate
visualization and reproducibility in metabarcoding analysis<a class="anchor" aria-label="anchor" href="#introduction-to-miscmetabar-an-r-package-to-facilitate-visualization-and-reproducibility-in-metabarcoding-analysis"></a>
</h2>
<div class="section level3">
<h3 id="raison-dêtre">Raison d’être<a class="anchor" aria-label="anchor" href="#raison-d%C3%AAtre"></a>
</h3>
<ul>
<li>Complete R packages dada2 and phyloseq</li>
<li>Useful visualizations (<code>biplot_pq</code>,
<code>circle_pq</code>, <code>upset_pq</code>,
<code>ggvenn_pq</code>)</li>
<li>Facilitate the use of targets package</li>
</ul>
</div>
<div class="section level3">
<h3 id="quick-overview">Quick overview<a class="anchor" aria-label="anchor" href="#quick-overview"></a>
</h3>
<p>For an introduction to metabarcoding in R, Please visite the <a href="https://adrientaudiere.github.io/MiscMetabar/articles/states_of_fields_in_R.html">state
of the field</a> vignettes. The <a href="https://adrientaudiere.github.io/MiscMetabar/articles/import_export_track.html">import,
export and track</a> vignette explains how import and export
<code>phyloseq</code> object. Its also show how to summarize useful
information (number of sequences, samples and clusters) across
bioinformatic pipelines.</p>
<p>If you are interested in ecological metrics, see the vignettes
describing <a href="https://adrientaudiere.github.io/MiscMetabar/articles/alpha-div.html">alpha-diversity</a>
and <a href="https://adrientaudiere.github.io/MiscMetabar/articles/beta-div.html">beta-diversity</a>
analysis. The vignette <a href="https://adrientaudiere.github.io/MiscMetabar/articles/filter.html">filter
taxa and samples</a> describes some data-filtering processes using
MiscMetabar and the <a href="https://adrientaudiere.github.io/MiscMetabar/articles/Reclustering.html">reclustering</a>
tutorial introduces the different way of clustering already-clustered
OTU/ASV. The vignette <a href="https://adrientaudiere.github.io/MiscMetabar/articles/tengeler.html">tengeler</a>
explore the dataset from Tengeler et al. (2020) using some MiscMetabar
functions.</p>
<p>For developers, I also wrote a vignette describing som <a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html">rules
of codes</a>.</p>
<div class="section level4">
<h4 id="summarize-a-physeq-object">Summarize a physeq object<a class="anchor" aria-label="anchor" href="#summarize-a-physeq-object"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">"MiscMetabar"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">"phyloseq"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"data_fungi"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></code></pre></div>
<p><img src="MiscMetabar_files/figure-html/example-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="create-an-interactive-table-of-the-tax_table">Create an interactive table of the tax_table<a class="anchor" aria-label="anchor" href="#create-an-interactive-table-of-the-tax_table"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"phyloseq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tax_datatable.html">tax_datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span></span>
<span>  <span class="va">GlobalPatterns</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">@</span><span class="va">otu_table</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">100000</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["12812","317658","329744","471141","326977","64396","317182","549656","327536","340216","279599","235567","557211","263681","360229","536311","9514","94166","550960","323845","534609","339261","322235","469778","158660","331820","244304","471122","361496","248140","259569","298875","359171","192573","470973","316732","325407","171551","189047","180658","316587","98605"],["Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria"],["Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Fusobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes"],["Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Fusobacteria","Chloroplast","Chloroplast","Chloroplast","Chloroplast","Nostocophycideae","Nostocophycideae","Synechococcophycideae","4C0d-2","Betaproteobacteria","Betaproteobacteria","Gammaproteobacteria","Gammaproteobacteria","Gammaproteobacteria","Gammaproteobacteria","Alphaproteobacteria","Flavobacteria","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Bacilli","Bacilli"],["koll13","Actinomycetales","Actinomycetales","Actinomycetales","Bifidobacteriales","Fusobacteriales","Stramenopiles","Stramenopiles","Stramenopiles","Stramenopiles","Nostocales","Nostocales","Synechococcales","YS2","Neisseriales","Neisseriales","Pasteurellales","Pasteurellales","Enterobacteriales","Oceanospirillales","Rhodobacterales",null,"Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Lactobacillales","Lactobacillales"],[null,"ACK-M1","ACK-M1","Micrococcaceae","Bifidobacteriaceae","Fusobacteriaceae",null,null,null,null,"Nostocaceae","Nostocaceae","Synechococcaceae",null,"Neisseriaceae","Neisseriaceae","Pasteurellaceae","Pasteurellaceae","Enterobacteriaceae","Halomonadaceae","Rhodobacteraceae",null,"Bacteroidaceae","Bacteroidaceae","Bacteroidaceae","Bacteroidaceae","Bacteroidaceae","Prevotellaceae","Bacteroidaceae","Bacteroidaceae","Rikenellaceae","Clostridiaceae","ClostridialesFamilyXI.IncertaeSedis","Ruminococcaceae","Lachnospiraceae","Lachnospiraceae","Ruminococcaceae","Ruminococcaceae","Ruminococcaceae","Ruminococcaceae","Streptococcaceae","Streptococcaceae"],[null,null,null,"Rothia","Bifidobacterium","Fusobacterium",null,null,null,null,"Dolichospermum",null,"Prochlorococcus",null,"Neisseria","Kingella","Actinobacillus","Haemophilus","Providencia","CandidatusPortiera",null,null,"Bacteroides","Bacteroides","Bacteroides","Bacteroides","Bacteroides","Prevotella","Bacteroides","Bacteroides","Alistipes","Clostridium","Anaerococcus",null,"Ruminococcus","Lachnospira","Faecalibacterium","Faecalibacterium",null,null,"Streptococcus","Streptococcus"],[null,null,null,"Rothiamucilaginosa","Bifidobacteriumadolescentis",null,null,null,null,null,null,null,null,null,null,null,"Actinobacillusporcinus","Haemophilusparainfluenzae",null,null,null,null,"Bacteroidesuniformis","Bacteroidescoprophilus",null,null,null,"Prevotellamelaninogenica",null,"Bacteroidescaccae","Alistipesputredinis",null,"Anaerococcushydrogenalis",null,"Ruminococcustorques",null,null,"Faecalibacteriumprausnitzii",null,null,null,"Streptococcussanguinis"],[196550,234024,436539,104260,291577,155864,528629,2481214,143627,104199,927850,122697,169240,212089,556441,196355,158228,444142,384205,139264,177026,124393,169042,133795,443938,1001492,222365,146594,109007,149891,149352,177879,106772,179312,137132,101484,110044,215470,312161,152940,106399,203583]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Kingdom<\/th>\n      <th>Phylum<\/th>\n      <th>Class<\/th>\n      <th>Order<\/th>\n      <th>Family<\/th>\n      <th>Genus<\/th>\n      <th>Species<\/th>\n      <th>nb_seq<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":8},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Kingdom","targets":1},{"name":"Phylum","targets":2},{"name":"Class","targets":3},{"name":"Order","targets":4},{"name":"Family","targets":5},{"name":"Genus","targets":6},{"name":"Species","targets":7},{"name":"nb_seq","targets":8}],"order":[],"autoWidth":false,"orderClasses":false,"rowCallback":"function(row, data, displayNum, displayIndex, dataIndex) {\nvar value=data[8]; $(this.api().cell(row, 8).node()).css({'background':isNaN(parseFloat(value)) || value <= 101484.000000 ? '' : 'linear-gradient(90.000000deg, transparent ' + Math.max(2481214.000000 - value, 0)/2379730.000000 * 100 + '%, steelblue ' + Math.max(2481214.000000 - value, 0)/2379730.000000 * 100 + '%)','background-size':'100% 90%','background-repeat':'no-repeat','background-position':'center'});\n}"}},"evals":["options.rowCallback"],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="sankey-diagram-of-the-tax_table">Sankey diagram of the tax_table<a class="anchor" aria-label="anchor" href="#sankey-diagram-of-the-tax_table"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, <span class="va">GlobalPatterns</span><span class="op">@</span><span class="va">tax_table</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">==</span> <span class="st">"Archaea"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/sankey_pq.html">sankey_pq</a></span><span class="op">(</span><span class="va">gp</span>, taxa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sankeyNetwork html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"links":{"source":[0,0,1,2,2,2,1,1,1,2,1,3,4,5,6,7,8,8,4,9,3,10,11,12,13,13,14,15,16,16,13,14,17,13,11,18,13,14],"target":[1,2,3,6,7,8,19,9,4,5,10,20,11,13,12,14,15,16,17,21,22,18,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38],"value":[106,102,27,7,27,24,8,8,57,44,3,7,37,44,7,27,12,12,20,5,18,1,34,7,10,1,25,2,5,6,3,1,20,16,3,1,12,1]},"nodes":{"name":["Archaea","Crenarchaeota","Euryarchaeota","C2","Thaumarchaeota","Thermoplasmata","Halobacteria","Methanobacteria","Methanomicrobia","SdNA","Thermoprotei","Cenarchaeales","Halobacteriales","E2","Methanobacteriales","Methanomicrobiales","Methanosarcinales","Nitrososphaerales","Sulfolobales","pMC2A209","B10","NRPJ","pGrfC26","Cenarchaeaceae","Halobacteriaceae","MarinegroupII","MarinegroupIII","Methanobacteriaceae","Methanocorpusculaceae","Methanosaetaceae","Methanosarcinaceae","MOB79","MSBL1","Nitrososphaeraceae","pMC2A33","SAGMAX","Sulfolobaceae","WCHD302","WSA2"],"group":["Archaea","Crenarchaeota","Euryarchaeota","C2","Thaumarchaeota","Thermoplasmata","Halobacteria","Methanobacteria","Methanomicrobia","SdNA","Thermoprotei","Cenarchaeales","Halobacteriales","E2","Methanobacteriales","Methanomicrobiales","Methanosarcinales","Nitrososphaerales","Sulfolobales","pMC2A209","B10","NRPJ","pGrfC26","Cenarchaeaceae","Halobacteriaceae","MarinegroupII","MarinegroupIII","Methanobacteriaceae","Methanocorpusculaceae","Methanosaetaceae","Methanosarcinaceae","MOB79","MSBL1","Nitrososphaeraceae","pMC2A33","SAGMAX","Sulfolobaceae","WCHD302","WSA2"]},"options":{"NodeID":"name","NodeGroup":"name","LinkGroup":null,"colourScale":"d3.scaleOrdinal(d3.schemeCategory20);","fontSize":7,"fontFamily":null,"nodeWidth":15,"nodePadding":10,"units":"OTUs","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="upset-plot-for-visualize-distribution-of-taxa-in-function-of-samples-variables">Upset plot for visualize distribution of taxa in function of samples
variables<a class="anchor" aria-label="anchor" href="#upset-plot-for-visualize-distribution-of-taxa-in-function-of-samples-variables"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/upset_pq.html">upset_pq</a></span><span class="op">(</span><span class="va">gp</span>, <span class="st">"SampleType"</span>, taxa <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span></span></code></pre></div>
<p><img src="MiscMetabar_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Tengeler, A.C., Dam, S.A., Wiesmann, M. et al. Gut microbiota from
persons with attention-deficit/hyperactivity disorder affects the brain
in mice. Microbiome 8, 44 (2020). <a href="https://doi.org/10.1186/s40168-020-00816-x" class="external-link uri">https://doi.org/10.1186/s40168-020-00816-x</a></p>
</div>
<div class="section level2">
<h2 id="session-inform">Session inform<a class="anchor" aria-label="anchor" href="#session-inform"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2025-03-10 r87922)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=fr_FR.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Paris</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] MiscMetabar_0.14.2 purrr_1.0.4        dplyr_1.1.4        dada2_1.34.0      </span></span>
<span><span class="co">#&gt; [5] Rcpp_1.0.14        ggplot2_3.5.1      phyloseq_1.50.0   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-9                deldir_2.0-4               </span></span>
<span><span class="co">#&gt;   [3] permute_0.9-7               rlang_1.1.5                </span></span>
<span><span class="co">#&gt;   [5] magrittr_2.0.3              ade4_1.7-22                </span></span>
<span><span class="co">#&gt;   [7] matrixStats_1.5.0           compiler_4.5.0             </span></span>
<span><span class="co">#&gt;   [9] mgcv_1.9-1                  png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [11] systemfonts_1.2.1           vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [13] reshape2_1.4.4              stringr_1.5.1              </span></span>
<span><span class="co">#&gt;  [15] pwalign_1.0.0               pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [17] crayon_1.5.3                fastmap_1.2.0              </span></span>
<span><span class="co">#&gt;  [19] XVector_0.44.0              labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [21] Rsamtools_2.20.0            rmarkdown_2.29             </span></span>
<span><span class="co">#&gt;  [23] UCSC.utils_1.0.0            ragg_1.3.3                 </span></span>
<span><span class="co">#&gt;  [25] xfun_0.51                   zlibbioc_1.50.0            </span></span>
<span><span class="co">#&gt;  [27] cachem_1.1.0                GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.9.1              biomformat_1.32.0          </span></span>
<span><span class="co">#&gt;  [31] rhdf5filters_1.16.0         DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt;  [33] Rhdf5lib_1.26.0             BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt;  [35] jpeg_0.1-10                 parallel_4.5.0             </span></span>
<span><span class="co">#&gt;  [37] cluster_2.1.8               R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [39] bslib_0.9.0                 stringi_1.8.4              </span></span>
<span><span class="co">#&gt;  [41] RColorBrewer_1.1-3          ComplexUpset_1.3.3         </span></span>
<span><span class="co">#&gt;  [43] GenomicRanges_1.56.2        jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [45] SummarizedExperiment_1.34.0 iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [47] knitr_1.49                  IRanges_2.38.1             </span></span>
<span><span class="co">#&gt;  [49] Matrix_1.7-3                splines_4.5.0              </span></span>
<span><span class="co">#&gt;  [51] igraph_2.1.4                tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [53] abind_1.4-8                 yaml_2.3.10                </span></span>
<span><span class="co">#&gt;  [55] vegan_2.6-10                codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [57] hwriter_1.3.2.1             lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [59] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [61] Biobase_2.64.0              withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [63] ShortRead_1.62.0            evaluate_1.0.3             </span></span>
<span><span class="co">#&gt;  [65] desc_1.4.3                  survival_3.8-3             </span></span>
<span><span class="co">#&gt;  [67] RcppParallel_5.1.10         Biostrings_2.72.1          </span></span>
<span><span class="co">#&gt;  [69] pillar_1.10.1               MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt;  [71] DT_0.33                     foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [73] stats4_4.5.0                generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [75] S4Vectors_0.42.1            munsell_0.5.1              </span></span>
<span><span class="co">#&gt;  [77] scales_1.3.0                glue_1.8.0                 </span></span>
<span><span class="co">#&gt;  [79] tools_4.5.0                 interp_1.1-6               </span></span>
<span><span class="co">#&gt;  [81] data.table_1.17.0           GenomicAlignments_1.40.0   </span></span>
<span><span class="co">#&gt;  [83] fs_1.6.5                    rhdf5_2.48.0               </span></span>
<span><span class="co">#&gt;  [85] grid_4.5.0                  tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [87] ape_5.8-1                   crosstalk_1.2.1            </span></span>
<span><span class="co">#&gt;  [89] latticeExtra_0.6-30         colorspace_2.1-1           </span></span>
<span><span class="co">#&gt;  [91] patchwork_1.3.0             networkD3_0.4              </span></span>
<span><span class="co">#&gt;  [93] nlme_3.1-167                GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt;  [95] cli_3.6.4                   textshaping_1.0.0          </span></span>
<span><span class="co">#&gt;  [97] S4Arrays_1.4.1              gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [99] sass_0.4.9                  digest_0.6.37              </span></span>
<span><span class="co">#&gt; [101] BiocGenerics_0.50.0         SparseArray_1.4.8          </span></span>
<span><span class="co">#&gt; [103] farver_2.1.2                htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt; [105] htmltools_0.5.8.1           pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt; [107] multtest_2.60.0             lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [109] httr_1.4.7                  MASS_7.3-65</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
