<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reclustering • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Reclustering">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reclustering</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/articles/Reclustering.Rmd" class="external-link"><code>vignettes/articles/Reclustering.Rmd</code></a></small>
      <div class="d-none name"><code>Reclustering.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">MiscMetabar</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="re-clustering-asvs">Re-clustering ASVs<a class="anchor" aria-label="anchor" href="#re-clustering-asvs"></a>
</h2>
<p><strong>ASV</strong> (stands for <em>Amplicon Sequence Variant</em>;
also called <strong>ESV</strong> for Exact Amplicon Variant) is a DNA
sequence obtained from high-throughput analysis of marker genes.
<strong>OTU</strong> (stands for <em>Operational Taxonomic Unit</em>) is
a group of closely related individuals created by clustering sequences
based on a threshold of similarity. An ASV is a special case of an OTU
with a similarity threshold of 100%. A third concept is the zero-radius
OTU <strong>zOTU</strong> <span class="citation">(Edgar 2016)</span>
which is the same concept than ASV but compute with other softwares than
<a href="https://github.com/benjjneb/dada2" class="external-link">dada</a> (e.g. <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a>).</p>
<p>The choice between ASV and OTU is important because they lead to
different results (<span class="citation">Joos et al. (2020)</span>, Box
2 in <span class="citation">Tedersoo et al. (2022)</span>, <span class="citation">Chiarello et al. (2022)</span>). Most articles
recommend making a choice depending on the question <span class="citation">(<strong>mclaren2018?</strong>)</span>, For example,
ASV may be better than OTU for describing a group of very closely
related species. In addition, ASV are comparable across different
datasets (obtained using identical marker genes). <span class="citation">(<strong>fasolo2024?</strong>)</span> showed that that
the OTUs clustering of 16S rDNA proportionally led to a marked
underestimation of the ecological indicators values for species
diversity and to a distorted behaviour of the dominance and evenness
indexes with respect to the direct use of the ASV data. On the other
hand, <span class="citation">(Tedersoo et al. 2022)</span> report that
ASV approaches overestimate the richness of common fungal species (due
to haplotype variation), but underestimate the richness of rare species.
They recommend the OTUs approach in metabarcoding analyses of fungal
communities. Finally, <span class="citation">(Kauserud 2023)</span>
argues that the ASV term falls within the original OTU term and
recommends adopting only the OTU terms, but with a concise and clear
report on how the OTUs were generated.</p>
<p>Recent articles <span class="citation">(Forster et al. 2019; Antich
et al. 2021; Brandt et al. 2021)</span> propose to use both approaches
together. They recommend (i) using ASV to denoise the dataset and (ii)
for some questions, clustering the ASV sequences into OTUs. <span class="citation">(García-García et al. 2019)</span> used both concept to
demonstrate that ecotypes (ASV within OTUs) are adapted to different
values of environmental factors favoring the persistence of OTU across
changing environmental conditions.</p>
<p>The goal of the function <code><a href="../reference/postcluster_pq.html">asv2otu()</a></code> is to facilitate the
reclustering of ASV into OTU, using either the
<code><a href="https://rdrr.io/pkg/DECIPHER/man/Clusterize.html" class="external-link">DECIPHER::Clusterize</a></code> function from R or the <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a> software.</p>
<div class="section level3">
<h3 id="using-decipher-or-vsearch-algorithm">Using decipher or Vsearch algorithm<a class="anchor" aria-label="anchor" href="#using-decipher-or-vsearch-algorithm"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span>
<span><span class="va">otu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/postcluster_pq.html">asv2otu</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span>, method <span class="op">=</span> <span class="st">"clusterize"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Partitioning sequences by 5-mer similarity:</span></span>
<span><span class="co">#&gt; ================================================================================</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Time difference of 0.21 secs</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Sorting by relatedness within 651 groups:</span></span>
<span><span class="co">#&gt; Clustering sequences by 9-mer similarity:</span></span>
<span><span class="co">#&gt; ================================================================================</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Time difference of 1.79 secs</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Clusters via relatedness sorting: 100% (0% exclusively)</span></span>
<span><span class="co">#&gt; Clusters via rare 5-mers: 100% (0% exclusively)</span></span>
<span><span class="co">#&gt; Estimated clustering effectiveness: 100%</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">otu_vs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/postcluster_pq.html">asv2otu</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span>, method <span class="op">=</span> <span class="st">"vsearch"</span><span class="op">)</span></span></code></pre></div>
<p>The vsearch method requires the installation of <a href="https://github.com/torognes/vsearch" class="external-link">Vsearch</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">otu</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="using-lulu-algorithm-link-to-lulu-article">Using lulu algorithm (<a href="https://www.nature.com/articles/s41467-017-01312-x" class="external-link">link to LULU
article</a>)<a class="anchor" aria-label="anchor" href="#using-lulu-algorithm-link-to-lulu-article"></a>
</h3>
<p>Another post-clustering transformation method is implemented in
<code><a href="../reference/lulu_pq.html">lulu_pq()</a></code>, which uses <span class="citation">Frøslev et al.
(2017)</span>’s method for curation of DNA amplicon data. The aim is
more to clean non-biological information than to make explicitly less
clusters. For examples, <span class="citation">(Brandt et al.
2021)</span> clustered amplicon sequence variants (ASVs) into
operational taxonomic units (OTUs) with swarm and choose to curate
ASVs/OTUs using LULU.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span>
<span><span class="va">lulu_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lulu_pq.html">lulu_pq</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">lulu_res</span><span class="op">$</span><span class="va">new_physeq</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="tracking-number-of-samples-sequences-and-clusters">Tracking number of samples, sequences and clusters<a class="anchor" aria-label="anchor" href="#tracking-number-of-samples-sequences-and-clusters"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"Raw data"</span> <span class="op">=</span> <span class="va">data_fungi_sp_known</span>,</span>
<span>  <span class="st">"OTU"</span> <span class="op">=</span> <span class="va">otu</span>,</span>
<span>  <span class="st">"OTU_vsearch"</span> <span class="op">=</span> <span class="va">otu_vs</span>,</span>
<span>  <span class="st">"LULU"</span> <span class="op">=</span> <span class="va">lulu_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;             nb_sequences nb_clusters nb_samples</span></span>
<span><span class="co">#&gt; Raw data         1106581         651        185</span></span>
<span><span class="co">#&gt; OTU              1106581         363        185</span></span>
<span><span class="co">#&gt; OTU_vsearch      1106581         362        185</span></span>
<span><span class="co">#&gt; LULU             1106581         549        185</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=fr_FR.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Paris</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] MiscMetabar_0.14.3 purrr_1.0.4        dplyr_1.1.4        dada2_1.34.0      </span></span>
<span><span class="co">#&gt; [5] Rcpp_1.0.14        ggplot2_3.5.1      phyloseq_1.50.0   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.3                   bitops_1.0-9               </span></span>
<span><span class="co">#&gt;   [3] pbapply_1.7-2               deldir_2.0-4               </span></span>
<span><span class="co">#&gt;   [5] permute_0.9-7               rlang_1.1.5                </span></span>
<span><span class="co">#&gt;   [7] magrittr_2.0.3              ade4_1.7-22                </span></span>
<span><span class="co">#&gt;   [9] matrixStats_1.5.0           compiler_4.4.3             </span></span>
<span><span class="co">#&gt;  [11] mgcv_1.9-1                  png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [13] systemfonts_1.2.1           vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [15] reshape2_1.4.4              stringr_1.5.1              </span></span>
<span><span class="co">#&gt;  [17] pwalign_1.0.0               pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [19] crayon_1.5.3                fastmap_1.2.0              </span></span>
<span><span class="co">#&gt;  [21] XVector_0.44.0              labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [23] Rsamtools_2.20.0            rmarkdown_2.29             </span></span>
<span><span class="co">#&gt;  [25] UCSC.utils_1.0.0            ragg_1.3.3                 </span></span>
<span><span class="co">#&gt;  [27] xfun_0.51                   zlibbioc_1.50.0            </span></span>
<span><span class="co">#&gt;  [29] cachem_1.1.0                GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [31] jsonlite_1.9.1              biomformat_1.32.0          </span></span>
<span><span class="co">#&gt;  [33] rhdf5filters_1.16.0         DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt;  [35] Rhdf5lib_1.26.0             BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt;  [37] jpeg_0.1-10                 parallel_4.4.3             </span></span>
<span><span class="co">#&gt;  [39] cluster_2.1.8               R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [41] bslib_0.9.0                 stringi_1.8.4              </span></span>
<span><span class="co">#&gt;  [43] RColorBrewer_1.1-3          GenomicRanges_1.56.2       </span></span>
<span><span class="co">#&gt;  [45] jquerylib_0.1.4             SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt;  [47] iterators_1.0.14            knitr_1.49                 </span></span>
<span><span class="co">#&gt;  [49] DECIPHER_3.2.0              IRanges_2.38.1             </span></span>
<span><span class="co">#&gt;  [51] Matrix_1.7-2                splines_4.4.3              </span></span>
<span><span class="co">#&gt;  [53] igraph_2.1.4                tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [55] abind_1.4-8                 yaml_2.3.10                </span></span>
<span><span class="co">#&gt;  [57] vegan_2.6-10                codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [59] hwriter_1.3.2.1             lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [61] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [63] Biobase_2.64.0              withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [65] ShortRead_1.62.0            evaluate_1.0.3             </span></span>
<span><span class="co">#&gt;  [67] desc_1.4.3                  survival_3.8-3             </span></span>
<span><span class="co">#&gt;  [69] RcppParallel_5.1.10         Biostrings_2.72.1          </span></span>
<span><span class="co">#&gt;  [71] pillar_1.10.1               MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt;  [73] foreach_1.5.2               stats4_4.4.3               </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3              S4Vectors_0.42.1           </span></span>
<span><span class="co">#&gt;  [77] munsell_0.5.1               scales_1.3.0               </span></span>
<span><span class="co">#&gt;  [79] glue_1.8.0                  tools_4.4.3                </span></span>
<span><span class="co">#&gt;  [81] interp_1.1-6                data.table_1.17.0          </span></span>
<span><span class="co">#&gt;  [83] GenomicAlignments_1.40.0    fs_1.6.5                   </span></span>
<span><span class="co">#&gt;  [85] rhdf5_2.48.0                grid_4.4.3                 </span></span>
<span><span class="co">#&gt;  [87] ape_5.8-1                   latticeExtra_0.6-30        </span></span>
<span><span class="co">#&gt;  [89] colorspace_2.1-1            nlme_3.1-167               </span></span>
<span><span class="co">#&gt;  [91] GenomeInfoDbData_1.2.12     cli_3.6.4                  </span></span>
<span><span class="co">#&gt;  [93] textshaping_1.0.0           S4Arrays_1.4.1             </span></span>
<span><span class="co">#&gt;  [95] gtable_0.3.6                sass_0.4.9                 </span></span>
<span><span class="co">#&gt;  [97] digest_0.6.37               BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt;  [99] SparseArray_1.4.8           farver_2.1.2               </span></span>
<span><span class="co">#&gt; [101] htmlwidgets_1.6.4           htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [103] pkgdown_2.1.1               multtest_2.60.0            </span></span>
<span><span class="co">#&gt; [105] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [107] MASS_7.3-65</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-antich2021" class="csl-entry">
Antich, Adrià, Creu Palacin, Owen S Wangensteen, and Xavier Turon. 2021.
<span>“To Denoise or to Cluster, That Is Not the Question: Optimizing
Pipelines for COI Metabarcoding and Metaphylogeography.”</span> <em>BMC
Bioinformatics</em> 22: 1–24. <a href="https://doi.org/10.1101/2021.01.08.425760" class="external-link">https://doi.org/10.1101/2021.01.08.425760</a>.
</div>
<div id="ref-brandt2021" class="csl-entry">
Brandt, Miriam I., Blandine Trouche, Laure Quintric, Babett Günther,
Patrick Wincker, Julie Poulain, and Sophie Arnaud-Haond. 2021.
<span>“Bioinformatic Pipelines Combining Denoising and Clustering Tools
Allow for More Comprehensive Prokaryotic and Eukaryotic
Metabarcoding.”</span> <em>Molecular Ecology Resources</em> 21 (6):
1904–21. https://doi.org/<a href="https://doi.org/10.1111/1755-0998.13398" class="external-link">https://doi.org/10.1111/1755-0998.13398</a>.
</div>
<div id="ref-chiarello2022" class="csl-entry">
Chiarello, Marlène, Mark McCauley, Sébastien Villéger, and Colin R
Jackson. 2022. <span>“Ranking the Biases: The Choice of OTUs Vs. ASVs in
16S rRNA Amplicon Data Analysis Has Stronger Effects on Diversity
Measures Than Rarefaction and OTU Identity Threshold.”</span> <em>PLoS
One</em> 17 (2): e0264443. <a href="https://doi.org/10.1371/journal.pone.0264443" class="external-link">https://doi.org/10.1371/journal.pone.0264443</a>.
</div>
<div id="ref-edgar2016" class="csl-entry">
Edgar, Robert C. 2016. <span>“UNOISE2: Improved Error-Correction for
Illumina 16S and ITS Amplicon Sequencing.”</span> <em>BioRxiv</em>,
081257.
</div>
<div id="ref-forster2019" class="csl-entry">
Forster, Dominik, Guillaume Lentendu, Sabine Filker, Elyssa Dubois,
Thomas A Wilding, and Thorsten Stoeck. 2019. <span>“Improving eDNA-Based
Protist Diversity Assessments Using Networks of Amplicon Sequence
Variants.”</span> <em>Environmental Microbiology</em> 21 (11): 4109–24.
<a href="https://doi.org/10.1111/1462-2920.14764" class="external-link">https://doi.org/10.1111/1462-2920.14764</a>.
</div>
<div id="ref-froslev2017" class="csl-entry">
Frøslev, Tobias Guldberg, Rasmus Kjøller, Hans Henrik Bruun, Rasmus
Ejrnæs, Ane Kirstine Brunbjerg, Carlotta Pietroni, and Anders Johannes
Hansen. 2017. <span>“Algorithm for Post-Clustering Curation of DNA
Amplicon Data Yields Reliable Biodiversity Estimates.”</span> <em>Nature
Communications</em> 8 (1): 1188. <a href="https://doi.org/10.1038/s41467-017-01312-x" class="external-link">https://doi.org/10.1038/s41467-017-01312-x</a>.
</div>
<div id="ref-garcia2019" class="csl-entry">
García-García, Natalia, Javier Tamames, Alexandra M Linz, Carlos
Pedrós-Alió, and Fernando Puente-Sánchez. 2019. <span>“<span class="nocase">Microdiversity ensures the maintenance of functional
microbial communities under changing environmental
conditions</span>.”</span> <em>The ISME Journal</em> 13 (12): 2969–83.
<a href="https://doi.org/10.1038/s41396-019-0487-8" class="external-link">https://doi.org/10.1038/s41396-019-0487-8</a>.
</div>
<div id="ref-joos2020" class="csl-entry">
Joos, Lisa, Stien Beirinckx, Annelies Haegeman, Jane Debode, Bart
Vandecasteele, Steve Baeyen, Sofie Goormachtig, Lieven Clement, and
Caroline De Tender. 2020. <span>“Daring to Be Differential:
Metabarcoding Analysis of Soil and Plant-Related Microbial Communities
Using Amplicon Sequence Variants and Operational Taxonomical
Units.”</span> <em>BMC Genomics</em> 21 (1): 1–17. <a href="https://doi.org/10.1186/s12864-020-07126-4" class="external-link">https://doi.org/10.1186/s12864-020-07126-4</a>.
</div>
<div id="ref-kauserud2023" class="csl-entry">
Kauserud, Håvard. 2023. <span>“ITS Alchemy: On the Use of ITS as a DNA
Marker in Fungal Ecology.”</span> <em>Fungal Ecology</em>, 101274. <a href="https://doi.org/10.1016/j.funeco.2023.101274" class="external-link">https://doi.org/10.1016/j.funeco.2023.101274</a>.
</div>
<div id="ref-tedersoo2022" class="csl-entry">
Tedersoo, Leho, Mohammad Bahram, Lucie Zinger, R Henrik Nilsson, Peter G
Kennedy, Teng Yang, Sten Anslan, and Vladimir Mikryukov. 2022.
<span>“Best Practices in Metabarcoding of Fungi: From Experimental
Design to Results.”</span> <em>Molecular Ecology</em> 31 (10): 2769–95.
<a href="https://doi.org/10.22541/au.163430390.04226544/v1" class="external-link">https://doi.org/10.22541/au.163430390.04226544/v1</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
