<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Examples with published dataset: Tengeler • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Examples with published dataset: Tengeler">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Examples with published dataset: Tengeler</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/articles/tengeler.Rmd" class="external-link"><code>vignettes/articles/tengeler.Rmd</code></a></small>
      <div class="d-none name"><code>tengeler.Rmd</code></div>
    </div>

    
    
<p>This tutorial explore a phyloseq version of the dataset from Tengeler
et al. (2020) available in the <code>mia</code> package.</p>
<div class="section level2">
<h2 id="load-library">Load library<a class="anchor" aria-label="anchor" href="#load-library"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/YuLab-SMU/MicrobiotaProcess/" class="external-link">"MicrobiotaProcess"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">"MiscMetabar"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://chao.stat.nthu.edu.tw/wordpress/software_download/" class="external-link">"iNEXT"</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">Tengeler2020</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="import-dataset-in-phyloseq-format">Import dataset in phyloseq format<a class="anchor" aria-label="anchor" href="#import-dataset-in-phyloseq-format"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Tengeler2020_pq</span><span class="op">)</span></span>
<span><span class="va">ten</span> <span class="op">&lt;-</span> <span class="va">Tengeler2020_pq</span></span>
<span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">ten</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-4-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="alpha-diversity-analysis">Alpha-diversity analysis<a class="anchor" aria-label="anchor" href="#alpha-diversity-analysis"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hill_pq.html">hill_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"patient_status"</span>, one_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-5-1.png" width="960"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_inext</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/iNEXT_pq.html">iNEXT_pq</a></span><span class="op">(</span><span class="va">ten</span>,</span>
<span>    datatype <span class="op">=</span> <span class="st">"abundance"</span>,</span>
<span>    merge_sample_by <span class="op">=</span> <span class="st">"patient_status_vs_cohort"</span>,</span>
<span>    nboot <span class="op">=</span> <span class="fl">5</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/iNEXT/man/ggiNEXT.html" class="external-link">ggiNEXT</a></span><span class="op">(</span><span class="va">res_inext</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-6-1.png" width="960"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/accu_plot.html">accu_plot</a></span><span class="op">(</span></span>
<span>  <span class="va">ten</span>,</span>
<span>  fact <span class="op">=</span> <span class="st">"sample_name"</span>,</span>
<span>  add_nb_seq <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  by.fact <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.8</span>, <span class="fl">.6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 9</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 17</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 25</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 7</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 8</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 6</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 14</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 28</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 7</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 2</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 15</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 5</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 5</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 6</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 4</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 6</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 21</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 5</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 13</span></span>
<span><span class="co">#&gt; Warning: A numeric `legend.position` argument in `theme()` was deprecated in ggplot2</span></span>
<span><span class="co">#&gt; 3.5.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `legend.position.inside` argument of `theme()` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-7-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="explore-taxonomy">Explore taxonomy<a class="anchor" aria-label="anchor" href="#explore-taxonomy"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library(metacoder)</span></span>
<span><span class="co"># heat_tree_pq(</span></span>
<span><span class="co">#   ten,</span></span>
<span><span class="co">#   node_size = n_obs,</span></span>
<span><span class="co">#   node_color = nb_sequences,</span></span>
<span><span class="co">#   node_label = taxon_names,</span></span>
<span><span class="co">#   tree_label = taxon_names,</span></span>
<span><span class="co">#   node_size_trans = "log10 area"</span></span>
<span><span class="co"># )</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/treemap_pq.html">treemap_pq</a></span><span class="op">(</span><span class="va">ten</span>, lvl1 <span class="op">=</span> <span class="st">"Order"</span>, lvl2 <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-9-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="beta-diversity-analysis-effect-of-patient-status-and-cohort">Beta-diversity analysis : effect of patient status and cohort<a class="anchor" aria-label="anchor" href="#beta-diversity-analysis-effect-of-patient-status-and-cohort"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/circle_pq.html">circle_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"patient_status"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-10-1.png" width="960"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/upset_pq.html">upset_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"patient_status_vs_cohort"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-11-1.png" width="960"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggvenn_pq.html">ggvenn_pq</a></span><span class="op">(</span><span class="fu"><a href="../reference/clean_pq.html">clean_pq</a></span><span class="op">(</span><span class="va">ten</span>, force_taxa_as_columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  <span class="st">"cohort"</span>,</span>
<span>  rarefy_before_merging <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ten_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clean_pq.html">clean_pq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">ten</span>, <span class="va">patient_status</span> <span class="op">==</span> <span class="st">"Control"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># p_control &lt;- heat_tree_pq(</span></span>
<span><span class="co">#   ten_control,</span></span>
<span><span class="co">#   node_size = n_obs,</span></span>
<span><span class="co">#   node_color = nb_sequences,</span></span>
<span><span class="co">#   node_label = taxon_names,</span></span>
<span><span class="co">#   tree_label = taxon_names,</span></span>
<span><span class="co">#   node_size_trans = "log10 area"</span></span>
<span><span class="co"># )</span></span>
<span></span>
<span><span class="va">ten_ADHD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clean_pq.html">clean_pq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">ten</span>, <span class="va">patient_status</span> <span class="op">==</span> <span class="st">"ADHD"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># p_ADHD &lt;- heat_tree_pq(</span></span>
<span><span class="co">#   ten_ADHD,</span></span>
<span><span class="co">#   node_size = n_obs,</span></span>
<span><span class="co">#   node_color = nb_sequences,</span></span>
<span><span class="co">#   node_label = taxon_names,</span></span>
<span><span class="co">#   tree_label = taxon_names,</span></span>
<span><span class="co">#   node_size_trans = "log10 area"</span></span>
<span><span class="co"># )</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># p_control + ggtitle("Control") + p_ADHD + ggtitle("ADHD")</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"All samples"</span> <span class="op">=</span> <span class="va">ten</span>,</span>
<span>  <span class="st">"Control samples"</span> <span class="op">=</span> <span class="va">ten_control</span>,</span>
<span>  <span class="st">"ADHD samples"</span> <span class="op">=</span> <span class="va">ten_ADHD</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">nb_sequences</th>
<th align="right">nb_clusters</th>
<th align="right">nb_samples</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">All samples</td>
<td align="right">485932</td>
<td align="right">151</td>
<td align="right">27</td>
</tr>
<tr class="even">
<td align="left">Control samples</td>
<td align="right">239329</td>
<td align="right">130</td>
<td align="right">14</td>
</tr>
<tr class="odd">
<td align="left">ADHD samples</td>
<td align="right">246603</td>
<td align="right">142</td>
<td align="right">13</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/adonis_pq.html">adonis_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"cohort + patient_status"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Permutation test for adonis under reduced model</span></span>
<span><span class="co">#&gt; Permutation: free</span></span>
<span><span class="co">#&gt; Number of permutations: 999</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; vegan::adonis2(formula = .formula, data = metadata)</span></span>
<span><span class="co">#&gt;          Df SumOfSqs      R2      F Pr(&gt;F)  </span></span>
<span><span class="co">#&gt; Model     3   1.2425 0.18483 1.7383  0.028 *</span></span>
<span><span class="co">#&gt; Residual 23   5.4799 0.81517                </span></span>
<span><span class="co">#&gt; Total    26   6.7223 1.00000                </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ten</span><span class="op">@</span><span class="va">tax_table</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="va">ten</span><span class="op">@</span><span class="va">tax_table</span>,</span>
<span>  <span class="st">"Species"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">ten</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/biplot_pq.html">biplot_pq</a></span><span class="op">(</span><span class="fu"><a href="../reference/subset_taxa_pq.html">subset_taxa_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="va">ten</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3000</span><span class="op">)</span>,</span>
<span>  merge_sample_by <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>  fact <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>  nudge_y <span class="op">=</span> <span class="fl">0.4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-16-1.png" width="960"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/multitax_bar_pq.html">multitax_bar_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"patient_status"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-17-1.png" width="960"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/multitax_bar_pq.html">multitax_bar_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"patient_status"</span>,</span>
<span>  nb_seq <span class="op">=</span> <span class="cn">FALSE</span>, log10trans <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_deseq2_pq.html">plot_deseq2_pq</a></span><span class="op">(</span><span class="va">ten</span>,</span>
<span>  contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"patient_status"</span>, <span class="st">"ADHD"</span>, <span class="st">"Control"</span><span class="op">)</span>,</span>
<span>  taxolev <span class="op">=</span> <span class="st">"Genus"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in</span></span>
<span><span class="co">#&gt; design formula are characters, converting to factors</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-19-1.png" width="960"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LEfSe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/diff_analysis.html" class="external-link">diff_analysis</a></span><span class="op">(</span></span>
<span>  <span class="va">ten</span>,</span>
<span>  classgroup <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>  mlfun <span class="op">=</span> <span class="st">"lda"</span>,</span>
<span>  ldascore <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  p.adjust.methods <span class="op">=</span> <span class="st">"bh"</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/ggeffectsize.html" class="external-link">ggeffectsize</a></span><span class="op">(</span><span class="va">LEfSe</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"#00AED7"</span>,</span>
<span>    <span class="st">"#FD9347"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-20-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Kali GNU/Linux Rolling</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.29.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=fr_FR.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Paris</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] iNEXT_3.0.1              patchwork_1.3.0          MiscMetabar_0.14.3      </span></span>
<span><span class="co">#&gt;  [4] purrr_1.0.4              dplyr_1.1.4              dada2_1.36.0            </span></span>
<span><span class="co">#&gt;  [7] Rcpp_1.0.14              ggplot2_3.5.2            phyloseq_1.52.0         </span></span>
<span><span class="co">#&gt; [10] MicrobiotaProcess_1.20.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] libcoin_1.0-10              RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [3] shape_1.4.6.1               rstudioapi_0.17.1          </span></span>
<span><span class="co">#&gt;   [5] jsonlite_2.0.0              magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [7] TH.data_1.1-3               modeltools_0.2-24          </span></span>
<span><span class="co">#&gt;   [9] farver_2.1.2                rmarkdown_2.29             </span></span>
<span><span class="co">#&gt;  [11] GlobalOptions_0.1.2         fs_1.6.6                   </span></span>
<span><span class="co">#&gt;  [13] ragg_1.4.0                  vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [15] multtest_2.64.0             Rsamtools_2.24.0           </span></span>
<span><span class="co">#&gt;  [17] ggtree_3.16.0               htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt;  [19] S4Arrays_1.8.1              ComplexUpset_1.3.3         </span></span>
<span><span class="co">#&gt;  [21] Rhdf5lib_1.30.0             SparseArray_1.8.0          </span></span>
<span><span class="co">#&gt;  [23] rhdf5_2.52.1                gridGraphics_0.5-1         </span></span>
<span><span class="co">#&gt;  [25] sass_0.4.10                 bslib_0.9.0                </span></span>
<span><span class="co">#&gt;  [27] htmlwidgets_1.6.4           desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [29] plyr_1.8.9                  sandwich_3.1-1             </span></span>
<span><span class="co">#&gt;  [31] zoo_1.8-14                  cachem_1.1.0               </span></span>
<span><span class="co">#&gt;  [33] ggfittext_0.10.2            GenomicAlignments_1.44.0   </span></span>
<span><span class="co">#&gt;  [35] igraph_2.1.4                lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt;  [37] iterators_1.0.14            pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [39] Matrix_1.7-3                R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [41] fastmap_1.2.0               GenomeInfoDbData_1.2.14    </span></span>
<span><span class="co">#&gt;  [43] MatrixGenerics_1.20.0       digest_0.6.37              </span></span>
<span><span class="co">#&gt;  [45] aplot_0.2.5                 colorspace_2.1-1           </span></span>
<span><span class="co">#&gt;  [47] ggnewscale_0.5.1            ShortRead_1.66.0           </span></span>
<span><span class="co">#&gt;  [49] S4Vectors_0.46.0            DESeq2_1.48.1              </span></span>
<span><span class="co">#&gt;  [51] textshaping_1.0.1           GenomicRanges_1.60.0       </span></span>
<span><span class="co">#&gt;  [53] hwriter_1.3.2.1             vegan_2.7-1                </span></span>
<span><span class="co">#&gt;  [55] labeling_0.4.3              httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [57] abind_1.4-8                 mgcv_1.9-3                 </span></span>
<span><span class="co">#&gt;  [59] compiler_4.5.0              withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [61] BiocParallel_1.42.1         ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt;  [63] MASS_7.3-65                 DelayedArray_0.34.1        </span></span>
<span><span class="co">#&gt;  [65] biomformat_1.36.0           permute_0.9-7              </span></span>
<span><span class="co">#&gt;  [67] tools_4.5.0                 ape_5.8-1                  </span></span>
<span><span class="co">#&gt;  [69] glue_1.8.0                  treemapify_2.5.6           </span></span>
<span><span class="co">#&gt;  [71] nlme_3.1-168                rhdf5filters_1.20.0        </span></span>
<span><span class="co">#&gt;  [73] grid_4.5.0                  cluster_2.1.8.1            </span></span>
<span><span class="co">#&gt;  [75] reshape2_1.4.4              ade4_1.7-23                </span></span>
<span><span class="co">#&gt;  [77] generics_0.1.4              gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [79] tidyr_1.3.1                 ggVennDiagram_1.5.2        </span></span>
<span><span class="co">#&gt;  [81] data.table_1.17.4           coin_1.4-3                 </span></span>
<span><span class="co">#&gt;  [83] XVector_0.48.0              BiocGenerics_0.54.0        </span></span>
<span><span class="co">#&gt;  [85] ggrepel_0.9.6               foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [87] pillar_1.10.2               stringr_1.5.1              </span></span>
<span><span class="co">#&gt;  [89] yulab.utils_0.2.0           circlize_0.4.16            </span></span>
<span><span class="co">#&gt;  [91] splines_4.5.0               treeio_1.32.0              </span></span>
<span><span class="co">#&gt;  [93] lattice_0.22-7              survival_3.8-3             </span></span>
<span><span class="co">#&gt;  [95] deldir_2.0-4                tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [97] locfit_1.5-9.12             pbapply_1.7-2              </span></span>
<span><span class="co">#&gt;  [99] Biostrings_2.76.0           knitr_1.50                 </span></span>
<span><span class="co">#&gt; [101] gridExtra_2.3               IRanges_2.42.0             </span></span>
<span><span class="co">#&gt; [103] SummarizedExperiment_1.38.1 ggtreeExtra_1.18.0         </span></span>
<span><span class="co">#&gt; [105] stats4_4.5.0                xfun_0.52                  </span></span>
<span><span class="co">#&gt; [107] Biobase_2.68.0              matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; [109] stringi_1.8.7               UCSC.utils_1.4.0           </span></span>
<span><span class="co">#&gt; [111] lazyeval_0.2.2              ggfun_0.1.8                </span></span>
<span><span class="co">#&gt; [113] yaml_2.3.10                 evaluate_1.0.3             </span></span>
<span><span class="co">#&gt; [115] codetools_0.2-20            interp_1.1-6               </span></span>
<span><span class="co">#&gt; [117] tibble_3.3.0                ggplotify_0.1.2            </span></span>
<span><span class="co">#&gt; [119] cli_3.6.5                   RcppParallel_5.1.10        </span></span>
<span><span class="co">#&gt; [121] systemfonts_1.2.3           jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [123] GenomeInfoDb_1.44.0         png_0.1-8                  </span></span>
<span><span class="co">#&gt; [125] parallel_4.5.0              ggh4x_0.3.1                </span></span>
<span><span class="co">#&gt; [127] pkgdown_2.1.2               latticeExtra_0.6-30        </span></span>
<span><span class="co">#&gt; [129] jpeg_0.1-11                 bitops_1.0-9               </span></span>
<span><span class="co">#&gt; [131] ggstar_1.0.4                pwalign_1.4.0              </span></span>
<span><span class="co">#&gt; [133] mvtnorm_1.3-3               tidytree_0.4.6             </span></span>
<span><span class="co">#&gt; [135] scales_1.4.0                crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [137] rlang_1.1.6                 multcomp_1.4-28</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Tengeler, A.C., Dam, S.A., Wiesmann, M. et al. Gut microbiota from
persons with attention-deficit/hyperactivity disorder affects the brain
in mice. Microbiome 8, 44 (2020). <a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-020-00816-x" class="external-link uri">https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-020-00816-x</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
