<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot accumulation curves with balanced modality and depth rarefaction — accu_plot_balanced_modality • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot accumulation curves with balanced modality and depth rarefaction — accu_plot_balanced_modality"><meta name="description" content="


This function (i) rarefy (equalize) the number of samples per modality of a
factor and (ii) rarefy the number of sequences per sample (depth). The
seed is set to 1:nperm. Thus, with exacly the same parameter, including
nperm values, results must be identical."><meta property="og:description" content="


This function (i) rarefy (equalize) the number of samples per modality of a
factor and (ii) rarefy the number of sequences per sample (depth). The
seed is set to 1:nperm. Thus, with exacly the same parameter, including
nperm values, results must be identical."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.10.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot accumulation curves with balanced modality and depth rarefaction</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/plot_functions.R" class="external-link"><code>R/plot_functions.R</code></a></small>
      <div class="d-none name"><code>accu_plot_balanced_modality.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>This function (i) rarefy (equalize) the number of samples per modality of a
factor and (ii) rarefy the number of sequences per sample (depth). The
seed is set to 1:nperm. Thus, with exacly the same parameter, including
nperm values, results must be identical.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">accu_plot_balanced_modality</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  <span class="va">fact</span>,</span>
<span>  nperm <span class="op">=</span> <span class="fl">99</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  by.fact <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  progress_bar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  quantile_prob <span class="op">=</span> <span class="fl">0.975</span>,</span>
<span>  rarefy_by_sample_before_merging <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sample.size <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-fact">fact<a class="anchor" aria-label="anchor" href="#arg-fact"></a></dt>
<dd><p>(required) The variable to rarefy. Must be present in
the <code>sam_data</code> slot of the physeq object.</p></dd>


<dt id="arg-nperm">nperm<a class="anchor" aria-label="anchor" href="#arg-nperm"></a></dt>
<dd><p>(int) The number of permutations to perform.</p></dd>


<dt id="arg-step">step<a class="anchor" aria-label="anchor" href="#arg-step"></a></dt>
<dd><p>(int) distance among points calculated to plot lines.
A low value give better plot but is more time consuming.</p></dd>


<dt id="arg-by-fact">by.fact<a class="anchor" aria-label="anchor" href="#arg-by-fact"></a></dt>
<dd><p>(logical, default TRUE)
First merge the OTU table by factor to plot only one line by factor</p></dd>


<dt id="arg-progress-bar">progress_bar<a class="anchor" aria-label="anchor" href="#arg-progress-bar"></a></dt>
<dd><p>(logical, default TRUE) Do we print progress during
the calculation?</p></dd>


<dt id="arg-quantile-prob">quantile_prob<a class="anchor" aria-label="anchor" href="#arg-quantile-prob"></a></dt>
<dd><p>(float, <code>[0:1]</code>) the value to compute the quantile.
Minimum quantile is compute using 1-quantile_prob.</p></dd>


<dt id="arg-rarefy-by-sample-before-merging">rarefy_by_sample_before_merging<a class="anchor" aria-label="anchor" href="#arg-rarefy-by-sample-before-merging"></a></dt>
<dd><p>(logical, default TRUE):
rarefy_by_sample_before_merging = FALSE is buggy for the moment.Please
only use rarefy_by_sample_before_merging = TRUE</p></dd>


<dt id="arg-sample-size">sample.size<a class="anchor" aria-label="anchor" href="#arg-sample-size"></a></dt>
<dd><p>(int) A single integer value equal to the number of
reads being simulated, also known as the depth. See
<code><a href="https://rdrr.io/pkg/phyloseq/man/rarefy_even_depth.html" class="external-link">phyloseq::rarefy_even_depth()</a></code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(logical). If TRUE, print additional informations.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other params for be passed on to <code><a href="accu_plot.html">accu_plot()</a></code> function</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot2 plot representing the richness accumulation plot</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="accu_plot.html">accu_plot()</a></code>, <code><a href="rarefy_sample_count_by_modality.html">rarefy_sample_count_by_modality()</a></code>, <code><a href="https://rdrr.io/pkg/phyloseq/man/rarefy_even_depth.html" class="external-link">phyloseq::rarefy_even_depth()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">data_fungi_woNA4Time</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Time</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data_fungi_woNA4Time</span><span class="op">@</span><span class="va">sam_data</span><span class="op">$</span><span class="va">Time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"time-"</span>, <span class="va">data_fungi_woNA4Time</span><span class="op">@</span><span class="va">sam_data</span><span class="op">$</span><span class="va">Time</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">accu_plot_balanced_modality</span><span class="op">(</span><span class="va">data_fungi_woNA4Time</span>, <span class="st">"Time"</span>, nperm <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `set.seed(1)` was used to initialize repeatable random subsampling.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Please record this for your records so others can reproduce.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try `set.seed(1); .Random.seed` for the full vector</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>no non-missing arguments to max; returning -Inf</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                          |                                                  |   0%  |                                                          |=================                                 |  33%  |                                                          |=================================                 |  67%  |                                                          |==================================================| 100%</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 4 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_line()`).</span>
<span class="r-plt img"><img src="accu_plot_balanced_modality-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data_fungi_woNA4Height</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Height</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">accu_plot_balanced_modality</span><span class="op">(</span><span class="va">data_fungi_woNA4Height</span>, <span class="st">"Height"</span>, nperm <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `set.seed(1)` was used to initialize repeatable random subsampling.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Please record this for your records so others can reproduce.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try `set.seed(1); .Random.seed` for the full vector</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>no non-missing arguments to max; returning -Inf</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                          |                                                  |   0%  |                                                          |=================                                 |  33%  |                                                          |=================================                 |  67%  |                                                          |==================================================| 100%</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 3 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_line()`).</span>
<span class="r-plt img"><img src="accu_plot_balanced_modality-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

