<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>A wrapper of IdTaxa — assign_idtaxa • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="A wrapper of IdTaxa — assign_idtaxa"><meta name="description" content="


This function is basically a wrapper of functions DECIPHER::IdTaxa() and
DECIPHER::LearnTaxa(), please cite the DECIPHER package if you use this
function. Note that if you want to specify parameters for the learning step
you must used the trainingSet param instead of the a fasta_for_training. The
training file can be obtain using the function learn_idtaxa().
It requires:
either a physeq or seq2search object.
either a trainingSet or a fasta_for_training

"><meta property="og:description" content="


This function is basically a wrapper of functions DECIPHER::IdTaxa() and
DECIPHER::LearnTaxa(), please cite the DECIPHER package if you use this
function. Note that if you want to specify parameters for the learning step
you must used the trainingSet param instead of the a fasta_for_training. The
training file can be obtain using the function learn_idtaxa().
It requires:
either a physeq or seq2search object.
either a trainingSet or a fasta_for_training

"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>A wrapper of <code><a href="https://rdrr.io/pkg/DECIPHER/man/IdTaxa.html" class="external-link">IdTaxa</a></code></h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/dada_phyloseq.R" class="external-link"><code>R/dada_phyloseq.R</code></a></small>
      <div class="d-none name"><code>assign_idtaxa.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>This function is basically a wrapper of functions <code><a href="https://rdrr.io/pkg/DECIPHER/man/IdTaxa.html" class="external-link">DECIPHER::IdTaxa()</a></code> and
<code><a href="https://rdrr.io/pkg/DECIPHER/man/LearnTaxa.html" class="external-link">DECIPHER::LearnTaxa()</a></code>, please cite the DECIPHER package if you use this
function. Note that if you want to specify parameters for the learning step
you must used the trainingSet param instead of the a fasta_for_training. The
training file can be obtain using the function <code><a href="learn_idtaxa.html">learn_idtaxa()</a></code>.</p>
<p>It requires:</p><ul><li><p>either a physeq or seq2search object.</p></li>
<li><p>either a trainingSet or a fasta_for_training</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">assign_idtaxa</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  trainingSet <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seq2search <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fasta_for_training <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  behavior <span class="op">=</span> <span class="st">"return_matrix"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">60</span>,</span>
<span>  column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span><span class="op">)</span>,</span>
<span>  suffix <span class="op">=</span> <span class="st">"_idtaxa"</span>,</span>
<span>  nproc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  unite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required) a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-trainingset">trainingSet<a class="anchor" aria-label="anchor" href="#arg-trainingset"></a></dt>
<dd><p>An object of class Taxa and subclass
Train compatible with the class of test.</p></dd>


<dt id="arg-seq-search">seq2search<a class="anchor" aria-label="anchor" href="#arg-seq-search"></a></dt>
<dd><p>A DNAStringSet object of sequences to search for. Replace
the physeq object.</p></dd>


<dt id="arg-fasta-for-training">fasta_for_training<a class="anchor" aria-label="anchor" href="#arg-fasta-for-training"></a></dt>
<dd><p>A fasta file (can be gzip) to train the trainingSet
using the function <code><a href="learn_idtaxa.html">learn_idtaxa()</a></code>. Only used if trainingSet is NULL.</p>
<p>The reference database must contain
taxonomic information in the header of each sequence in the form of a string
starting with ";tax=" and followed by a comma-separated list of up to nine
taxonomic identifiers.</p>
<p>The only exception is if <code>unite=TRUE</code>. In that case the UNITE taxonomy is
automatically formatted.</p></dd>


<dt id="arg-behavior">behavior<a class="anchor" aria-label="anchor" href="#arg-behavior"></a></dt>
<dd><p>Either "return_matrix" (default), or "add_to_phyloseq":</p><ul><li><p>"return_matrix" return a list of two objects. The first element is
the taxonomic matrix and the second element is the raw results from
DECIPHER::IdTaxa() function.</p></li>
<li><p>"add_to_phyloseq" return a phyloseq object with amended slot <code>@taxtable</code>.
Only available if using physeq input and not seq2search input.</p></li>
</ul></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>(Int, default 60) Numeric specifying the confidence at which
to truncate the output taxonomic classifications.
Lower values of threshold will classify deeper into the taxonomic tree at
the expense of accuracy, and vise-versa for higher values of threshold. See
<code><a href="https://rdrr.io/pkg/DECIPHER/man/IdTaxa.html" class="external-link">DECIPHER::IdTaxa()</a></code> man page.</p></dd>


<dt id="arg-column-names">column_names<a class="anchor" aria-label="anchor" href="#arg-column-names"></a></dt>
<dd><p>(vector of character) names for the column of the
taxonomy</p></dd>


<dt id="arg-suffix">suffix<a class="anchor" aria-label="anchor" href="#arg-suffix"></a></dt>
<dd><p>(character) The suffix to name the new columns.
Default to "_idtaxa".</p></dd>


<dt id="arg-nproc">nproc<a class="anchor" aria-label="anchor" href="#arg-nproc"></a></dt>
<dd><p>(default: 1)
Set to number of cpus/processors to use</p></dd>


<dt id="arg-unite">unite<a class="anchor" aria-label="anchor" href="#arg-unite"></a></dt>
<dd><p>(logical, default FALSE). If set to TRUE, the fasta_for_training
file is formatted from UNITE format to sintax one, needed in
fasta_for_training. Only used if trainingSet is NULL.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(logical). If TRUE, print additional information.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed on to <code><a href="https://rdrr.io/pkg/DECIPHER/man/IdTaxa.html" class="external-link">IdTaxa</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Either a new phyloseq object with additional information in
the @tax_table slot or a list of two objects if behavior is "return_matrix"</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is mainly a wrapper of the work of others.
Please make a reference to <code><a href="https://rdrr.io/pkg/DECIPHER/man/IdTaxa.html" class="external-link">DECIPHER::IdTaxa()</a></code> if you
use this function.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="assign_sintax.html">assign_sintax()</a></code>, <code><a href="add_new_taxonomy_pq.html">add_new_taxonomy_pq()</a></code>, <code><a href="assign_vsearch_lca.html">assign_vsearch_lca()</a></code>, <code><a href="assign_blastn.html">assign_blastn()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># /!\ The value of threshold must be change for real database (recommend</span></span></span>
<span class="r-in"><span><span class="co">#  value are between 50 and 70).</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data_fungi_mini_new</span> <span class="op">&lt;-</span> <span class="fu">assign_idtaxa</span><span class="op">(</span><span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>  fasta_for_training <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mini_UNITE_fungi.fasta.gz"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"MiscMetabar"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>, threshold <span class="op">=</span> <span class="fl">20</span>, behavior <span class="op">=</span> <span class="st">"add_to_phyloseq"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">result_idtaxa</span> <span class="op">&lt;-</span> <span class="fu">assign_idtaxa</span><span class="op">(</span><span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>  fasta_for_training <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mini_UNITE_fungi.fasta.gz"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"MiscMetabar"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>, threshold <span class="op">=</span> <span class="fl">20</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">result_idtaxa</span><span class="op">$</span><span class="va">idtaxa_raw</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

