<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Assign taxonomy with dada2 using 2 steps assignTaxonomy and assignSpecies — assign_dada2 • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Assign taxonomy with dada2 using 2 steps assignTaxonomy and assignSpecies — assign_dada2"><meta name="description" content="


Mainly a wrapper of dada2::assignTaxonomy() and dada2::assignSpecies()"><meta property="og:description" content="


Mainly a wrapper of dada2::assignTaxonomy() and dada2::assignSpecies()"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Assign taxonomy with dada2 using 2 steps assignTaxonomy and assignSpecies</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/dada_phyloseq.R" class="external-link"><code>R/dada_phyloseq.R</code></a></small>
      <div class="d-none name"><code>assign_dada2.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>Mainly a wrapper of <code><a href="https://rdrr.io/pkg/dada2/man/assignTaxonomy.html" class="external-link">dada2::assignTaxonomy()</a></code> and <code><a href="https://rdrr.io/pkg/dada2/man/assignSpecies.html" class="external-link">dada2::assignSpecies()</a></code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">assign_dada2</span><span class="op">(</span></span>
<span>  physeq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_fasta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seq2search <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_bootstrap <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  tryRC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  taxa_ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span>,</span>
<span>    <span class="st">"taxId"</span><span class="op">)</span>,</span>
<span>  use_assignSpecies <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  trunc_absent_ranks <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nproc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  suffix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seq_at_one_time <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  allowMultiple <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_sintax <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-ref-fasta">ref_fasta<a class="anchor" aria-label="anchor" href="#arg-ref-fasta"></a></dt>
<dd><p>(required) A link to a database in fasta.</p></dd>


<dt id="arg-seq-search">seq2search<a class="anchor" aria-label="anchor" href="#arg-seq-search"></a></dt>
<dd><p>A DNAStringSet object of sequences to search for. Replace
the physeq object.</p></dd>


<dt id="arg-min-bootstrap">min_bootstrap<a class="anchor" aria-label="anchor" href="#arg-min-bootstrap"></a></dt>
<dd><p>(Float [0:1], default 0.5), See <code><a href="https://rdrr.io/pkg/dada2/man/assignTaxonomy.html" class="external-link">dada2::assignTaxonomy()</a></code></p></dd>


<dt id="arg-tryrc">tryRC<a class="anchor" aria-label="anchor" href="#arg-tryrc"></a></dt>
<dd><p>See <code><a href="https://rdrr.io/pkg/dada2/man/assignTaxonomy.html" class="external-link">dada2::assignTaxonomy()</a></code></p></dd>


<dt id="arg-taxa-ranks">taxa_ranks<a class="anchor" aria-label="anchor" href="#arg-taxa-ranks"></a></dt>
<dd><p>(vector of character) names for the column of the
taxonomy</p></dd>


<dt id="arg-use-assignspecies">use_assignSpecies<a class="anchor" aria-label="anchor" href="#arg-use-assignspecies"></a></dt>
<dd><p>(logical, default TRUE) Do the Species rank is
obtained using <code><a href="https://rdrr.io/pkg/dada2/man/assignSpecies.html" class="external-link">dada2::assignSpecies()</a></code> ?</p></dd>


<dt id="arg-trunc-absent-ranks">trunc_absent_ranks<a class="anchor" aria-label="anchor" href="#arg-trunc-absent-ranks"></a></dt>
<dd><p>(logical, default FALSE) Do ranks present in
taxa_ranks but not present in the database are removed ?</p></dd>


<dt id="arg-nproc">nproc<a class="anchor" aria-label="anchor" href="#arg-nproc"></a></dt>
<dd><p>(Float [0:1], default 0.5)</p></dd>


<dt id="arg-suffix">suffix<a class="anchor" aria-label="anchor" href="#arg-suffix"></a></dt>
<dd><p>(character) The suffix to name the new columns.
Default to "_idtaxa".</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(logical). If TRUE, print additional information.</p></dd>


<dt id="arg-seq-at-one-time">seq_at_one_time<a class="anchor" aria-label="anchor" href="#arg-seq-at-one-time"></a></dt>
<dd><p>How many sequences are treated at one time.
See param <code>n</code> in <code><a href="https://rdrr.io/pkg/dada2/man/assignSpecies.html" class="external-link">dada2::assignSpecies()</a></code></p></dd>


<dt id="arg-allowmultiple">allowMultiple<a class="anchor" aria-label="anchor" href="#arg-allowmultiple"></a></dt>
<dd><p>(logical, default FALSE). Unchanged from
<code><a href="https://rdrr.io/pkg/dada2/man/assignSpecies.html" class="external-link">dada2::assignSpecies()</a></code>.
Defines the behavior when multiple exact matches
against different species are returned. By default only unambiguous
identifications are return. If TRUE, a concatenated string of all
exactly matched species is returned. If an integer is provided,
multiple identifications up to that many are returned as
a concatenated string.</p></dd>


<dt id="arg-from-sintax">from_sintax<a class="anchor" aria-label="anchor" href="#arg-from-sintax"></a></dt>
<dd><p>(logical, default FALSE). Set to TRUE
if the ref_fasta database is in sintax format. See <code><a href="assign_sintax.html">assign_sintax()</a></code>
for more information about the sintax format.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Either a an object of class phyloseq (if <code>physeq</code> is not NULL),
or a taxonomic table if <code>seq2search</code> is used in place of <code>physeq</code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">data_fungi_mini2</span> <span class="op">&lt;-</span> <span class="fu">assign_dada2</span><span class="op">(</span><span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>  ref_fasta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mini_UNITE_fungi.fasta.gz"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"MiscMetabar"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>, suffix <span class="op">=</span> <span class="st">"_dada2"</span>,</span></span>
<span class="r-in"><span>  from_sintax <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

