<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Assign taxonomy using blastn algorithm and the blast software — assign_blastn • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Assign taxonomy using blastn algorithm and the blast software — assign_blastn"><meta name="description" content="


Use the blast software."><meta property="og:description" content="


Use the blast software."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.13.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Assign taxonomy using blastn algorithm and the blast software</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/blast.R" class="external-link"><code>R/blast.R</code></a></small>
      <div class="d-none name"><code>assign_blastn.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>Use the blast software.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">assign_blastn</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  ref_fasta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  database <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blastpath <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  behavior <span class="op">=</span> <span class="st">"return_matrix"</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vote"</span>, <span class="st">"top-hit"</span><span class="op">)</span>,</span>
<span>  suffix <span class="op">=</span> <span class="st">"_blastn"</span>,</span>
<span>  min_id <span class="op">=</span> <span class="fl">95</span>,</span>
<span>  min_bit_score <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  min_cover <span class="op">=</span> <span class="fl">95</span>,</span>
<span>  min_e_value <span class="op">=</span> <span class="fl">1e-30</span>,</span>
<span>  nb_voting <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span><span class="op">)</span>,</span>
<span>  vote_algorithm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"consensus"</span>, <span class="st">"rel_majority"</span>, <span class="st">"abs_majority"</span>, <span class="st">"unanimity"</span><span class="op">)</span>,</span>
<span>  strict <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nb_agree_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  preference_index <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  collapse_string <span class="op">=</span> <span class="st">"/"</span>,</span>
<span>  replace_collapsed_rank_by_NA <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  simplify_taxo <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  keep_blast_metrics <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-ref-fasta">ref_fasta<a class="anchor" aria-label="anchor" href="#arg-ref-fasta"></a></dt>
<dd><p>Either a DNAStringSet object or a path to a fasta
file to make the blast database. It must be in sintax format.
See <code><a href="assign_sintax.html">assign_sintax()</a></code>.</p></dd>


<dt id="arg-database">database<a class="anchor" aria-label="anchor" href="#arg-database"></a></dt>
<dd><p>path to a blast database. Only used if ref_fasta
is not set.</p></dd>


<dt id="arg-blastpath">blastpath<a class="anchor" aria-label="anchor" href="#arg-blastpath"></a></dt>
<dd><p>path to blast program.</p></dd>


<dt id="arg-behavior">behavior<a class="anchor" aria-label="anchor" href="#arg-behavior"></a></dt>
<dd><p>Either "return_matrix" (default), or "add_to_phyloseq":</p><ul><li><p>"return_matrix" return a list of two matrix with taxonomic value in the
first element of the list and bootstrap value in the second one.</p></li>
<li><p>"add_to_phyloseq" return a phyloseq object with amended slot <code>@taxtable</code>.
Only available if using physeq input and not seq2search input.</p></li>
</ul></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>(One of "vote" or "top-hit"). If top-hit, only the
better match is used to assign taxonomy. If vote, the algorithm
takes all (or <code>nb_voting</code> if <code>nb_voting</code> is not null) select assignation
and resolve the conflict using the function <code><a href="resolve_vector_ranks.html">resolve_vector_ranks()</a></code>.</p></dd>


<dt id="arg-suffix">suffix<a class="anchor" aria-label="anchor" href="#arg-suffix"></a></dt>
<dd><p>(character) The suffix to name the new columns.
If set to "" (the default), the taxo_rank algorithm is used
without suffix.</p></dd>


<dt id="arg-min-id">min_id<a class="anchor" aria-label="anchor" href="#arg-min-id"></a></dt>
<dd><p>(default: 95) the identity percent to take into account
a references taxa</p></dd>


<dt id="arg-min-bit-score">min_bit_score<a class="anchor" aria-label="anchor" href="#arg-min-bit-score"></a></dt>
<dd><p>(default: 50) the minimum bit score to take
into account a references taxa</p></dd>


<dt id="arg-min-cover">min_cover<a class="anchor" aria-label="anchor" href="#arg-min-cover"></a></dt>
<dd><p>(default: 50) cut of in query cover (%) to keep result</p></dd>


<dt id="arg-min-e-value">min_e_value<a class="anchor" aria-label="anchor" href="#arg-min-e-value"></a></dt>
<dd><p>(default: 1e-30)  cut of in e-value (%) to keep result
The BLAST E-value is the number of expected hits of similar quality (score)
that could be found just by chance.</p></dd>


<dt id="arg-nb-voting">nb_voting<a class="anchor" aria-label="anchor" href="#arg-nb-voting"></a></dt>
<dd><p>(Int, default NULL). The number of taxa to keep before apply
a vote to resolve conflict. If NULL all taxa passing the filters (min_id,
min_bit_score, min_cover and min_e_value) are selected.</p></dd>


<dt id="arg-column-names">column_names<a class="anchor" aria-label="anchor" href="#arg-column-names"></a></dt>
<dd><p>A vector of names for taxonomic ranks. Must
correspond to names in the ref_fasta files.</p></dd>


<dt id="arg-vote-algorithm">vote_algorithm<a class="anchor" aria-label="anchor" href="#arg-vote-algorithm"></a></dt>
<dd><p>the method to vote among "consensus", "rel_majority",
"abs_majority" and "unanimity". See <code><a href="resolve_vector_ranks.html">resolve_vector_ranks()</a></code> for more details.</p></dd>


<dt id="arg-strict">strict<a class="anchor" aria-label="anchor" href="#arg-strict"></a></dt>
<dd><p>(Logical, default FALSE). See <code><a href="resolve_vector_ranks.html">resolve_vector_ranks()</a></code> for more details.</p></dd>


<dt id="arg-nb-agree-threshold">nb_agree_threshold<a class="anchor" aria-label="anchor" href="#arg-nb-agree-threshold"></a></dt>
<dd><p>See <code><a href="resolve_vector_ranks.html">resolve_vector_ranks()</a></code> for more details.</p></dd>


<dt id="arg-preference-index">preference_index<a class="anchor" aria-label="anchor" href="#arg-preference-index"></a></dt>
<dd><p>See <code><a href="resolve_vector_ranks.html">resolve_vector_ranks()</a></code> for more details.</p></dd>


<dt id="arg-collapse-string">collapse_string<a class="anchor" aria-label="anchor" href="#arg-collapse-string"></a></dt>
<dd><p>See <code><a href="resolve_vector_ranks.html">resolve_vector_ranks()</a></code> for more details.</p></dd>


<dt id="arg-replace-collapsed-rank-by-na">replace_collapsed_rank_by_NA<a class="anchor" aria-label="anchor" href="#arg-replace-collapsed-rank-by-na"></a></dt>
<dd><p>(Logical, default TRUE) See <code><a href="resolve_vector_ranks.html">resolve_vector_ranks()</a></code> for more details.</p></dd>


<dt id="arg-simplify-taxo">simplify_taxo<a class="anchor" aria-label="anchor" href="#arg-simplify-taxo"></a></dt>
<dd><p>(logical default TRUE). Do we apply the
function <code><a href="simplify_taxo.html">simplify_taxo()</a></code> to the phyloseq object?</p></dd>


<dt id="arg-keep-blast-metrics">keep_blast_metrics<a class="anchor" aria-label="anchor" href="#arg-keep-blast-metrics"></a></dt>
<dd><p>(Logical, default FALSE). If TRUE, the blast metrics
("Query seq. length", "Taxa seq. length", "Alignment length",  "% id. match", "e-value",
"bit score" and "Query cover") are stored in the tax_table.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other arguments passed on to <code><a href="blast_pq.html">blast_pq()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>

<ul><li><p>If behavior == "return_matrix" :</p><ul><li><p>If method = "top-hit" a matrix of taxonomic assignation</p></li>
<li><p>If method = "vote", a list of two matrix, the first is the
raw taxonomic assignation (before vote). The second one is
the taxonomic assignation in which conflicts are resolved
using vote.</p></li>
</ul></li>
<li><p>If behavior == "vote", return a new phyloseq object</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">ref_fasta</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"mini_UNITE_fungi.fasta.gz"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"MiscMetabar"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># assign_blastn(data_fungi_mini, ref_fasta = ref_fasta) # error because not</span></span></span>
<span class="r-in"><span><span class="co"># enough sequences in db so none blast query passed the filters.</span></span></span>
<span class="r-in"><span><span class="co"># So we used low score filter hereafter.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">assign_blastn</span><span class="op">(</span><span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>  ref_fasta <span class="op">=</span> <span class="va">ref_fasta</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"top-hit"</span>, min_id <span class="op">=</span> <span class="fl">70</span>, min_e_value <span class="op">=</span> <span class="fl">1e-3</span>, min_cover <span class="op">=</span> <span class="fl">50</span>,</span></span>
<span class="r-in"><span>  min_bit_score <span class="op">=</span> <span class="fl">20</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">assign_blastn</span><span class="op">(</span><span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>  ref_fasta <span class="op">=</span> <span class="va">ref_fasta</span>, method <span class="op">=</span> <span class="st">"vote"</span>,</span></span>
<span class="r-in"><span>  vote_algorithm <span class="op">=</span> <span class="st">"rel_majority"</span>, min_id <span class="op">=</span> <span class="fl">90</span>, min_cover <span class="op">=</span> <span class="fl">50</span>,</span></span>
<span class="r-in"><span>  behavior <span class="op">=</span> <span class="st">"add_to_phyloseq"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">@</span><span class="va">tax_table</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">assign_blastn</span><span class="op">(</span><span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>  ref_fasta <span class="op">=</span> <span class="va">ref_fasta</span>, method <span class="op">=</span> <span class="st">"vote"</span>,</span></span>
<span class="r-in"><span>  vote_algorithm <span class="op">=</span> <span class="st">"consensus"</span>, replace_collapsed_rank_by_NA <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  min_id <span class="op">=</span> <span class="fl">90</span>, min_cover <span class="op">=</span> <span class="fl">50</span>, behavior <span class="op">=</span> <span class="st">"add_to_phyloseq"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">@</span><span class="va">tax_table</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

