<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Build phylogenetic trees from refseq slot of a phyloseq object — build_phytree_pq • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Build phylogenetic trees from refseq slot of a phyloseq object — build_phytree_pq"><meta name="description" content="


This function build tree phylogenetic tree and if nb_bootstrap is
set, it build also the 3 corresponding bootstrapped tree.
Default parameters are based on doi:10.12688/f1000research.8986.2

and phangorn vignette Estimating phylogenetic trees with phangorn. You should understand your data, especially the markers,
before using this function.
Note that phylogenetic reconstruction with markers used for metabarcoding are
not robust. You must verify the robustness of your phylogenetic tree using
taxonomic classification (see vignette Tree visualization) and bootstrap or multi-tree visualization"><meta property="og:description" content="


This function build tree phylogenetic tree and if nb_bootstrap is
set, it build also the 3 corresponding bootstrapped tree.
Default parameters are based on doi:10.12688/f1000research.8986.2

and phangorn vignette Estimating phylogenetic trees with phangorn. You should understand your data, especially the markers,
before using this function.
Note that phylogenetic reconstruction with markers used for metabarcoding are
not robust. You must verify the robustness of your phylogenetic tree using
taxonomic classification (see vignette Tree visualization) and bootstrap or multi-tree visualization"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.9.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Build phylogenetic trees from refseq slot of a phyloseq object</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/dada_phyloseq.R" class="external-link"><code>R/dada_phyloseq.R</code></a></small>
      <div class="d-none name"><code>build_phytree_pq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>This function build tree phylogenetic tree and if nb_bootstrap is
set, it build also the 3 corresponding bootstrapped tree.</p>
<p>Default parameters are based on <a href="https://doi.org/10.12688/f1000research.8986.2" class="external-link">doi:10.12688/f1000research.8986.2</a>

and phangorn vignette <a href="https://klausvigo.github.io/phangorn/articles/Trees.html" class="external-link">Estimating phylogenetic trees with phangorn</a>. You should understand your data, especially the markers,
before using this function.</p>
<p>Note that phylogenetic reconstruction with markers used for metabarcoding are
not robust. You must verify the robustness of your phylogenetic tree using
taxonomic classification (see vignette <a href="https://adrientaudiere.github.io/MiscMetabar/articles/tree_visualization.html">Tree visualization</a>) and bootstrap or multi-tree visualization</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">build_phytree_pq</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  nb_bootstrap <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"GTR"</span>,</span>
<span>  optInv <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  optGamma <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rearrangement <span class="op">=</span> <span class="st">"NNI"</span>,</span>
<span>  control <span class="op">=</span> <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/pml.control.html" class="external-link">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>  optNni <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  multicore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-nb-bootstrap">nb_bootstrap<a class="anchor" aria-label="anchor" href="#arg-nb-bootstrap"></a></dt>
<dd><p>(default 0): If a positive number is set,
the function also build 3 bootstrapped trees using <code>nb_bootstrap</code>
bootstrap samples</p></dd>


<dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>allows to choose an amino acid models or nucleotide model,
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt id="arg-optinv">optInv<a class="anchor" aria-label="anchor" href="#arg-optinv"></a></dt>
<dd><p>Logical value indicating whether topology gets optimized
(NNI). See <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt id="arg-optgamma">optGamma<a class="anchor" aria-label="anchor" href="#arg-optgamma"></a></dt>
<dd><p>Logical value indicating whether gamma rate parameter gets
optimized. See <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt id="arg-rearrangement">rearrangement<a class="anchor" aria-label="anchor" href="#arg-rearrangement"></a></dt>
<dd><p>type of tree tree rearrangements to perform, one of
"NNI", "stochastic" or "ratchet"
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt id="arg-control">control<a class="anchor" aria-label="anchor" href="#arg-control"></a></dt>
<dd><p>A list of parameters for controlling the fitting process.
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt id="arg-optnni">optNni<a class="anchor" aria-label="anchor" href="#arg-optnni"></a></dt>
<dd><p>Logical value indicating whether topology gets optimized (NNI).
see <code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> for more details</p></dd>


<dt id="arg-multicore">multicore<a class="anchor" aria-label="anchor" href="#arg-multicore"></a></dt>
<dd><p>(logical) whether models should estimated in parallel.
see <code><a href="https://klausvigo.github.io/phangorn/reference/bootstrap.pml.html" class="external-link">phangorn::bootstrap.pml()</a></code> for more details</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other params for be passed on to
<code><a href="https://klausvigo.github.io/phangorn/reference/pml.html" class="external-link">phangorn::optim.pml()</a></code> function</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of phylogenetic tree</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is mainly a wrapper of the work of others.
Please make a reference to <code>phangorn</code> package if you
use this function.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"phangorn"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="subset_taxa_pq.html">subset_taxa_pq</a></span><span class="op">(</span><span class="va">data_fungi_mini</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="va">data_fungi_mini</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">9000</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">df_tree</span> <span class="op">&lt;-</span> <span class="fu">build_phytree_pq</span><span class="op">(</span><span class="va">df</span>, nb_bootstrap <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">UPGMA</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">UPGMA</span>, <span class="va">df_tree</span><span class="op">$</span><span class="va">UPGMA_bs</span>, main <span class="op">=</span> <span class="st">"UPGMA"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">NJ</span>, <span class="st">"unrooted"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">$</span><span class="va">tree</span>, <span class="va">df_tree</span><span class="op">$</span><span class="va">ML_bs</span>, p <span class="op">=</span> <span class="fl">20</span>, frame <span class="op">=</span> <span class="st">"circle"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/plotBS.html" class="external-link">plotBS</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">$</span><span class="va">tree</span>,</span></span>
<span class="r-in"><span>    <span class="va">df_tree</span><span class="op">$</span><span class="va">ML_bs</span>,</span></span>
<span class="r-in"><span>    p <span class="op">=</span> <span class="fl">20</span>,</span></span>
<span class="r-in"><span>    frame <span class="op">=</span> <span class="st">"circle"</span>,</span></span>
<span class="r-in"><span>    method <span class="op">=</span> <span class="st">"TBE"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/consensusNet.html" class="external-link">consensusNet</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">ML_bs</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">phangorn</span><span class="fu">::</span><span class="fu"><a href="https://klausvigo.github.io/phangorn/reference/consensusNet.html" class="external-link">consensusNet</a></span><span class="op">(</span><span class="va">df_tree</span><span class="op">$</span><span class="va">NJ_bs</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">ps_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/merge_phyloseq.html" class="external-link">merge_phyloseq</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">df_tree</span><span class="op">$</span><span class="va">ML</span><span class="op">$</span><span class="va">tree</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 6 sample(s) ( AD26-005-H_S10_MERGED.fastq.gz / CB8-019-H_S70_MERGED.fastq.gz / DY5-004-H_S97_MERGED.fastq.gz / N23-002-B_S130_MERGED.fastq.gz / NVABM0244-M_S137_MERGED.fastq.gz / T28-ABM602-B_S162_MERGED.fastq.gz ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 45</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 23</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 22</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on shared 8-mers:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Aligning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.14 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iteration 1 of 2:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reclustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Realigning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.09 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iteration 2 of 2:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reclustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Realigning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.07 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Refining the alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4411.398 --&gt; -4275.686 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4275.686 --&gt; -4275.67 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -4275.67 --&gt; -4255.759  NNI moves:  4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4255.759 --&gt; -4255.758 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -4255.758 --&gt; -4255.758  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4255.758 --&gt; -4255.758 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4383.247 --&gt; -4224.549 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4224.549 --&gt; -4224.543 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -4224.543 --&gt; -4221.228  NNI moves:  3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4221.228 --&gt; -4221.176 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -4221.176 --&gt; -4221.176  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -4221.176 --&gt; -4221.176 </span>
<span class="r-plt img"><img src="build_phytree_pq-1.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="build_phytree_pq-2.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="build_phytree_pq-3.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="build_phytree_pq-4.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="build_phytree_pq-5.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="build_phytree_pq-6.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="build_phytree_pq-7.png" alt="" width="700" height="433"></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘RNeXML’ ‘tidytree’</span>
<span class="r-plt img"><img src="build_phytree_pq-8.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

