<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>A wrapper of the function dada2::filterAndTrim() to use in targets pipeline — filter_trim • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="A wrapper of the function dada2::filterAndTrim() to use in targets pipeline — filter_trim"><meta name="description" content="


This function filter and trim (with parameters passed on to
dada2::filterAndTrim() function) forward sequences or paired end
sequence if 'rev' parameter is set. It return the list of files to
subsequent analysis in a targets pipeline."><meta property="og:description" content="


This function filter and trim (with parameters passed on to
dada2::filterAndTrim() function) forward sequences or paired end
sequence if 'rev' parameter is set. It return the list of files to
subsequent analysis in a targets pipeline."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>A wrapper of the function <code><a href="https://rdrr.io/pkg/dada2/man/filterAndTrim.html" class="external-link">dada2::filterAndTrim()</a></code> to use in <a href="https://books.ropensci.org/targets/" class="external-link">targets</a> pipeline</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/targets_misc.R" class="external-link"><code>R/targets_misc.R</code></a></small>
      <div class="d-none name"><code>filter_trim.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-maturing-blue" alt="lifecycle-maturing"></a></p>

<p>This function filter and trim (with parameters passed on to
<code><a href="https://rdrr.io/pkg/dada2/man/filterAndTrim.html" class="external-link">dada2::filterAndTrim()</a></code> function) forward sequences or paired end
sequence if 'rev' parameter is set. It return the list of files to
subsequent analysis in a targets pipeline.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filter_trim</span><span class="op">(</span></span>
<span>  fw <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rev <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_fw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"/output/filterAndTrim_fwd"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  output_rev <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"/output/filterAndTrim_rev"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  return_a_vector <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fw">fw<a class="anchor" aria-label="anchor" href="#arg-fw"></a></dt>
<dd><p>(required) a list of forward fastq files</p></dd>


<dt id="arg-rev">rev<a class="anchor" aria-label="anchor" href="#arg-rev"></a></dt>
<dd><p>a list of reverse fastq files for paired end trimming</p></dd>


<dt id="arg-output-fw">output_fw<a class="anchor" aria-label="anchor" href="#arg-output-fw"></a></dt>
<dd><p>Path to output folder for forward files. By default,
this function will create
a folder "output/filterAndTrim_fwd" in the current working directory.</p></dd>


<dt id="arg-output-rev">output_rev<a class="anchor" aria-label="anchor" href="#arg-output-rev"></a></dt>
<dd><p>Path to output folder for reverse files. By default,
this function will create
a folder "output/filterAndTrim_fwd" in the current working directory.</p></dd>


<dt id="arg-return-a-vector">return_a_vector<a class="anchor" aria-label="anchor" href="#arg-return-a-vector"></a></dt>
<dd><p>(logical, default FALSE) If true, the return is
a vector of path (usefull when used with
targets::tar_targets(..., format="file"))</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other parameters passed on to <code><a href="https://rdrr.io/pkg/dada2/man/filterAndTrim.html" class="external-link">dada2::filterAndTrim()</a></code> function.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of files. If rev is set, will return a list of two lists.
The first list is a list of forward files, and the second one
is a list of reverse files.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/dada2/man/filterAndTrim.html" class="external-link">dada2::filterAndTrim()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">testFastqs_fw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sam1F.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"dada2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sam2F.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"dada2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">testFastqs_rev</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sam1R.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"dada2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sam2R.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"dada2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">filt_fastq_fw</span> <span class="op">&lt;-</span> <span class="fu">filter_trim</span><span class="op">(</span><span class="va">testFastqs_fw</span>, output_fw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">derep_fw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dada2/man/derepFastq.html" class="external-link">derepFastq</a></span><span class="op">(</span><span class="va">filt_fastq_fw</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">derep_fw</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam1F.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 896 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  896 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=36, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  4 155 627 265 5 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam2F.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 909 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  909 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=36, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  890 14 2 246 797 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">filt_fastq_pe</span> <span class="op">&lt;-</span> <span class="fu">filter_trim</span><span class="op">(</span><span class="va">testFastqs_fw</span>,</span></span>
<span class="r-in"><span>  <span class="va">testFastqs_rev</span>,</span></span>
<span class="r-in"><span>  output_fw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="st">"fw"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  output_rev <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="st">"rev"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">derep_fw_pe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dada2/man/derepFastq.html" class="external-link">derepFastq</a></span><span class="op">(</span><span class="va">filt_fastq_pe</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">derep_rv_pe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dada2/man/derepFastq.html" class="external-link">derepFastq</a></span><span class="op">(</span><span class="va">filt_fastq_pe</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">derep_fw_pe</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam1F.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 896 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  896 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=36, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  4 155 627 265 5 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam1R.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 1373 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  1373 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=34, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  2 1029 664 983 156 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam2F.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 909 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  909 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=36, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  890 14 2 246 797 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam2R.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 1401 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  1401 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=34, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  1353 1171 1059 947 293 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">derep_rv_pe</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam1F.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 896 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  896 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=36, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  4 155 627 265 5 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam1R.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 1373 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  1373 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=34, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  2 1029 664 983 156 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam2F.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 909 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  909 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=36, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  890 14 2 246 797 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sam2R.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> derep-class: R object describing dereplicated sequencing reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $uniques: 1500 reads in 1401 unique sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sequence lengths: min=250, median=250, max=250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $quals: Quality matrix dimension:  1401 250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Consensus quality scores: min=7, median=34, max=38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $map: Map from reads to unique sequences:  1353 1171 1059 947 293 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

