<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Filter taxa by cleaning taxa with NA at given taxonomic rank(s) — filt_taxa_wo_NA • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Filter taxa by cleaning taxa with NA at given taxonomic rank(s) — filt_taxa_wo_NA"><meta name="description" content="


Basically a wrapper of subset_taxa_pq()"><meta property="og:description" content="


Basically a wrapper of subset_taxa_pq()"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Filter taxa by cleaning taxa with NA at given taxonomic rank(s)</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/dada_phyloseq.R" class="external-link"><code>R/dada_phyloseq.R</code></a></small>
      <div class="d-none name"><code>filt_taxa_wo_NA.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>Basically a wrapper of subset_taxa_pq()</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filt_taxa_wo_NA</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  taxa_ranks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_NA <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  NA_equivalent <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clean_pq <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-taxa-ranks">taxa_ranks<a class="anchor" aria-label="anchor" href="#arg-taxa-ranks"></a></dt>
<dd><p>A vector of taxonomic ranks. For examples c("Family","Genus").
If taxa_ranks is NULL (default), all ranks are used, i.e. all taxa with at least 1 NA
will be filtered out. Numeric position of taxonomic ranks can also be used.</p></dd>


<dt id="arg-n-na">n_NA<a class="anchor" aria-label="anchor" href="#arg-n-na"></a></dt>
<dd><p>(int default = 0). Number of allowed NA by taxa in the list of the
taxonomic ranks</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(logical). If TRUE, print additional information.</p></dd>


<dt id="arg-na-equivalent">NA_equivalent<a class="anchor" aria-label="anchor" href="#arg-na-equivalent"></a></dt>
<dd><p>(vector of character, default NULL). Exact matching of
the character listed in the vector are converted as NA before to filter out
taxa.</p></dd>


<dt id="arg-clean-pq">clean_pq<a class="anchor" aria-label="anchor" href="#arg-clean-pq"></a></dt>
<dd><p>(logical, default TRUE)
If set to TRUE, empty samples are discarded after filtering. See <code><a href="clean_pq.html">clean_pq()</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object of class phyloseq</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="subset_taxa_pq.html">subset_taxa_pq()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data_fungi_wo_NA</span> <span class="op">&lt;-</span> <span class="fu">filt_taxa_wo_NA</span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 1 sample(s) ( N23-002-M_S132_MERGED.fastq.gz ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 771</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 649</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 184</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You filtered out 771 taxa, leading to a phyloseq object including 649 taxa without NA in the taxonomic ranks: 1 2 3 4 5 6 7 8 9 10 11 12.</span>
<span class="r-in"><span><span class="fu">filt_taxa_wo_NA</span><span class="op">(</span><span class="va">data_fungi</span>, n_NA <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 0 sample(s) (  ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 410</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 1010</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 185</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You filtered out 410 taxa, leading to a phyloseq object including 1010 taxa without NA in the taxonomic ranks: 1 2 3 4 5 6 7 8 9 10 11 12.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 1010 taxa and 185 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 185 samples by 7 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 1010 taxa by 12 taxonomic ranks ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> refseq()      DNAStringSet:      [ 1010 reference sequences ]</span>
<span class="r-in"><span><span class="fu">filt_taxa_wo_NA</span><span class="op">(</span><span class="va">data_fungi</span>, taxa_ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 0 sample(s) (  ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 93</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 1327</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 185</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You filtered out 93 taxa, leading to a phyloseq object including 1327 taxa without NA in the taxonomic ranks: 1 2 3.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 1327 taxa and 185 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 185 samples by 7 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 1327 taxa by 12 taxonomic ranks ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> refseq()      DNAStringSet:      [ 1327 reference sequences ]</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">filt_taxa_wo_NA</span><span class="op">(</span><span class="va">data_fungi</span>, taxa_ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Trait"</span>, <span class="st">"Confidence.Ranking"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 0 sample(s) (  ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 185</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You filtered out 0 taxa, leading to a phyloseq object including 1420 taxa without NA in the taxonomic ranks: Trait Confidence.Ranking.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 1420 taxa and 185 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 185 samples by 7 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 1420 taxa by 12 taxonomic ranks ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> refseq()      DNAStringSet:      [ 1420 reference sequences ]</span>
<span class="r-in"><span><span class="fu">filt_taxa_wo_NA</span><span class="op">(</span><span class="va">data_fungi</span>,</span></span>
<span class="r-in"><span>  taxa_ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Trait"</span>, <span class="st">"Confidence.Ranking"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  NA_equivalent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">"NULL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 4 sample(s) ( N23-002-M_S132_MERGED.fastq.gz / O26-004-M_S150_MERGED.fastq.gz / Y28-002-H_S179_MERGED.fastq.gz / Y28-002-M_S180_MERGED.fastq.gz ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 1198</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 222</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 181</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You filtered out 1198 taxa, leading to a phyloseq object including 222 taxa without NA in the taxonomic ranks: Trait Confidence.Ranking.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 222 taxa and 181 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 181 samples by 7 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 222 taxa by 12 taxonomic ranks ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> refseq()      DNAStringSet:      [ 222 reference sequences ]</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

