<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Alluvial plot for taxonomy and samples factor vizualisation — ggaluv_pq • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Alluvial plot for taxonomy and samples factor vizualisation — ggaluv_pq"><meta name="description" content="


Basically a wrapper of ggalluvial
package"><meta property="og:description" content="


Basically a wrapper of ggalluvial
package"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.13.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Alluvial plot for taxonomy and samples factor vizualisation</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/plot_functions.R" class="external-link"><code>R/plot_functions.R</code></a></small>
      <div class="d-none name"><code>ggaluv_pq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>Basically a wrapper of <a href="https://corybrunson.github.io/ggalluvial/index.html" class="external-link">ggalluvial</a>
package</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ggaluv_pq</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  taxa_ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>  wrap_factor <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  by_sample <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rarefy_by_sample <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fact <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"nb_seq"</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>  min.size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  na_remove <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_ggfittext <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_geom_label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  size_lab <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-taxa-ranks">taxa_ranks<a class="anchor" aria-label="anchor" href="#arg-taxa-ranks"></a></dt>
<dd><p>A vector of taxonomic ranks. For examples c("Family","Genus").
If taxa ranks is not set
(default value = c("Phylum", "Class", "Order", "Family")).</p></dd>


<dt id="arg-wrap-factor">wrap_factor<a class="anchor" aria-label="anchor" href="#arg-wrap-factor"></a></dt>
<dd><p>A name to determine
which samples to merge using <code><a href="merge_samples2.html">merge_samples2()</a></code> function.
Need to be in <code>physeq@sam_data</code>.
Need to be use when you want to wrap by factor the final plot
with the number of taxa (type="nb_taxa")</p></dd>


<dt id="arg-by-sample">by_sample<a class="anchor" aria-label="anchor" href="#arg-by-sample"></a></dt>
<dd><p>(logical) If FALSE (default), sample information is not taking
into account, so the taxonomy is studied globally. If fact is not NULL, by_sample
is automatically set to TRUE.</p></dd>


<dt id="arg-rarefy-by-sample">rarefy_by_sample<a class="anchor" aria-label="anchor" href="#arg-rarefy-by-sample"></a></dt>
<dd><p>(logical, default FALSE) If TRUE, rarefy
samples using <code><a href="https://rdrr.io/pkg/phyloseq/man/rarefy_even_depth.html" class="external-link">phyloseq::rarefy_even_depth()</a></code> function.</p></dd>


<dt id="arg-fact">fact<a class="anchor" aria-label="anchor" href="#arg-fact"></a></dt>
<dd><p>(required) Name of the factor in <code>physeq@sam_data</code> used to plot  the last column</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>If "nb_seq" (default), the number of sequences is
used in plot. If "nb_taxa", the number of ASV is plotted.</p></dd>


<dt id="arg-width">width<a class="anchor" aria-label="anchor" href="#arg-width"></a></dt>
<dd><p>(passed on to <code><a href="http://corybrunson.github.io/ggalluvial/reference/geom_flow.html" class="external-link">ggalluvial::geom_flow()</a></code>) the width of each stratum,
as a proportion of the distance between axes. Defaults to 1/3.</p></dd>


<dt id="arg-min-size">min.size<a class="anchor" aria-label="anchor" href="#arg-min-size"></a></dt>
<dd><p>(passed on to <code><a href="https://wilkox.org/ggfittext/reference/geom_fit_text.html" class="external-link">ggfittext::geom_fit_text()</a></code>) Minimum font size,
in points. Text that would need to be shrunk below this size to fit the box will
be hidden. Defaults to 4 pt.</p></dd>


<dt id="arg-na-remove">na_remove<a class="anchor" aria-label="anchor" href="#arg-na-remove"></a></dt>
<dd><p>(logical, default FALSE) If set to TRUE, remove samples with
NA in the variables set in formula.</p></dd>


<dt id="arg-use-ggfittext">use_ggfittext<a class="anchor" aria-label="anchor" href="#arg-use-ggfittext"></a></dt>
<dd><p>(logical, default FALSE) Do we use ggfittext to plot labels?</p></dd>


<dt id="arg-use-geom-label">use_geom_label<a class="anchor" aria-label="anchor" href="#arg-use-geom-label"></a></dt>
<dd><p>(logical, default FALSE) Do we use geom_label to plot labels?</p></dd>


<dt id="arg-size-lab">size_lab<a class="anchor" aria-label="anchor" href="#arg-size-lab"></a></dt>
<dd><p>Size for label if use_ggfittext is FALSE</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other arguments passed on to <code><a href="http://corybrunson.github.io/ggalluvial/reference/geom_flow.html" class="external-link">ggalluvial::geom_flow()</a></code> function.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is mainly a wrapper of the work of others.
Please make a reference to <code>ggalluvial</code> package if you
use this function.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="sankey_pq.html">sankey_pq()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ggalluvial"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">ggaluv_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required namespace: ggalluvial</span>
<span class="r-plt img"><img src="ggaluv_pq-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ggalluvial"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">ggaluv_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, type <span class="op">=</span> <span class="st">"nb_taxa"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">ggaluv_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, wrap_factor <span class="op">=</span> <span class="st">"Height"</span>, by_sample <span class="op">=</span> <span class="cn">TRUE</span>, type <span class="op">=</span> <span class="st">"nb_taxa"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="st">"Height"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">ggaluv_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>    width <span class="op">=</span> <span class="fl">0.9</span>, min.size <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>    type <span class="op">=</span> <span class="st">"nb_taxa"</span>, taxa_ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="va">rev</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>`group` has missing values; corresponding samples will be dropped</span>
<span class="r-plt img"><img src="ggaluv_pq-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

