<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Venn diagram of phyloseq-class object using ggVennDiagram::ggVennDiagram function — ggvenn_pq • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Venn diagram of phyloseq-class object using ggVennDiagram::ggVennDiagram function — ggvenn_pq"><meta name="description" content='


Note that you can use ggplot2 function to customize the plot
for ex. + scale_fill_distiller(palette = "BuPu", direction = 1)
and + scale_x_continuous(expand = expansion(mult = 0.5)). See
examples.'><meta property="og:description" content='


Note that you can use ggplot2 function to customize the plot
for ex. + scale_fill_distiller(palette = "BuPu", direction = 1)
and + scale_x_continuous(expand = expansion(mult = 0.5)). See
examples.'></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Venn diagram of <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object using <code>ggVennDiagram::ggVennDiagram</code> function</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/plot_functions.R" class="external-link"><code>R/plot_functions.R</code></a></small>
      <div class="d-none name"><code>ggvenn_pq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-maturing-blue" alt="lifecycle-maturing"></a></p>

<p>Note that you can use ggplot2 function to customize the plot
for ex. <code>+ scale_fill_distiller(palette = "BuPu", direction = 1)</code>
and <code>+ scale_x_continuous(expand = expansion(mult = 0.5))</code>. See
examples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ggvenn_pq</span><span class="op">(</span></span>
<span>  physeq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fact <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_nb_seq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  taxonomic_rank <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  split_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_nb_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_nb_seq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rarefy_before_merging <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rarefy_after_merging <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  return_data_for_venn <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"nb_taxa"</span>,</span>
<span>  na_remove <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-fact">fact<a class="anchor" aria-label="anchor" href="#arg-fact"></a></dt>
<dd><p>(required): Name of the factor to cluster samples by modalities.
Need to be in <code>physeq@sam_data</code>.</p></dd>


<dt id="arg-min-nb-seq">min_nb_seq<a class="anchor" aria-label="anchor" href="#arg-min-nb-seq"></a></dt>
<dd><p>minimum number of sequences by OTUs by
samples to take into count this OTUs in this sample. For example,
if min_nb_seq=2,each value of 2 or less in the OTU table
will not count in the venn diagram</p></dd>


<dt id="arg-taxonomic-rank">taxonomic_rank<a class="anchor" aria-label="anchor" href="#arg-taxonomic-rank"></a></dt>
<dd><p>Name (or number) of a taxonomic rank
to count. If set to Null (the default) the number of OTUs is counted.</p></dd>


<dt id="arg-split-by">split_by<a class="anchor" aria-label="anchor" href="#arg-split-by"></a></dt>
<dd><p>Split into multiple plot using variable split_by.
The name of a variable must be present in <code>sam_data</code> slot
of the physeq object.</p></dd>


<dt id="arg-add-nb-samples">add_nb_samples<a class="anchor" aria-label="anchor" href="#arg-add-nb-samples"></a></dt>
<dd><p>(logical, default TRUE) Add the number of samples to
levels names</p></dd>


<dt id="arg-add-nb-seq">add_nb_seq<a class="anchor" aria-label="anchor" href="#arg-add-nb-seq"></a></dt>
<dd><p>(logical, default FALSE) Add the number of sequences to
levels names</p></dd>


<dt id="arg-rarefy-before-merging">rarefy_before_merging<a class="anchor" aria-label="anchor" href="#arg-rarefy-before-merging"></a></dt>
<dd><p>Rarefy each sample before merging by the
modalities of args <code>fact</code>. Use <code><a href="https://rdrr.io/pkg/phyloseq/man/rarefy_even_depth.html" class="external-link">phyloseq::rarefy_even_depth()</a></code> function</p></dd>


<dt id="arg-rarefy-after-merging">rarefy_after_merging<a class="anchor" aria-label="anchor" href="#arg-rarefy-after-merging"></a></dt>
<dd><p>Rarefy each sample after merging by the
modalities of args <code>fact</code>.</p></dd>


<dt id="arg-return-data-for-venn">return_data_for_venn<a class="anchor" aria-label="anchor" href="#arg-return-data-for-venn"></a></dt>
<dd><p>(logical, default FALSE) If TRUE, the plot is
not returned, but the resulting dataframe to plot with ggVennDiagram package
is returned.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(logical, default TRUE) If TRUE, prompt some messages.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>If "nb_taxa" (default), the number of taxa (ASV, OTU or
taxonomic_rank if <code>taxonomic_rank</code> is not NULL) is
used in plot. If "nb_seq", the number of sequences is plotted.
<code>taxonomic_rank</code> is never used if type = "nb_seq".</p></dd>


<dt id="arg-na-remove">na_remove<a class="anchor" aria-label="anchor" href="#arg-na-remove"></a></dt>
<dd><p>(logical, default TRUE) If set to TRUE, remove samples with
NA in the variables set in <code>fact</code> param</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other arguments for the <code><a href="https://gaospecial.github.io/ggVennDiagram/reference/ggVennDiagram.html" class="external-link">ggVennDiagram::ggVennDiagram</a></code> function
for ex. <code>category.names</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></code>2 plot representing Venn diagram of
modalities of the argument <code>factor</code> or if split_by is set a list
of plots.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="upset_pq.html">upset_pq()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ggVennDiagram"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 54 were discarded due to NA in variable fact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 501 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 457 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 469 taxa and 0 samples.</span>
<span class="r-plt img"><img src="ggvenn_pq-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ggVennDiagram"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"BuPu"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">pl</span> <span class="op">&lt;-</span> <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span>, split_by <span class="op">=</span> <span class="st">"Time"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">pl</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">p</span> <span class="op">&lt;-</span> <span class="va">pl</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span></span></span>
<span class="r-in"><span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"BuPu"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">22</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="va">data_fungi2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="va">data_fungi</span><span class="op">@</span><span class="va">sam_data</span><span class="op">$</span><span class="va">Tree_name</span> <span class="op">==</span> <span class="st">"A10-005"</span> <span class="op">|</span></span></span>
<span class="r-in"><span>    <span class="va">data_fungi</span><span class="op">@</span><span class="va">sam_data</span><span class="op">$</span><span class="va">Height</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Low"</span>, <span class="st">"High"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi2</span>, fact <span class="op">=</span> <span class="st">"Height"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi2</span>, fact <span class="op">=</span> <span class="st">"Height"</span>, type <span class="op">=</span> <span class="st">"nb_seq"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span>, add_nb_seq <span class="op">=</span> <span class="cn">TRUE</span>, set_size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span>, rarefy_before_merging <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span>, rarefy_after_merging <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># For more flexibility, you can save the dataset for more precise construction</span></span></span>
<span class="r-in"><span>  <span class="co"># with ggplot2 and ggVennDiagramm</span></span></span>
<span class="r-in"><span>  <span class="co"># (https://gaospecial.github.io/ggVennDiagram/articles/fully-customed.html)</span></span></span>
<span class="r-in"><span>  <span class="va">res_venn</span> <span class="op">&lt;-</span> <span class="fu">ggvenn_pq</span><span class="op">(</span><span class="va">data_fungi</span>, fact <span class="op">=</span> <span class="st">"Height"</span>, return_data_for_venn <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="co"># 1. region count layer</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_polygon.html" class="external-link">geom_polygon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, group <span class="op">=</span> <span class="va">id</span>, fill <span class="op">=</span> <span class="va">name</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      data <span class="op">=</span> <span class="fu">ggVennDiagram</span><span class="fu">::</span><span class="fu"><a href="https://gaospecial.github.io/ggVennDiagram/reference/venn_plot_data.html" class="external-link">venn_regionedge</a></span><span class="op">(</span><span class="va">res_venn</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="funky_color.html">funky_color</a></span><span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="co"># 2. set edge layer</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_path</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, color <span class="op">=</span> <span class="va">id</span>, group <span class="op">=</span> <span class="va">id</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      data <span class="op">=</span> <span class="fu">ggVennDiagram</span><span class="fu">::</span><span class="fu"><a href="https://gaospecial.github.io/ggVennDiagram/reference/venn_plot_data.html" class="external-link">venn_setedge</a></span><span class="op">(</span><span class="va">res_venn</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      show.legend <span class="op">=</span> <span class="cn">FALSE</span>, linewidth <span class="op">=</span> <span class="fl">2</span></span></span>
<span class="r-in"><span>    <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="co"># 3. set label layer</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, label <span class="op">=</span> <span class="va">name</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      data <span class="op">=</span> <span class="fu">ggVennDiagram</span><span class="fu">::</span><span class="fu"><a href="https://gaospecial.github.io/ggVennDiagram/reference/venn_plot_data.html" class="external-link">venn_setlabel</a></span><span class="op">(</span><span class="va">res_venn</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="co"># 4. region label layer</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        <span class="va">count</span>, <span class="st">" ("</span>,</span></span>
<span class="r-in"><span>        <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">(</span><span class="va">count</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>, accuracy <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, <span class="st">")"</span></span></span>
<span class="r-in"><span>      <span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      data <span class="op">=</span> <span class="fu">ggVennDiagram</span><span class="fu">::</span><span class="fu"><a href="https://gaospecial.github.io/ggVennDiagram/reference/venn_plot_data.html" class="external-link">venn_regionlabel</a></span><span class="op">(</span><span class="va">res_venn</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 54 were discarded due to NA in variable fact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 501 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 457 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 469 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 54 were discarded due to NA in variable fact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 501 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 457 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 469 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 335 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 276 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 299 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 360 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 404 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 374 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 341 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 221 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 360 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 343 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 218 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 408 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Two modalities differ greatly (more than x2) in their number of sequences (159635 vs 53355). You may be interested by the parameter rarefy_after_merging</span>
<span class="r-plt img"><img src="ggvenn_pq-2.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="ggvenn_pq-3.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="ggvenn_pq-4.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="ggvenn_pq-5.png" alt="" width="700" height="433"></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 501 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 457 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 1366 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Two modalities differ greatly (more than x2) in their number of sequences (432919 vs 3961). You may be interested by the parameter rarefy_after_merging</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 501 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 457 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 1366 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Two modalities differ greatly (more than x2) in their number of sequences (432919 vs 3961). You may be interested by the parameter rarefy_after_merging</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 54 were discarded due to NA in variable fact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 501 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 457 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 469 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 54 were discarded due to NA in variable fact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You set `rngseed` to FALSE. Make sure you've set &amp; recorded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  the random seed of your session for reproducibility.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> See `?set.seed`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1130OTUs were removed because they are no longer </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> present in any sample after random subsampling</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 173 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 157 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 156 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 54 were discarded due to NA in variable fact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You set `rngseed` to FALSE. Make sure you've set &amp; recorded</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  the random seed of your session for reproducibility.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> See `?set.seed`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 157OTUs were removed because they are no longer </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> present in any sample after random subsampling</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 386 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 342 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 338 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 54 were discarded due to NA in variable fact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 501 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 457 taxa and 0 samples.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 469 taxa and 0 samples.</span>
<span class="r-plt img"><img src="ggvenn_pq-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

