<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.9.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6>
    <a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="describe-phyloseq-object">Describe phyloseq object<a class="anchor" aria-label="anchor" href="#describe-phyloseq-object"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="see-also-vignette-on-exploring-data-articles-explore-data-html-">See also vignette on <a href="articles/explore_data.html">exploring data</a><a class="anchor" aria-label="anchor" href="#see-also-vignette-on-exploring-data-articles-explore-data-html-"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="summary_plot_pq.html">summary_plot_pq()</a></code> 
        </dt>
        <dd>Summarize a <code>phyloseq-class</code> object using a plot.</dd>
      </dl><dl><dt>
          
          <code><a href="tbl_sum_samdata.html">tbl_sum_samdata()</a></code> 
        </dt>
        <dd>Summarize information from sample data in a table</dd>
      </dl><dl><dt>
          
          <code><a href="verify_pq.html">verify_pq()</a></code> 
        </dt>
        <dd>Verify the validity of a phyloseq object</dd>
      </dl></div><div class="section level2">
      <h2 id="transform-phyloseq-object">Transform phyloseq object<a class="anchor" aria-label="anchor" href="#transform-phyloseq-object"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="taxonomy">Taxonomy<a class="anchor" aria-label="anchor" href="#taxonomy"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="add_new_taxonomy_pq.html">add_new_taxonomy_pq()</a></code> 
        </dt>
        <dd>Add new taxonomic rank to a phyloseq object.</dd>
      </dl><dl><dt>
          
          <code><a href="add_blast_info.html">add_blast_info()</a></code> 
        </dt>
        <dd>Add information from <code><a href="../reference/blast_pq.html">blast_pq()</a></code> to the <code>tax_table</code> slot of a <em>phyloseq</em> object</dd>
      </dl><dl><dt>
          
          <code><a href="blast_to_phyloseq.html">blast_to_phyloseq()</a></code> 
        </dt>
        <dd>Blast some sequence against <code>refseq</code> slot of a <code>phyloseq-class</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="blast_pq.html">blast_pq()</a></code> 
        </dt>
        <dd>Blast all sequence of <code>refseq</code> slot of a <code>phyloseq-class</code> object against a custom database.</dd>
      </dl><dl><dt>
          
          <code><a href="blast_to_derep.html">blast_to_derep()</a></code> 
        </dt>
        <dd>Blast some sequence against sequences from of a <code>derep-class</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="filter_asv_blast.html">filter_asv_blast()</a></code> 
        </dt>
        <dd>Filter undesirable taxa using blast against a custom database.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="function-to-add-information-from-taxonomy">Function to add information from taxonomy<a class="anchor" aria-label="anchor" href="#function-to-add-information-from-taxonomy"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="add_funguild_info.html">add_funguild_info()</a></code> 
        </dt>
        <dd>Add information about Guild for FUNGI the FUNGuild databse</dd>
      </dl><dl><dt>
          
          <code><a href="funguild_assign.html">funguild_assign()</a></code> 
        </dt>
        <dd>Assign Guilds to Organisms Based on Taxonomic Classification</dd>
      </dl><dl><dt>
          
          <code><a href="get_funguild_db.html">get_funguild_db()</a></code> 
        </dt>
        <dd>Retrieve the FUNGuild database</dd>
      </dl></div><div class="section level2">
      
      <h3 id="transform-otu-table-only">Transform OTU table only<a class="anchor" aria-label="anchor" href="#transform-otu-table-only"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="as_binary_otu_table.html">as_binary_otu_table()</a></code> 
        </dt>
        <dd>Transform the otu_table of a <code>phyloseq-class</code> object into a <code>phyloseq-class</code> object with a binary otu_table.</dd>
      </dl><dl><dt>
          
          <code><a href="normalize_prop_pq.html">normalize_prop_pq()</a></code> 
        </dt>
        <dd>Normalize OTU table using samples depth</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_as_rows.html">taxa_as_rows()</a></code> 
        </dt>
        <dd>Force taxa to be in columns in the otu_table of a physeq object</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_as_columns.html">taxa_as_columns()</a></code> 
        </dt>
        <dd>Force taxa to be in columns in the otu_table of a physeq object</dd>
      </dl></div><div class="section level2">
      
      <h3 id="subset-merge-taxa">Subset/merge taxa<a class="anchor" aria-label="anchor" href="#subset-merge-taxa"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="subset_taxa_pq.html">subset_taxa_pq()</a></code> 
        </dt>
        <dd>Subset taxa using a conditional named boolean vector.</dd>
      </dl><dl><dt>
          
          <code><a href="subset_taxa_tax_control.html">subset_taxa_tax_control()</a></code> 
        </dt>
        <dd>Subset taxa using a taxa control (e.g. truffle root tips) through 3 methods.</dd>
      </dl><dl><dt>
          
          <code><a href="select_taxa-methods.html">select_taxa()</a></code> 
        </dt>
        <dd>Select a subset of taxa in a specified order where possible</dd>
      </dl><dl><dt>
          
          <code><a href="merge_taxa_vec.html">merge_taxa_vec()</a></code> 
        </dt>
        <dd>Merge taxa in groups (vectorized version)</dd>
      </dl></div><div class="section level2">
      
      <h3 id="subset-merge-samples">Subset/merge samples<a class="anchor" aria-label="anchor" href="#subset-merge-samples"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="merge_samples2.html">merge_samples2()</a></code> 
        </dt>
        <dd>Merge samples by a sample variable or factor</dd>
      </dl><dl><dt>
          
          <code><a href="rarefy_sample_count_by_modality.html">rarefy_sample_count_by_modality()</a></code> 
        </dt>
        <dd>Rarefy (equalize) the number of samples per modality of a factor</dd>
      </dl><dl><dt>
          
          <code><a href="select_one_sample.html">select_one_sample()</a></code> 
        </dt>
        <dd>Select one sample from a physeq object</dd>
      </dl><dl><dt>
          
          <code><a href="subset_samples_pq.html">subset_samples_pq()</a></code> 
        </dt>
        <dd>Subset samples using a conditional boolean vector.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="-re-clustering-otu-table">(Re)clustering OTU table<a class="anchor" aria-label="anchor" href="#-re-clustering-otu-table"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="asv2otu.html">asv2otu()</a></code> 
        </dt>
        <dd>Recluster sequences of an object of class <code>physeq</code> or a list of DNA sequences</dd>
      </dl><dl><dt>
          
          <code><a href="swarm_clustering.html">swarm_clustering()</a></code> 
        </dt>
        <dd>Re-cluster sequences of an object of class <code>physeq</code> or cluster a list of DNA sequences using SWARM</dd>
      </dl><dl><dt>
          
          <code><a href="vsearch_clustering.html">vsearch_clustering()</a></code> 
        </dt>
        <dd>Recluster sequences of an object of class <code>physeq</code> or cluster a list of DNA sequences using vsearch software</dd>
      </dl><dl><dt>
          
          <code><a href="lulu_pq.html">lulu_pq()</a></code> 
        </dt>
        <dd>Lulu reclustering of class <code>physeq</code></dd>
      </dl><dl><dt>
          
          <code><a href="lulu.html">lulu()</a></code> 
        </dt>
        <dd>Post Clustering Curation of Amplicon Data.</dd>
      </dl><dl><dt>
          
          <code><a href="mumu_pq.html">mumu_pq()</a></code> 
        </dt>
        <dd>MUMU reclustering of class <code>physeq</code></dd>
      </dl></div><div class="section level2">
      
      <h3 id="remove-chimera-using-vsearch">Remove chimera using vsearch<a class="anchor" aria-label="anchor" href="#remove-chimera-using-vsearch"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="chimera_removal_vs.html">chimera_removal_vs()</a></code> 
        </dt>
        <dd>Search for a list of sequence in an object to remove chimera taxa using <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a></dd>
      </dl><dl><dt>
          
          <code><a href="chimera_detection_vs.html">chimera_detection_vs()</a></code> 
        </dt>
        <dd>Detect for chimera taxa using <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a></dd>
      </dl></div><div class="section level2">
      
      <h3 id="sample-data">Sample data<a class="anchor" aria-label="anchor" href="#sample-data"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="add_info_to_sam_data.html">add_info_to_sam_data()</a></code> 
        </dt>
        <dd>Add information to sample_data slot of a phyloseq-class object</dd>
      </dl></div><div class="section level2">
      
      <h3 id="phylogenetic-tree">Phylogenetic tree<a class="anchor" aria-label="anchor" href="#phylogenetic-tree"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="build_phytree_pq.html">build_phytree_pq()</a></code> 
        </dt>
        <dd>Build phylogenetic trees from refseq slot of a phyloseq object</dd>
      </dl></div><div class="section level2">
      
      <h3 id="others">Others<a class="anchor" aria-label="anchor" href="#others"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="add_dna_to_phyloseq.html">add_dna_to_phyloseq()</a></code> 
        </dt>
        <dd>Add dna in <code>refseq</code> slot of a <code>physeq</code> object using taxa names and renames taxa using ASV_1, ASV_2, …</dd>
      </dl><dl><dt>
          
          <code><a href="clean_pq.html">clean_pq()</a></code> 
        </dt>
        <dd>Clean phyloseq object by removing empty samples and taxa</dd>
      </dl></div><div class="section level2">
      <h2 id="explore-phyloseq-object">Explore phyloseq object<a class="anchor" aria-label="anchor" href="#explore-phyloseq-object"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="alpha-diversity">alpha-diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="accu_plot.html">accu_plot()</a></code> 
        </dt>
        <dd>Plot accumulation curves for <code>phyloseq-class</code> object</dd>
      </dl><dl><dt>
          
          <code><a href="accu_plot_balanced_modality.html">accu_plot_balanced_modality()</a></code> 
        </dt>
        <dd>Plot accumulation curves with balanced modality and depth rarefaction</dd>
      </dl><dl><dt>
          
          <code><a href="accu_samp_threshold.html">accu_samp_threshold()</a></code> 
        </dt>
        <dd>Compute the number of sequence to obtain a given proportion of ASV in accumulation curves</dd>
      </dl><dl><dt>
          
          <code><a href="ggbetween_pq.html">ggbetween_pq()</a></code> 
        </dt>
        <dd>Box/Violin plots for between-subjects comparisons of Hill Number</dd>
      </dl><dl><dt>
          
          <code><a href="ggscatt_pq.html">ggscatt_pq()</a></code> 
        </dt>
        <dd>Scatterplot with marginal distributions and statistical results against Hill diversity of phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="glmutli_pq.html">glmutli_pq()</a></code> 
        </dt>
        <dd>Automated model selection and multimodel inference with (G)LMs for phyloseq</dd>
      </dl><dl><dt>
          
          <code><a href="hill_pq.html">hill_pq()</a></code> 
        </dt>
        <dd>Graphical representation of hill number 0, 1 and 2 across a factor</dd>
      </dl><dl><dt>
          
          <code><a href="hill_test_rarperm_pq.html">hill_test_rarperm_pq()</a></code> 
        </dt>
        <dd>Test multiple times effect of factor on Hill diversity with different rarefaction even depth</dd>
      </dl><dl><dt>
          
          <code><a href="hill_tuckey_pq.html">hill_tuckey_pq()</a></code> 
        </dt>
        <dd>Calculate hill number and compute Tuckey post-hoc test</dd>
      </dl><dl><dt>
          
          <code><a href="iNEXT_pq.html">iNEXT_pq()</a></code> 
        </dt>
        <dd>iNterpolation and EXTrapolation of Hill numbers (with iNEXT)</dd>
      </dl></div><div class="section level2">
      
      <h3 id="beta-diversity">beta-diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="adonis_pq.html">adonis_pq()</a></code> 
        </dt>
        <dd>Permanova on a phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="adonis_rarperm_pq.html">adonis_rarperm_pq()</a></code> 
        </dt>
        <dd>Permanova (adonis) on permutations of rarefaction even depth</dd>
      </dl><dl><dt>
          
          <code><a href="biplot_pq.html">biplot_pq()</a></code> 
        </dt>
        <dd>Visualization of two samples for comparison</dd>
      </dl><dl><dt>
          
          <code><a href="circle_pq.html">circle_pq()</a></code> 
        </dt>
        <dd>Plot OTU circle for <code>phyloseq-class</code> object</dd>
      </dl><dl><dt>
          
          <code><a href="compare_pairs_pq.html">compare_pairs_pq()</a></code> 
        </dt>
        <dd>Compare samples in pairs using diversity and number of ASV including shared ASV.</dd>
      </dl><dl><dt>
          
          <code><a href="formattable_pq.html">formattable_pq()</a></code> 
        </dt>
        <dd>Create an visualization table to describe taxa distribution across a modality</dd>
      </dl><dl><dt>
          
          <code><a href="ggvenn_pq.html">ggvenn_pq()</a></code> 
        </dt>
        <dd>Venn diagram of <code>phyloseq-class</code> object using <code><a href="https://gaospecial.github.io/ggVennDiagram/reference/ggVennDiagram.html" class="external-link">ggVennDiagram::ggVennDiagram</a></code> function</dd>
      </dl><dl><dt>
          
          <code><a href="graph_test_pq.html">graph_test_pq()</a></code> 
        </dt>
        <dd>Performs graph-based permutation tests on phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="LCBD_pq.html">LCBD_pq()</a></code> 
        </dt>
        <dd>Compute and test local contributions to beta diversity (LCBD) of samples</dd>
      </dl><dl><dt>
          
          <code><a href="multi_biplot_pq.html">multi_biplot_pq()</a></code> 
        </dt>
        <dd>Visualization of a collection of couples of samples for comparison</dd>
      </dl><dl><dt>
          
          <code><a href="plot_LCBD_pq.html">plot_LCBD_pq()</a></code> 
        </dt>
        <dd>Plot and test local contributions to beta diversity (LCBD) of samples</dd>
      </dl><dl><dt>
          
          <code><a href="plot_SCBD_pq.html">plot_SCBD_pq()</a></code> 
        </dt>
        <dd>Plot species contributions to beta diversity (SCBD) of samples</dd>
      </dl><dl><dt>
          
          <code><a href="plot_tsne_pq.html">plot_tsne_pq()</a></code> 
        </dt>
        <dd>Plot a tsne low dimensional representation of a phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="ridges_pq.html">ridges_pq()</a></code> 
        </dt>
        <dd>Ridge plot of a phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="SRS_curve_pq.html">SRS_curve_pq()</a></code> 
        </dt>
        <dd>Scaling with ranked subsampling (SRS) curve of phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="tsne_pq.html">tsne_pq()</a></code> 
        </dt>
        <dd>Compute tSNE position of samples from a phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="upset_pq.html">upset_pq()</a></code> 
        </dt>
        <dd>Make upset plot for phyloseq object.</dd>
      </dl><dl><dt>
          
          <code><a href="upset_test_pq.html">upset_test_pq()</a></code> 
        </dt>
        <dd>Test for differences between intersections</dd>
      </dl><dl><dt>
          
          <code><a href="var_par_pq.html">var_par_pq()</a></code> 
        </dt>
        <dd>Partition the Variation of a phyloseq object by 2, 3, or 4 Explanatory Matrices</dd>
      </dl><dl><dt>
          
          <code><a href="var_par_rarperm_pq.html">var_par_rarperm_pq()</a></code> 
        </dt>
        <dd>Partition the Variation of a phyloseq object with rarefaction permutations</dd>
      </dl><dl><dt>
          
          <code><a href="plot_var_part_pq.html">plot_var_part_pq()</a></code> 
        </dt>
        <dd>Plot the partition the variation of a phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="venn_pq.html">venn_pq()</a></code> 
        </dt>
        <dd>Venn diagram of <code>phyloseq-class</code> object</dd>
      </dl></div><div class="section level2">
      
      <h3 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="multipatt_pq.html">multipatt_pq()</a></code> 
        </dt>
        <dd>Test and plot multipatt result</dd>
      </dl><dl><dt>
          
          <code><a href="phyloseq_to_edgeR.html">phyloseq_to_edgeR()</a></code> 
        </dt>
        <dd>Convert phyloseq OTU count data into DGEList for edgeR package</dd>
      </dl><dl><dt>
          
          <code><a href="plot_edgeR_pq.html">plot_edgeR_pq()</a></code> 
        </dt>
        <dd>Plot edgeR results for a phyloseq or a edgeR object.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_deseq2_pq.html">plot_deseq2_pq()</a></code> 
        </dt>
        <dd>Plot DESeq2 results for a phyloseq or a DESeq2 object.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_mt.html">plot_mt()</a></code> 
        </dt>
        <dd>Plot the result of a mt test <code><a href="https://rdrr.io/pkg/phyloseq/man/mt-methods.html" class="external-link">phyloseq::mt()</a></code></dd>
      </dl><dl><dt>
          
          <code><a href="ancombc_pq.html">ancombc_pq()</a></code> 
        </dt>
        <dd>Run ANCOMBC2 on phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="signif_ancombc.html">signif_ancombc()</a></code> 
        </dt>
        <dd>Filter ancombc_pq results</dd>
      </dl><dl><dt>
          
          <code><a href="plot_ancombc_pq.html">plot_ancombc_pq()</a></code> 
        </dt>
        <dd>Plot ANCOMBC2 result for phyloseq object</dd>
      </dl></div><div class="section level2">
      
      <h3 id="taxonomy-1">Taxonomy<a class="anchor" aria-label="anchor" href="#taxonomy-1"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="ggaluv_pq.html">ggaluv_pq()</a></code> 
        </dt>
        <dd>Alluvial plot for taxonomy and samples factor vizualisation</dd>
      </dl><dl><dt>
          
          <code><a href="heat_tree_pq.html">heat_tree_pq()</a></code> 
        </dt>
        <dd>Heat tree from <code>metacoder</code> package using <code>tax_table</code> slot</dd>
      </dl><dl><dt>
          
          <code><a href="krona.html">krona()</a></code> 
        </dt>
        <dd>Make Krona files using <a href="https://github.com/marbl/Krona/wiki" class="external-link">KronaTools</a>.</dd>
      </dl><dl><dt>
          
          <code><a href="merge_krona.html">merge_krona()</a></code> 
        </dt>
        <dd>Merge Krona files using <a href="https://github.com/marbl/Krona/wiki" class="external-link">KronaTools</a>.</dd>
      </dl><dl><dt>
          
          <code><a href="multitax_bar_pq.html">multitax_bar_pq()</a></code> 
        </dt>
        <dd>Plot taxonomic distribution across 3 taxonomic levels and optionally one sample factor</dd>
      </dl><dl><dt>
          
          <code><a href="physeq_or_string_to_dna.html">physeq_or_string_to_dna()</a></code> 
        </dt>
        <dd>Return a DNAStringSet object from either a character vector of DNA sequences or the <code>refseq</code> slot of a phyloseq-class object</dd>
      </dl><dl><dt>
          
          <code><a href="plot_guild_pq.html">plot_guild_pq()</a></code> 
        </dt>
        <dd>Plot information about Guild from tax_table slot previously created with <code><a href="../reference/add_funguild_info.html">add_funguild_info()</a></code></dd>
      </dl><dl><dt>
          
          <code><a href="plot_tax_pq.html">plot_tax_pq()</a></code> 
        </dt>
        <dd>Plot taxonomic distribution in function of a factor with stacked bar in %</dd>
      </dl><dl><dt>
          
          <code><a href="rotl_pq.html">rotl_pq()</a></code> 
        </dt>
        <dd>rotl wrapper for phyloseq data</dd>
      </dl><dl><dt>
          
          <code><a href="sankey_pq.html">sankey_pq()</a></code> 
        </dt>
        <dd>Sankey plot of <code>phyloseq-class</code> object</dd>
      </dl><dl><dt>
          
          <code><a href="search_exact_seq_pq.html">search_exact_seq_pq()</a></code> 
        </dt>
        <dd>Search for exact matching of sequences using complement, reverse and reverse-complement</dd>
      </dl><dl><dt>
          
          <code><a href="tax_bar_pq.html">tax_bar_pq()</a></code> 
        </dt>
        <dd>Plot the distribution of sequences or ASV in one taxonomic levels</dd>
      </dl><dl><dt>
          
          <code><a href="tax_datatable.html">tax_datatable()</a></code> 
        </dt>
        <dd>Make a datatable with the taxonomy of a <code>phyloseq-class</code> object</dd>
      </dl><dl><dt>
          
          <code><a href="treemap_pq.html">treemap_pq()</a></code> 
        </dt>
        <dd>Plot treemap of 2 taxonomic levels</dd>
      </dl><dl><dt>
          
          <code><a href="vs_search_global.html">vs_search_global()</a></code> 
        </dt>
        <dd>Search for a list of sequence in a fasta file against physeq reference sequences using <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a></dd>
      </dl></div><div class="section level2">
      
      <h3 id="zoom-on-given-taxa">Zoom on given taxa<a class="anchor" aria-label="anchor" href="#zoom-on-given-taxa"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="distri_1_taxa.html">distri_1_taxa()</a></code> 
        </dt>
        <dd>Distribution of sequences across a factor for one taxon</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_only_in_one_level.html">taxa_only_in_one_level()</a></code> 
        </dt>
        <dd>Show taxa which are present in only one given level of a modality</dd>
      </dl></div><div class="section level2">
      <h2 id="improve-reproducibility-using-targets-https-books-ropensci-org-targets-pipeline">Improve reproducibility using <a href="https://books.ropensci.org/targets/" class="external-link">targets</a> pipeline<a class="anchor" aria-label="anchor" href="#improve-reproducibility-using-targets-https-books-ropensci-org-targets-pipeline"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="processing-fastq-files">Processing fastq files<a class="anchor" aria-label="anchor" href="#processing-fastq-files"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="list_fastq_files.html">list_fastq_files()</a></code> 
        </dt>
        <dd>List fastq files</dd>
      </dl><dl><dt>
          
          <code><a href="filter_trim.html">filter_trim()</a></code> 
        </dt>
        <dd>A wrapper of the function <code><a href="https://rdrr.io/pkg/dada2/man/filterAndTrim.html" class="external-link">dada2::filterAndTrim()</a></code> to use in <a href="https://books.ropensci.org/targets/" class="external-link">targets</a> pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="cutadapt_remove_primers.html">cutadapt_remove_primers()</a></code> 
        </dt>
        <dd>Remove primers using <a href="https://github.com/marcelm/cutadapt/" class="external-link">cutadapt</a></dd>
      </dl><dl><dt>
          
          <code><a href="subsample_fastq.html">subsample_fastq()</a></code> 
        </dt>
        <dd>Subsample a fastq file copying the n_seq first sequences in a given folder</dd>
      </dl></div><div class="section level2">
      
      <h3 id="import-export-phyloseq-object">Import/export phyloseq object<a class="anchor" aria-label="anchor" href="#import-export-phyloseq-object"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="read_pq.html">read_pq()</a></code> 
        </dt>
        <dd>Read phyloseq object from multiple csv tables and a phylogenetic tree in Newick format.</dd>
      </dl><dl><dt>
          
          <code><a href="save_pq.html">save_pq()</a></code> 
        </dt>
        <dd>A wrapper of write_pq to save in all three possible formats</dd>
      </dl><dl><dt>
          
          <code><a href="write_pq.html">write_pq()</a></code> 
        </dt>
        <dd>Save phyloseq object in the form of multiple csv tables.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="track-info-along-pipeline">Track info along pipeline<a class="anchor" aria-label="anchor" href="#track-info-along-pipeline"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="count_seq.html">count_seq()</a></code> 
        </dt>
        <dd>Count sequences in fasta or fastq file</dd>
      </dl><dl><dt>
          
          <code><a href="track_wkflow.html">track_wkflow()</a></code> 
        </dt>
        <dd>Track the number of reads (= sequences), samples and cluster (e.g. ASV) from various objects including dada-class and derep-class.</dd>
      </dl><dl><dt>
          
          <code><a href="track_wkflow_samples.html">track_wkflow_samples()</a></code> 
        </dt>
        <dd>Track the number of reads (= sequences), samples and cluster (e.g. ASV) for each sample</dd>
      </dl></div><div class="section level2">
      
      <h3 id="sample-data-management">Sample data management<a class="anchor" aria-label="anchor" href="#sample-data-management"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="rename_samples_otu_table.html">rename_samples_otu_table()</a></code> 
        </dt>
        <dd>Rename samples of an otu_table</dd>
      </dl><dl><dt>
          
          <code><a href="rename_samples.html">rename_samples()</a></code> 
        </dt>
        <dd>Rename the samples of a phyloseq slot</dd>
      </dl><dl><dt>
          
          <code><a href="sample_data_with_new_names.html">sample_data_with_new_names()</a></code> 
        </dt>
        <dd>Load sample data from file and rename samples using names of samples and an optional order</dd>
      </dl><dl><dt>
          
          <code><a href="psmelt_samples_pq.html">psmelt_samples_pq()</a></code> 
        </dt>
        <dd>Build a sample information tibble from physeq object</dd>
      </dl></div><div class="section level2">
      
      <h3 id="taxonomy-management">Taxonomy management<a class="anchor" aria-label="anchor" href="#taxonomy-management"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="simplify_taxo.html">simplify_taxo()</a></code> 
        </dt>
        <dd>Simplify taxonomy by removing some unused characters such as "k__"</dd>
      </dl><dl><dt>
          
          <code><a href="reorder_taxa_pq.html">reorder_taxa_pq()</a></code> 
        </dt>
        <dd>Reorder taxa in otu_table/tax_table/refseq slot of a phyloseq object</dd>
      </dl></div><div class="section level2">
      <h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="data_fungi.html">data_fungi</a></code> 
        </dt>
        <dd>Fungal OTU in phyloseq format</dd>
      </dl><dl><dt>
          
          <code><a href="data_fungi_sp_known.html">data_fungi_sp_known</a></code> 
        </dt>
        <dd>Fungal OTU in phyloseq format</dd>
      </dl><dl><dt>
          
          <code><a href="data_fungi_mini.html">data_fungi_mini</a></code> 
        </dt>
        <dd>Fungal OTU in phyloseq format</dd>
      </dl><dl><dt>
          
          <code><a href="Tengeler2020_pq.html">Tengeler2020_pq</a></code> 
        </dt>
        <dd>This tutorial explores the dataset from Tengeler et al. (2020) available in the <code>mia</code> package. obtained using <code>mia::makePhyloseqFromTreeSE(Tengeler2020)</code></dd>
      </dl></div><div class="section level2">
      <h2 id="other-utilities">Other utilities<a class="anchor" aria-label="anchor" href="#other-utilities"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="color-management">Color management<a class="anchor" aria-label="anchor" href="#color-management"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="funky_color.html">funky_color()</a></code> 
        </dt>
        <dd>Funky palette color</dd>
      </dl><dl><dt>
          
          <code><a href="fac2col.html">fac2col()</a></code> 
        </dt>
        <dd>Translates a factor into colors.</dd>
      </dl><dl><dt>
          
          <code><a href="transp.html">transp()</a></code> 
        </dt>
        <dd>Adds transparency to a vector of colors</dd>
      </dl></div><div class="section level2">
      
      <h3 id="file-management">File management<a class="anchor" aria-label="anchor" href="#file-management"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="get_file_extension.html">get_file_extension()</a></code> 
        </dt>
        <dd>Get the extension of a file</dd>
      </dl></div><div class="section level2">
      
      <h3 id="external-software-management">External software management<a class="anchor" aria-label="anchor" href="#external-software-management"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="is_cutadapt_installed.html">is_cutadapt_installed()</a></code> 
        </dt>
        <dd>Test if cutadapt is installed.</dd>
      </dl><dl><dt>
          
          <code><a href="is_falco_installed.html">is_falco_installed()</a></code> 
        </dt>
        <dd>Test if falco is installed.</dd>
      </dl><dl><dt>
          
          <code><a href="is_krona_installed.html">is_krona_installed()</a></code> 
        </dt>
        <dd>Test if krona is installed.</dd>
      </dl><dl><dt>
          
          <code><a href="is_mumu_installed.html">is_mumu_installed()</a></code> 
        </dt>
        <dd>Test if mumu is installed.</dd>
      </dl><dl><dt>
          
          <code><a href="is_swarm_installed.html">is_swarm_installed()</a></code> 
        </dt>
        <dd>Test if swarm is installed.</dd>
      </dl><dl><dt>
          
          <code><a href="is_vsearch_installed.html">is_vsearch_installed()</a></code> 
        </dt>
        <dd>Test if vsearch is installed.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="plot-management">Plot management<a class="anchor" aria-label="anchor" href="#plot-management"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="multiplot.html">multiplot()</a></code> 
        </dt>
        <dd>Multiple plot function</dd>
      </dl></div><div class="section level2">
      
      <h3 id="variable-management">Variable management<a class="anchor" aria-label="anchor" href="#variable-management"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="diff_fct_diff_class.html">diff_fct_diff_class()</a></code> 
        </dt>
        <dd>Compute different functions for different class of vector.</dd>
      </dl><dl><dt>
          
          <code><a href="perc.html">perc()</a></code> 
        </dt>
        <dd>Convert a value (or a fraction x/y) in percentage</dd>
      </dl><dl><dt>
          
          <code><a href="unique_or_na.html">unique_or_na()</a></code> 
        </dt>
        <dd>Get the unique value in x or NA if none</dd>
      </dl></div><div class="section level2">
      
      <h3 id="others-1">Others<a class="anchor" aria-label="anchor" href="#others-1"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="all_object_size.html">all_object_size()</a></code> 
        </dt>
        <dd>List the size of all objects of the GlobalEnv.</dd>
      </dl><dl><dt>
          
          <code><a href="dist_bycol.html">dist_bycol()</a></code> 
        </dt>
        <dd>Compute paired distances among matrix (e.g. otu_table)</dd>
      </dl><dl><dt>
          
          <code><a href="dist_pos_control.html">dist_pos_control()</a></code> 
        </dt>
        <dd>Calculate ecological distance among positive controls vs distance for all samples</dd>
      </dl><dl><dt>
          
          <code><a href="are_modality_even_depth.html">are_modality_even_depth()</a></code> 
        </dt>
        <dd>Test if the mean number of sequences by samples is link to the modality of a factor</dd>
      </dl><dl><dt>
          
          <code><a href="MiscMetabar-package.html">MiscMetabar-package</a></code> 
        </dt>
        <dd><code>MiscMetabar</code> package</dd>
      </dl></div><div class="section level2">
      <h2 id="deprecated">Deprecated<a class="anchor" aria-label="anchor" href="#deprecated"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="MiscMetabar-deprecated.html">physeq_graph_test()</a></code> 
        </dt>
        <dd>Deprecated function(s) in the MiscMetabar package</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

