<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Post Clustering Curation of Amplicon Data. — lulu • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Post Clustering Curation of Amplicon Data. — lulu"><meta name="description" content="


The original function and documentation was written by Tobias Guldberg Frøslev
in the lulu package.
This algorithm lulu consumes an OTU table and a matchlist, and
evaluates cooccurence of 'daughters' (potential analytical artefacts) and
their 'parents' (~= real biological species/OTUs). The algorithm requires an
OTU table (species/site matrix), and a match list. The OTU table can be
made with various r-packages (e.g. DADA2) or
external pipelines (VSEARCH, USEARCH, QIIME, etc.), and the
match-list can be made with external bioinformatic tools like
VSEARCH, USEARCH, BLASTN or another algorithm
for pair-wise sequence matching."><meta property="og:description" content="


The original function and documentation was written by Tobias Guldberg Frøslev
in the lulu package.
This algorithm lulu consumes an OTU table and a matchlist, and
evaluates cooccurence of 'daughters' (potential analytical artefacts) and
their 'parents' (~= real biological species/OTUs). The algorithm requires an
OTU table (species/site matrix), and a match list. The OTU table can be
made with various r-packages (e.g. DADA2) or
external pipelines (VSEARCH, USEARCH, QIIME, etc.), and the
match-list can be made with external bioinformatic tools like
VSEARCH, USEARCH, BLASTN or another algorithm
for pair-wise sequence matching."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.12.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Post Clustering Curation of Amplicon Data.</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/lulu.R" class="external-link"><code>R/lulu.R</code></a></small>
      <div class="d-none name"><code>lulu.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-stable-green" alt="lifecycle-stable"></a></p>

<p>The original function and documentation was written by Tobias Guldberg Frøslev
in the <a href="https://github.com/tobiasgf/lulu" class="external-link">lulu</a> package.</p>
<p>This algorithm <code>lulu</code> consumes an OTU table and a matchlist, and
evaluates cooccurence of 'daughters' (potential analytical artefacts) and
their 'parents' (~= real biological species/OTUs). The algorithm requires an
OTU table (species/site matrix), and a match list. The OTU table can be
made with various r-packages (e.g. <code>DADA2</code>) or
external pipelines (<code>VSEARCH, USEARCH, QIIME</code>, etc.), and the
match-list can be made with external bioinformatic tools like
<code>VSEARCH, USEARCH, BLASTN</code> or another algorithm
for pair-wise sequence matching.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">lulu</span><span class="op">(</span></span>
<span>  <span class="va">otu_table</span>,</span>
<span>  <span class="va">matchlist</span>,</span>
<span>  minimum_ratio_type <span class="op">=</span> <span class="st">"min"</span>,</span>
<span>  minimum_ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minimum_match <span class="op">=</span> <span class="fl">84</span>,</span>
<span>  minimum_relative_cooccurence <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  progress_bar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  log_conserved <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-otu-table">otu_table<a class="anchor" aria-label="anchor" href="#arg-otu-table"></a></dt>
<dd><p>a data.frame with with an OTU table that has sites/samples as
columns and OTUs (unique OTU id's) as rows, and observations as read
counts.</p></dd>


<dt id="arg-matchlist">matchlist<a class="anchor" aria-label="anchor" href="#arg-matchlist"></a></dt>
<dd><p>a data.frame containing three columns: (1) OTU id of
potential child, (2) OTU id of potential parent, (3) match - % identiti
between the sequences of the potential parent and potential child OTUs.
<strong>NB: The matchlist is the product of a mapping of OTU sequences against each other. This is
currently carried out by an external script in e.g. Blastn or VSEARCH, prior to running lulu!</strong></p></dd>


<dt id="arg-minimum-ratio-type">minimum_ratio_type<a class="anchor" aria-label="anchor" href="#arg-minimum-ratio-type"></a></dt>
<dd><p>sets whether a potential error must have lower
abundance than the parent in all samples <code>min</code> (default), or if an error
just needs to have lower abundance on average <code>avg</code>. Choosing lower
abundance on average over globally lower abundance will greatly increase
the number of designated errors. This option was introduced to make it
possible to account for non-sufficiently clustered intraspecific variation,
but is not generally recommended, as it will also increase the potential of
cluster well-separated, but co-occuring, sequence similar species.</p></dd>


<dt id="arg-minimum-ratio">minimum_ratio<a class="anchor" aria-label="anchor" href="#arg-minimum-ratio"></a></dt>
<dd><p>sets the minimim abundance ratio between a potential error
and a potential parent to be identified as an error. If the <code>minimum_ratio_type</code> is
set to <code>min</code> (default), the <code>minimum_ratio</code> applies to the lowest observed
ration across the samples.  If the <code>minimum_ratio_type</code> is
set to <code>avg</code> (default), the <code>minimum_ratio</code> applies to the mean of observed
ration across the samples.<code>avg</code>. (default is 1).</p></dd>


<dt id="arg-minimum-match">minimum_match<a class="anchor" aria-label="anchor" href="#arg-minimum-match"></a></dt>
<dd><p>minimum threshold of sequence similarity
for considering any OTU as an error of another can be set (default 84%).</p></dd>


<dt id="arg-minimum-relative-cooccurence">minimum_relative_cooccurence<a class="anchor" aria-label="anchor" href="#arg-minimum-relative-cooccurence"></a></dt>
<dd><p>minimum co-occurrence rate, i.e. the
lower rate of occurrence of the potential error explained by co-occurrence
with the potential parent for considering error state.</p></dd>


<dt id="arg-progress-bar">progress_bar<a class="anchor" aria-label="anchor" href="#arg-progress-bar"></a></dt>
<dd><p>(Logical, default TRUE) print progress during the calculation or not.</p></dd>


<dt id="arg-log-conserved">log_conserved<a class="anchor" aria-label="anchor" href="#arg-log-conserved"></a></dt>
<dd><p>(Logical, default FALSE) conserved log files writed in the disk</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Function <code>lulu</code> returns a list of results based on the input OTU
table and match list.</p><ul><li><p><code>curated_table</code> - a curated
OTU table with daughters merged with their matching parents.</p></li>
<li><p><code>curated_count</code> - number of curated (parent) OTUs.</p></li>
<li><p><code>curated_otus</code> - ids of the OTUs that were accepted as valid OTUs.</p></li>
<li><p><code>discarded_count</code> - number of discarded (merged with parent) OTUs.</p></li>
<li><p><code>discarded_otus</code> - ids of the OTUs that were identified as
errors (daughters) and merged with respective parents.</p></li>
<li><p><code>runtime</code> - time used by the script.</p></li>
<li><p><code>minimum_match</code> - the id threshold
(minimum match \<!-- % between parent and daughter) for evaluating co-occurence (set -->
by user).</p></li>
<li><p><code>minimum_relative_cooccurence</code> - minimum ratio of
daughter-occurences explained by co-occurence with parent (set by user).</p></li>
<li><p><code>otu_map</code> - information of which daughters were mapped to which
parents.</p></li>
<li><p><code>original_table</code> - original OTU table.</p></li>
</ul><p>The matchlist is the product of a mapping of OTU sequences against each other. This is
currently carried out by an external script in e.g. BLASTN or VSEARCH, prior to running <code>lulu</code>!
Producing the match list requires a file with all the OTU sequences (centroids) - e.g. <code>OTUcentroids.fasta</code>. The matchlist can be produced by mapping all OTUs against each other with an external algorithm like VSEARCH or BLASTN. In <code>VSEARCH</code> a matchlist can be produced e.g. with the following command: <code>vsearch --usearch_global OTUcentroids.fasta --db OTUcentroids.fasta --strand plus --self --id .80 --iddef 1 --userout matchlist.txt --userfields query+target+id --maxaccepts 0 --query_cov .9 --maxhits 10</code>. In <code>BLASTN</code> a matchlist can be produces e.g. with the following commands. First we produce a blast-database from the fasta file: <code>makeblastdb -in OTUcentroids.fasta -parse_seqids -dbtype nucl</code>, then we match the centroids against that database: <code>blastn -db OTUcentoids.fasta -num_threads 10 -outfmt'6 qseqid sseqid pident' -out matchlist.txt -qcov_hsp_perc .90 -perc_identity .84 -query OTUcentroids.fasta</code></p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Please cite the lulu original paper: https://www.nature.com/articles/s41467-017-01312-x</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Tobias Guldberg Frøslev (orcid: <a href="https://orcid.org/0000-0002-3530-013X" class="external-link">0000-0002-3530-013X</a>),
modified by Adrien Taudière</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

