<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Merge taxa in groups (vectorized version) — merge_taxa_vec • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Merge taxa in groups (vectorized version) — merge_taxa_vec"><meta name="description" content="


Firstly release in the speedyseq R
package by Michael R. McLaren.
Merge taxa in x into a smaller set of taxa defined by the vector group.
Taxa whose value in group is NA will be dropped. New taxa will be named
according to the most abundant taxon in each group (phyloseq and
otu_table objects) or the first taxon in each group (all other phyloseq
component objects).
If x is a phyloseq object with a phylogenetic tree, then the new taxa will
be ordered as they are in the tree. Otherwise, the taxa order can be
controlled by the reorder argument, which behaves like the reorder
argument in base::rowsum(). reorder = FALSE will keep taxa in
the original order determined by when the member of each group first appears
in taxa_names(x); reorder = TRUE will order new taxa according to their
corresponding value in group.
The tax_adjust argument controls the handling of taxonomic disagreements
within groups. Setting tax_adjust == 0 causes no adjustment; the taxonomy
of the new group is set to the archetype taxon (see below). Otherwise,
disagreements within a group at a given rank cause the values at lower ranks
to be set to NA. If tax_adjust == 1 (the default), then a rank where all
taxa in the group are already NA is not counted as a disagreement, and lower
ranks may be kept if the taxa agree. This corresponds to the original
phyloseq behavior. If tax_adjust == 2, then these NAs are treated as a
disagreement; all ranks are set to NA after the first disagreement or NA."><meta property="og:description" content="


Firstly release in the speedyseq R
package by Michael R. McLaren.
Merge taxa in x into a smaller set of taxa defined by the vector group.
Taxa whose value in group is NA will be dropped. New taxa will be named
according to the most abundant taxon in each group (phyloseq and
otu_table objects) or the first taxon in each group (all other phyloseq
component objects).
If x is a phyloseq object with a phylogenetic tree, then the new taxa will
be ordered as they are in the tree. Otherwise, the taxa order can be
controlled by the reorder argument, which behaves like the reorder
argument in base::rowsum(). reorder = FALSE will keep taxa in
the original order determined by when the member of each group first appears
in taxa_names(x); reorder = TRUE will order new taxa according to their
corresponding value in group.
The tax_adjust argument controls the handling of taxonomic disagreements
within groups. Setting tax_adjust == 0 causes no adjustment; the taxonomy
of the new group is set to the archetype taxon (see below). Otherwise,
disagreements within a group at a given rank cause the values at lower ranks
to be set to NA. If tax_adjust == 1 (the default), then a rank where all
taxa in the group are already NA is not counted as a disagreement, and lower
ranks may be kept if the taxa agree. This corresponds to the original
phyloseq behavior. If tax_adjust == 2, then these NAs are treated as a
disagreement; all ranks are set to NA after the first disagreement or NA."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Merge taxa in groups (vectorized version)</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/speedyseq_functions.R" class="external-link"><code>R/speedyseq_functions.R</code></a></small>
      <div class="d-none name"><code>merge_taxa_vec.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-stable-green" alt="lifecycle-stable"></a></p>

<p>Firstly release in the <a href="https://github.com/mikemc/speedyseq/" class="external-link">speedyseq</a> R
package by Michael R. McLaren.</p>
<p>Merge taxa in <code>x</code> into a smaller set of taxa defined by the vector <code>group</code>.
Taxa whose value in <code>group</code> is NA will be dropped. New taxa will be named
according to the most abundant taxon in each group (<code>phyloseq</code> and
<code>otu_table</code> objects) or the first taxon in each group (all other phyloseq
component objects).</p>
<p>If <code>x</code> is a phyloseq object with a phylogenetic tree, then the new taxa will
be ordered as they are in the tree. Otherwise, the taxa order can be
controlled by the <code>reorder</code> argument, which behaves like the <code>reorder</code>
argument in <code><a href="https://rdrr.io/r/base/rowsum.html" class="external-link">base::rowsum()</a></code>. <code>reorder = FALSE</code> will keep taxa in
the original order determined by when the member of each group first appears
in <code>taxa_names(x)</code>; <code>reorder = TRUE</code> will order new taxa according to their
corresponding value in <code>group</code>.</p>
<p>The <code>tax_adjust</code> argument controls the handling of taxonomic disagreements
within groups. Setting <code>tax_adjust == 0</code> causes no adjustment; the taxonomy
of the new group is set to the archetype taxon (see below). Otherwise,
disagreements within a group at a given rank cause the values at lower ranks
to be set to <code>NA</code>. If <code>tax_adjust == 1</code> (the default), then a rank where all
taxa in the group are already NA is not counted as a disagreement, and lower
ranks may be kept if the taxa agree. This corresponds to the original
phyloseq behavior. If <code>tax_adjust == 2</code>, then these NAs are treated as a
disagreement; all ranks are set to NA after the first disagreement or NA.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">merge_taxa_vec</span><span class="op">(</span><span class="va">x</span>, <span class="va">group</span>, reorder <span class="op">=</span> <span class="cn">FALSE</span>, tax_adjust <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'phyloseq'</span></span>
<span><span class="fu">merge_taxa_vec</span><span class="op">(</span><span class="va">x</span>, <span class="va">group</span>, reorder <span class="op">=</span> <span class="cn">FALSE</span>, tax_adjust <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'otu_table'</span></span>
<span><span class="fu">merge_taxa_vec</span><span class="op">(</span><span class="va">x</span>, <span class="va">group</span>, reorder <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'taxonomyTable'</span></span>
<span><span class="fu">merge_taxa_vec</span><span class="op">(</span><span class="va">x</span>, <span class="va">group</span>, reorder <span class="op">=</span> <span class="cn">FALSE</span>, tax_adjust <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'phylo'</span></span>
<span><span class="fu">merge_taxa_vec</span><span class="op">(</span><span class="va">x</span>, <span class="va">group</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'XStringSet'</span></span>
<span><span class="fu">merge_taxa_vec</span><span class="op">(</span><span class="va">x</span>, <span class="va">group</span>, reorder <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A phyloseq object or component object</p></dd>


<dt id="arg-group">group<a class="anchor" aria-label="anchor" href="#arg-group"></a></dt>
<dd><p>A vector with one element for each taxon in <code>physeq</code> that
defines the new groups. see <code><a href="https://rdrr.io/r/base/rowsum.html" class="external-link">base::rowsum()</a></code>.</p></dd>


<dt id="arg-reorder">reorder<a class="anchor" aria-label="anchor" href="#arg-reorder"></a></dt>
<dd><p>Logical specifying whether to reorder the taxa by their
<code>group</code> values. Ignored if <code>x</code> has (or is) a phylogenetic tree.</p></dd>


<dt id="arg-tax-adjust">tax_adjust<a class="anchor" aria-label="anchor" href="#arg-tax-adjust"></a></dt>
<dd><p>0: no adjustment; 1: phyloseq-compatible adjustment; 2:
conservative adjustment</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A new phyloseq-class, otu_table, tax_table, XStringset or
sam_data object depending on the class of the x param</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Function in MiscMetabar that use this function: <code><a href="postcluster_pq.html">postcluster_pq()</a></code></p>
<p><code><a href="https://rdrr.io/r/base/rowsum.html" class="external-link">base::rowsum()</a></code></p>
<p><code><a href="https://rdrr.io/pkg/phyloseq/man/merge_taxa-methods.html" class="external-link">phyloseq::merge_taxa()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Michael R. McLaren (orcid: <a href="https://orcid.org/0000-0003-1575-473X" class="external-link">0000-0003-1575-473X</a>) modified by Adrien Taudiere</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

