<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot the nucleotide proportion at both extremity of the sequences — plot_refseq_extremity_pq • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot the nucleotide proportion at both extremity of the sequences — plot_refseq_extremity_pq"><meta name="description" content="


It is a useful function to check for the absence of unwanted patterns caused
for example by Illumina adaptator or bad removal of primers.
If hill_scale is not null, Hill diversity number are used to represent the distribution
of the diversity (equitability) along the sequences."><meta property="og:description" content="


It is a useful function to check for the absence of unwanted patterns caused
for example by Illumina adaptator or bad removal of primers.
If hill_scale is not null, Hill diversity number are used to represent the distribution
of the diversity (equitability) along the sequences."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot the nucleotide proportion at both extremity of the sequences</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/plot_functions.R" class="external-link"><code>R/plot_functions.R</code></a></small>
      <div class="d-none name"><code>plot_refseq_extremity_pq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>It is a useful function to check for the absence of unwanted patterns caused
for example by Illumina adaptator or bad removal of primers.</p>
<p>If <code>hill_scale</code> is not null, Hill diversity number are used to represent the distribution
of the diversity (equitability) along the sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_refseq_extremity_pq</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  first_n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  last_n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  hill_scales <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  min_width <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required) a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-first-n">first_n<a class="anchor" aria-label="anchor" href="#arg-first-n"></a></dt>
<dd><p>(int, default 10) The number of nucleotides to plot the 5' extremity.</p></dd>


<dt id="arg-last-n">last_n<a class="anchor" aria-label="anchor" href="#arg-last-n"></a></dt>
<dd><p>(int, default 10) The number of nucleotides to plot the 3' extremity.</p></dd>


<dt id="arg-hill-scales">hill_scales<a class="anchor" aria-label="anchor" href="#arg-hill-scales"></a></dt>
<dd><p>(vector) A vector defining the Hill number wanted. Set to NULL if
you don't want to plot Hill diversity metrics.</p></dd>


<dt id="arg-min-width">min_width<a class="anchor" aria-label="anchor" href="#arg-min-width"></a></dt>
<dd><p>(int, default 0) Select only the sequences from physeq@refseq with using a
minimum length threshold. If <code>first_n</code> is superior to the minimum length of the
references sequences, you must use min_width to filter out the narrower sequences</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of 4 objects</p><ul><li><p>p_start and p_last are the ggplot object representing respectively the start and
the end of the sequences.</p></li>
<li><p>df_start and df_last are the data.frame corresponding to the ggplot object.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">res1</span> <span class="op">&lt;-</span> <span class="fu">plot_refseq_extremity_pq</span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "plot_start" "plot_last"  "df_start"   "df_end"    </span>
<span class="r-in"><span><span class="va">res1</span><span class="op">$</span><span class="va">plot_start</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 1652 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_point()`).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 826 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_line()`).</span>
<span class="r-plt img"><img src="plot_refseq_extremity_pq-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">res1</span><span class="op">$</span><span class="va">plot_last</span></span></span>
<span class="r-plt img"><img src="plot_refseq_extremity_pq-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu">plot_refseq_extremity_pq</span><span class="op">(</span><span class="va">data_fungi</span>, first_n <span class="op">=</span> <span class="fl">200</span>, last_n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res2</span><span class="op">$</span><span class="va">plot_start</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 892 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_point()`).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 446 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_line()`).</span>
<span class="r-plt img"><img src="plot_refseq_extremity_pq-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">res2</span><span class="op">$</span><span class="va">plot_last</span></span></span>
<span class="r-plt img"><img src="plot_refseq_extremity_pq-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">plot_refseq_extremity_pq</span><span class="op">(</span><span class="va">data_fungi</span>,</span></span>
<span class="r-in"><span>  first_n <span class="op">=</span> <span class="fl">400</span>,</span></span>
<span class="r-in"><span>  min_width <span class="op">=</span> <span class="fl">400</span>,</span></span>
<span class="r-in"><span>  hill_scales <span class="op">=</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">$</span><span class="va">plot_start</span> <span class="op">+</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">value</span>, x <span class="op">=</span> <span class="va">seq_id</span>, color <span class="op">=</span> <span class="va">name</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, linewidth <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 165 sample(s) ( A10-005-B_S188_MERGED.fastq.gz / A10-005-H_S189_MERGED.fastq.gz / A10-005-M_S190_MERGED.fastq.gz / A12-007_S191_MERGED.fastq.gz / A12-007-B_S2_MERGED.fastq.gz / A8-005_S4_MERGED.fastq.gz / A9-012_S5_MERGED.fastq.gz / AB29-ABMX-H_S6_MERGED.fastq.gz / AC29033_S8_MERGED.fastq.gz / AD26-005-B_S9_MERGED.fastq.gz / AD26-005-H_S10_MERGED.fastq.gz / AD30-ABMX-M_S12_MERGED.fastq.gz / AD32-007-M_S13_MERGED.fastq.gz / ADABM30X-H_S15_MERGED.fastq.gz / ADABM30X-M_S16_MERGED.fastq.gz / B17-014_S18_MERGED.fastq.gz / B18-006-B_S19_MERGED.fastq.gz / BA16-036bis_S20_MERGED.fastq.gz / BA17-050-B_S21_MERGED.fastq.gz / BB19-006-H_S22_MERGED.fastq.gz / BB6-019-B_S23_MERGED.fastq.gz / BB6-019-H_S24_MERGED.fastq.gz / BD14-021_S26_MERGED.fastq.gz / BE9-006-B_S27_MERGED.fastq.gz / BE9-006-H_S28_MERGED.fastq.gz / BE9-006-M_S29_MERGED.fastq.gz / BG7-010-B_S30_MERGED.fastq.gz / BG7-010-H_S31_MERGED.fastq.gz / BG7-010-M_S32_MERGED.fastq.gz / BH9-021_S33_MERGED.fastq.gz / BJ17-007-M_S34_MERGED.fastq.gz / BJ8-ABM-003_S35_MERGED.fastq.gz / BL7-006-B_S36_MERGED.fastq.gz / BL7-006-H_S37_MERGED.fastq.gz / BL7-006-M_S38_MERGED.fastq.gz / BN11-041_S39_MERGED.fastq.gz / BO8-002_S41_MERGED.fastq.gz / BO8-005_S42_MERGED.fastq.gz / BP11-001-B_S43_MERGED.fastq.gz / BP11-001-H_S44_MERGED.fastq.gz / BP11-001-M_S45_MERGED.fastq.gz / BP12-025-B_S46_MERGED.fastq.gz / BP14-006_S47_MERGED.fastq.gz / BQ3-019_S48_MERGED.fastq.gz / BQ4-018-B_S49_MERGED.fastq.gz / BQ4-018-H_S50_MERGED.fastq.gz / BQ9ABM-002_S52_MERGED.fastq.gz / BS14-006_S54_MERGED.fastq.gz / BT-006-M_S55_MERGED.fastq.gz / BT7-006_S56_MERGED.fastq.gz / BV11-002-B_S57_MERGED.fastq.gz / BV11-002-M_S59_MERGED.fastq.gz / C1-001_S61_MERGED.fastq.gz / C21-NV1-H_S63_MERGED.fastq.gz / C21-NV1-M_S64_MERGED.fastq.gz / CA12-024_S66_MERGED.fastq.gz / CA9-027_S67_MERGED.fastq.gz / CA9-X_S68_MERGED.fastq.gz / CB8-019-B_S69_MERGED.fastq.gz / CB8-019-H_S70_MERGED.fastq.gz / CB8-019-M_S71_MERGED.fastq.gz / CB9-013_S72_MERGED.fastq.gz / CC3-044_S73_MERGED.fastq.gz / CC8-003_S74_MERGED.fastq.gz / D18-003-B_S78_MERGED.fastq.gz / D18-003-H_S79_MERGED.fastq.gz / D18-003-M_S80_MERGED.fastq.gz / D22-NVABM1_S81_MERGED.fastq.gz / D61-010-B_S82_MERGED.fastq.gz / D9-027-H_S84_MERGED.fastq.gz / D9-027-M_S85_MERGED.fastq.gz / DBM-ABM-001_S86_MERGED.fastq.gz / DJ2-008-B_S87_MERGED.fastq.gz / DJ2-008-H_S88_MERGED.fastq.gz / DP4-ABM001_S90_MERGED.fastq.gz / DS1-ABM002-B_S91_MERGED.fastq.gz / DS1-ABM002-H_S92_MERGED.fastq.gz / DS1-ABM002-M_S93_MERGED.fastq.gz / DU3-045-B_S94_MERGED.fastq.gz / DW4-007_S95_MERGED.fastq.gz / DY5-004-B_S96_MERGED.fastq.gz / DY5-004-H_S97_MERGED.fastq.gz / DY5-004-M_S98_MERGED.fastq.gz / DZ6-ABM-001_S99_MERGED.fastq.gz / E9-009-B_S100_MERGED.fastq.gz / E9-009-H_S101_MERGED.fastq.gz / E9-009-M_S102_MERGED.fastq.gz / EA5-ABM-001_S103_MERGED.fastq.gz / EC2-013-B_S104_MERGED.fastq.gz / F6-ABM-001_S105_MERGED.fastq.gz / F7-015-M_S106_MERGED.fastq.gz / FOMES19-H_S108_MERGED.fastq.gz / FOMES19-M_S109_MERGED.fastq.gz / H10-018-M_S110_MERGED.fastq.gz / H24-NVABM1-H_S111_MERGED.fastq.gz / J18-004-B_S114_MERGED.fastq.gz / J18-004-H_S115_MERGED.fastq.gz / J18-004-M_S116_MERGED.fastq.gz / K18-002-H_S117_MERGED.fastq.gz / K26-NVABM1_S118_MERGED.fastq.gz / L19X-H_S120_MERGED.fastq.gz / L19X-M_S121_MERGED.fastq.gz / L23-002-B_S122_MERGED.fastq.gz / L23-002-H_S123_MERGED.fastq.gz / M22-001_S125_MERGED.fastq.gz / N19X-B_S126_MERGED.fastq.gz / N19X-H_S127_MERGED.fastq.gz / N19X-M_S128_MERGED.fastq.gz / N22-001-B_S129_MERGED.fastq.gz / N23-002-B_S130_MERGED.fastq.gz / N23-002-H_S131_MERGED.fastq.gz / N23-002-M_S132_MERGED.fastq.gz / N25-ABMX_S133_MERGED.fastq.gz / NVABM-0058_S134_MERGED.fastq.gz / NVABM-0163-H_S135_MERGED.fastq.gz / NVABM-0397_S138_MERGED.fastq.gz / NVABM0216_S136_MERGED.fastq.gz / NVABM0244-M_S137_MERGED.fastq.gz / O20-X-B_S139_MERGED.fastq.gz / O20-X-H_S140_MERGED.fastq.gz / O20-X-M_S141_MERGED.fastq.gz / O21-007-B_S142_MERGED.fastq.gz / O21-007-H_S143_MERGED.fastq.gz / O21-007-M_S144_MERGED.fastq.gz / O24-003-H_S146_MERGED.fastq.gz / O24-003-M_S147_MERGED.fastq.gz / O26-004-B_S148_MERGED.fastq.gz / O26-004-H_S149_MERGED.fastq.gz / O26-004-M_S150_MERGED.fastq.gz / O27-012_S151_MERGED.fastq.gz / O9-005-B_S152_MERGED.fastq.gz / P19-023-M_S153_MERGED.fastq.gz / P27-015-M_S154_MERGED.fastq.gz / P27-ABM001_S155_MERGED.fastq.gz / Q27-ABM003-B_S156_MERGED.fastq.gz / R25-ABMX_S157_MERGED.fastq.gz / R28-008-B_S158_MERGED.fastq.gz / R28-008-H_S159_MERGED.fastq.gz / R28-008-M_S160_MERGED.fastq.gz / T28-ABM602-B_S162_MERGED.fastq.gz / U27-ABM002_S163_MERGED.fastq.gz / W25-ABMX_S164_MERGED.fastq.gz / W26-001-B_S165_MERGED.fastq.gz / W26-001-H_S166_MERGED.fastq.gz / W26-001-M_S167_MERGED.fastq.gz / W30-006_S168_MERGED.fastq.gz / W9-025-M_S169_MERGED.fastq.gz / X24-009-B_S170_MERGED.fastq.gz / X24-009-H_S171_MERGED.fastq.gz / X24-009-M_S172_MERGED.fastq.gz / X24-010_S173_MERGED.fastq.gz / X29-004-B_S174_MERGED.fastq.gz / X29-004-H_S175_MERGED.fastq.gz / X29-004-M_S176_MERGED.fastq.gz / Y21-ABM484-H_S177_MERGED.fastq.gz / Y28-002-B_S178_MERGED.fastq.gz / Y28-002-H_S179_MERGED.fastq.gz / Y28-002-M_S180_MERGED.fastq.gz / Y29-007-B_S181_MERGED.fastq.gz / Y29-007-H_S182_MERGED.fastq.gz / Y29-007-M_S183_MERGED.fastq.gz / Y31-ABM484-B_S184_MERGED.fastq.gz / Z29-001-H_S185_MERGED.fastq.gz / Z30-002_S186_MERGED.fastq.gz / Z30-ABM560-M_S187_MERGED.fastq.gz ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 1413</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 7</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 20</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 92 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_point()`).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 92 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_line()`).</span>
<span class="r-plt img"><img src="plot_refseq_extremity_pq-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">plot_refseq_extremity_pq</span><span class="op">(</span><span class="va">data_fungi</span>,</span></span>
<span class="r-in"><span>  first_n <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  last_n <span class="op">=</span> <span class="fl">400</span>,</span></span>
<span class="r-in"><span>  min_width <span class="op">=</span> <span class="fl">400</span>,</span></span>
<span class="r-in"><span>  hill_scales <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">$</span><span class="va">plot_last</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cleaning suppress 0 taxa (  ) and 165 sample(s) ( A10-005-B_S188_MERGED.fastq.gz / A10-005-H_S189_MERGED.fastq.gz / A10-005-M_S190_MERGED.fastq.gz / A12-007_S191_MERGED.fastq.gz / A12-007-B_S2_MERGED.fastq.gz / A8-005_S4_MERGED.fastq.gz / A9-012_S5_MERGED.fastq.gz / AB29-ABMX-H_S6_MERGED.fastq.gz / AC29033_S8_MERGED.fastq.gz / AD26-005-B_S9_MERGED.fastq.gz / AD26-005-H_S10_MERGED.fastq.gz / AD30-ABMX-M_S12_MERGED.fastq.gz / AD32-007-M_S13_MERGED.fastq.gz / ADABM30X-H_S15_MERGED.fastq.gz / ADABM30X-M_S16_MERGED.fastq.gz / B17-014_S18_MERGED.fastq.gz / B18-006-B_S19_MERGED.fastq.gz / BA16-036bis_S20_MERGED.fastq.gz / BA17-050-B_S21_MERGED.fastq.gz / BB19-006-H_S22_MERGED.fastq.gz / BB6-019-B_S23_MERGED.fastq.gz / BB6-019-H_S24_MERGED.fastq.gz / BD14-021_S26_MERGED.fastq.gz / BE9-006-B_S27_MERGED.fastq.gz / BE9-006-H_S28_MERGED.fastq.gz / BE9-006-M_S29_MERGED.fastq.gz / BG7-010-B_S30_MERGED.fastq.gz / BG7-010-H_S31_MERGED.fastq.gz / BG7-010-M_S32_MERGED.fastq.gz / BH9-021_S33_MERGED.fastq.gz / BJ17-007-M_S34_MERGED.fastq.gz / BJ8-ABM-003_S35_MERGED.fastq.gz / BL7-006-B_S36_MERGED.fastq.gz / BL7-006-H_S37_MERGED.fastq.gz / BL7-006-M_S38_MERGED.fastq.gz / BN11-041_S39_MERGED.fastq.gz / BO8-002_S41_MERGED.fastq.gz / BO8-005_S42_MERGED.fastq.gz / BP11-001-B_S43_MERGED.fastq.gz / BP11-001-H_S44_MERGED.fastq.gz / BP11-001-M_S45_MERGED.fastq.gz / BP12-025-B_S46_MERGED.fastq.gz / BP14-006_S47_MERGED.fastq.gz / BQ3-019_S48_MERGED.fastq.gz / BQ4-018-B_S49_MERGED.fastq.gz / BQ4-018-H_S50_MERGED.fastq.gz / BQ9ABM-002_S52_MERGED.fastq.gz / BS14-006_S54_MERGED.fastq.gz / BT-006-M_S55_MERGED.fastq.gz / BT7-006_S56_MERGED.fastq.gz / BV11-002-B_S57_MERGED.fastq.gz / BV11-002-M_S59_MERGED.fastq.gz / C1-001_S61_MERGED.fastq.gz / C21-NV1-H_S63_MERGED.fastq.gz / C21-NV1-M_S64_MERGED.fastq.gz / CA12-024_S66_MERGED.fastq.gz / CA9-027_S67_MERGED.fastq.gz / CA9-X_S68_MERGED.fastq.gz / CB8-019-B_S69_MERGED.fastq.gz / CB8-019-H_S70_MERGED.fastq.gz / CB8-019-M_S71_MERGED.fastq.gz / CB9-013_S72_MERGED.fastq.gz / CC3-044_S73_MERGED.fastq.gz / CC8-003_S74_MERGED.fastq.gz / D18-003-B_S78_MERGED.fastq.gz / D18-003-H_S79_MERGED.fastq.gz / D18-003-M_S80_MERGED.fastq.gz / D22-NVABM1_S81_MERGED.fastq.gz / D61-010-B_S82_MERGED.fastq.gz / D9-027-H_S84_MERGED.fastq.gz / D9-027-M_S85_MERGED.fastq.gz / DBM-ABM-001_S86_MERGED.fastq.gz / DJ2-008-B_S87_MERGED.fastq.gz / DJ2-008-H_S88_MERGED.fastq.gz / DP4-ABM001_S90_MERGED.fastq.gz / DS1-ABM002-B_S91_MERGED.fastq.gz / DS1-ABM002-H_S92_MERGED.fastq.gz / DS1-ABM002-M_S93_MERGED.fastq.gz / DU3-045-B_S94_MERGED.fastq.gz / DW4-007_S95_MERGED.fastq.gz / DY5-004-B_S96_MERGED.fastq.gz / DY5-004-H_S97_MERGED.fastq.gz / DY5-004-M_S98_MERGED.fastq.gz / DZ6-ABM-001_S99_MERGED.fastq.gz / E9-009-B_S100_MERGED.fastq.gz / E9-009-H_S101_MERGED.fastq.gz / E9-009-M_S102_MERGED.fastq.gz / EA5-ABM-001_S103_MERGED.fastq.gz / EC2-013-B_S104_MERGED.fastq.gz / F6-ABM-001_S105_MERGED.fastq.gz / F7-015-M_S106_MERGED.fastq.gz / FOMES19-H_S108_MERGED.fastq.gz / FOMES19-M_S109_MERGED.fastq.gz / H10-018-M_S110_MERGED.fastq.gz / H24-NVABM1-H_S111_MERGED.fastq.gz / J18-004-B_S114_MERGED.fastq.gz / J18-004-H_S115_MERGED.fastq.gz / J18-004-M_S116_MERGED.fastq.gz / K18-002-H_S117_MERGED.fastq.gz / K26-NVABM1_S118_MERGED.fastq.gz / L19X-H_S120_MERGED.fastq.gz / L19X-M_S121_MERGED.fastq.gz / L23-002-B_S122_MERGED.fastq.gz / L23-002-H_S123_MERGED.fastq.gz / M22-001_S125_MERGED.fastq.gz / N19X-B_S126_MERGED.fastq.gz / N19X-H_S127_MERGED.fastq.gz / N19X-M_S128_MERGED.fastq.gz / N22-001-B_S129_MERGED.fastq.gz / N23-002-B_S130_MERGED.fastq.gz / N23-002-H_S131_MERGED.fastq.gz / N23-002-M_S132_MERGED.fastq.gz / N25-ABMX_S133_MERGED.fastq.gz / NVABM-0058_S134_MERGED.fastq.gz / NVABM-0163-H_S135_MERGED.fastq.gz / NVABM-0397_S138_MERGED.fastq.gz / NVABM0216_S136_MERGED.fastq.gz / NVABM0244-M_S137_MERGED.fastq.gz / O20-X-B_S139_MERGED.fastq.gz / O20-X-H_S140_MERGED.fastq.gz / O20-X-M_S141_MERGED.fastq.gz / O21-007-B_S142_MERGED.fastq.gz / O21-007-H_S143_MERGED.fastq.gz / O21-007-M_S144_MERGED.fastq.gz / O24-003-H_S146_MERGED.fastq.gz / O24-003-M_S147_MERGED.fastq.gz / O26-004-B_S148_MERGED.fastq.gz / O26-004-H_S149_MERGED.fastq.gz / O26-004-M_S150_MERGED.fastq.gz / O27-012_S151_MERGED.fastq.gz / O9-005-B_S152_MERGED.fastq.gz / P19-023-M_S153_MERGED.fastq.gz / P27-015-M_S154_MERGED.fastq.gz / P27-ABM001_S155_MERGED.fastq.gz / Q27-ABM003-B_S156_MERGED.fastq.gz / R25-ABMX_S157_MERGED.fastq.gz / R28-008-B_S158_MERGED.fastq.gz / R28-008-H_S159_MERGED.fastq.gz / R28-008-M_S160_MERGED.fastq.gz / T28-ABM602-B_S162_MERGED.fastq.gz / U27-ABM002_S163_MERGED.fastq.gz / W25-ABMX_S164_MERGED.fastq.gz / W26-001-B_S165_MERGED.fastq.gz / W26-001-H_S166_MERGED.fastq.gz / W26-001-M_S167_MERGED.fastq.gz / W30-006_S168_MERGED.fastq.gz / W9-025-M_S169_MERGED.fastq.gz / X24-009-B_S170_MERGED.fastq.gz / X24-009-H_S171_MERGED.fastq.gz / X24-009-M_S172_MERGED.fastq.gz / X24-010_S173_MERGED.fastq.gz / X29-004-B_S174_MERGED.fastq.gz / X29-004-H_S175_MERGED.fastq.gz / X29-004-M_S176_MERGED.fastq.gz / Y21-ABM484-H_S177_MERGED.fastq.gz / Y28-002-B_S178_MERGED.fastq.gz / Y28-002-H_S179_MERGED.fastq.gz / Y28-002-M_S180_MERGED.fastq.gz / Y29-007-B_S181_MERGED.fastq.gz / Y29-007-H_S182_MERGED.fastq.gz / Y29-007-M_S183_MERGED.fastq.gz / Y31-ABM484-B_S184_MERGED.fastq.gz / Z29-001-H_S185_MERGED.fastq.gz / Z30-002_S186_MERGED.fastq.gz / Z30-ABM560-M_S187_MERGED.fastq.gz ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of non-matching ASV 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of matching ASV 1420</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of filtered-out ASV 1413</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept ASV 7</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of kept samples 20</span>
<span class="r-plt img"><img src="plot_refseq_extremity_pq-6.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

