<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot ANCOMBC2 result for phyloseq object — plot_ancombc_pq • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot ANCOMBC2 result for phyloseq object — plot_ancombc_pq"><meta name="description" content="


Graphical representation of ANCOMBC2 result."><meta property="og:description" content="


Graphical representation of ANCOMBC2 result."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.13.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot ANCOMBC2 result for phyloseq object</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/beta_div_test.R" class="external-link"><code>R/beta_div_test.R</code></a></small>
      <div class="d-none name"><code>plot_ancombc_pq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-experimental-orange" alt="lifecycle-experimental"></a></p>

<p>Graphical representation of ANCOMBC2 result.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_ancombc_pq</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  <span class="va">ancombc_res</span>,</span>
<span>  filter_passed <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_diff <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_abs_lfc <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  tax_col <span class="op">=</span> <span class="st">"Genus"</span>,</span>
<span>  tax_label <span class="op">=</span> <span class="st">"Species"</span>,</span>
<span>  add_marginal_vioplot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_hline_cut_lfc <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-ancombc-res">ancombc_res<a class="anchor" aria-label="anchor" href="#arg-ancombc-res"></a></dt>
<dd><p>(required) the result of the ancombc_pq function
For the moment only bimodal factors are possible.</p></dd>


<dt id="arg-filter-passed">filter_passed<a class="anchor" aria-label="anchor" href="#arg-filter-passed"></a></dt>
<dd><p>(logical, default TRUE) Do we filter using the column
passed_ss? The passed_ss value is  TRUE if the taxon passed the sensitivity
analysis, i.e., adding different pseudo-counts to 0s would not change the results.</p></dd>


<dt id="arg-filter-diff">filter_diff<a class="anchor" aria-label="anchor" href="#arg-filter-diff"></a></dt>
<dd><p>(logical, default TRUE) Do we filter using the column
diff? The diff value is TRUE if the taxon is significant
(has q less than alpha)</p></dd>


<dt id="arg-min-abs-lfc">min_abs_lfc<a class="anchor" aria-label="anchor" href="#arg-min-abs-lfc"></a></dt>
<dd><p>(integer, default 0) Minimum absolute value to filter
results based on Log Fold Change. For ex. a value of 1 filter out taxa
for which the abundance in a given level of the modalty is not at least
the double of the abundance in the other level.</p></dd>


<dt id="arg-tax-col">tax_col<a class="anchor" aria-label="anchor" href="#arg-tax-col"></a></dt>
<dd><p>The taxonomic level (must be present in <code>tax_table</code> slot)
to color the points</p></dd>


<dt id="arg-tax-label">tax_label<a class="anchor" aria-label="anchor" href="#arg-tax-label"></a></dt>
<dd><p>The taxonomic level (must be present in <code>tax_table</code> slot)
to add label</p></dd>


<dt id="arg-add-marginal-vioplot">add_marginal_vioplot<a class="anchor" aria-label="anchor" href="#arg-add-marginal-vioplot"></a></dt>
<dd><p>(logical, default TRUE) Do we add a marginal
vioplot representing all the taxa lfc from ancombc_res.</p></dd>


<dt id="arg-add-label">add_label<a class="anchor" aria-label="anchor" href="#arg-add-label"></a></dt>
<dd><p>(logical, default TRUE) Do we add a label?</p></dd>


<dt id="arg-add-hline-cut-lfc">add_hline_cut_lfc<a class="anchor" aria-label="anchor" href="#arg-add-hline-cut-lfc"></a></dt>
<dd><p>(logical, default NULL) Do we add two horizontal
lines when min_abs_lfc is set (different from zero)?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot2 object. If add_marginal_vioplot is TRUE, this is a
patchworks of plot made using <code><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">patchwork::plot_layout()</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is mainly a wrapper of the work of others.
Please make a reference to <code><a href="https://rdrr.io/pkg/ANCOMBC/man/ancombc2.html" class="external-link">ANCOMBC::ancombc2()</a></code> if you
use this function.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"mia"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">data_fungi_mini</span><span class="op">@</span><span class="va">tax_table</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">data_fungi_mini</span><span class="op">@</span><span class="va">tax_table</span>,</span></span>
<span class="r-in"><span>    <span class="st">"taxon"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">data_fungi_mini</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="va">res_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="ancombc_pq.html">ancombc_pq</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">data_fungi_mini</span>,</span></span>
<span class="r-in"><span>    fact <span class="op">=</span> <span class="st">"Time"</span>,</span></span>
<span class="r-in"><span>    levels_fact <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"15"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    tax_level <span class="op">=</span> <span class="st">"taxon"</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">plot_ancombc_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, <span class="va">res_time</span>,</span></span>
<span class="r-in"><span>    filter_passed <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    tax_label <span class="op">=</span> <span class="st">"Genus"</span>, tax_col <span class="op">=</span> <span class="st">"Order"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">plot_ancombc_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, <span class="va">res_time</span>, tax_col <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">plot_ancombc_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, <span class="va">res_time</span>,</span></span>
<span class="r-in"><span>    filter_passed <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    filter_diff <span class="op">=</span> <span class="cn">FALSE</span>, tax_col <span class="op">=</span> <span class="st">"Family"</span>, add_label <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking the input data type ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The input data is of type: TreeSummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> PASS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking the sample metadata ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The specified variables in the formula: Time</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The available variables in the sample metadata: X, Sample_names, Tree_name, Sample_id, Height, Diameter, Time</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> PASS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking other arguments ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of groups of interest is: 2</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The group variable has &lt; 3 categories </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The sample size per group is: 0 = 30, 15 = 34</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> PASS</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The number of taxa used for estimating sample-specific biases is: 9</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> A large number of taxa (&gt;50) is required for the consistent estimation of biases</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Obtaining initial estimates ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Estimating sample-specific biases ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Estimation of sampling fractions failed for the following samples:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> BT-006-M_S55_MERGED.fastq.gz, C21-NV1-B_S62_MERGED.fastq.gz, CB8-019-H_S70_MERGED.fastq.gz, DY5-004-B_S96_MERGED.fastq.gz, DY5-004-H_S97_MERGED.fastq.gz, J18-004-M_S116_MERGED.fastq.gz, W26-001-B_S165_MERGED.fastq.gz, X24-009-B_S170_MERGED.fastq.gz</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> These samples may have an excessive number of zero values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Conducting sensitivity analysis for pseudo-count addition to 0s ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> For taxa that are significant but do not pass the sensitivity analysis,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> please flag them and proceed with caution, as they are likely false positives.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> For detailed instructions on performing sensitivity analysis,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> please refer to the package vignette.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ANCOM-BC2 primary results ...</span>
<span class="r-plt img"><img src="plot_ancombc_pq-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

