<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Recluster sequences of an object of class physeq or a list of DNA sequences — postcluster_pq • MiscMetabar</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Recluster sequences of an object of class physeq or a list of DNA sequences — postcluster_pq"><meta name="description" content="


This function use the merge_taxa_vec function to  merge taxa into clusters."><meta property="og:description" content="


This function use the merge_taxa_vec function to  merge taxa into clusters."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.14.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MiscMetabar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a></li>
    <li><a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a></li>
    <li><a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a></li>
    <li><a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a></li>
    <li><a class="dropdown-item" href="../articles/tree_visualization.html">Tree building and visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a></li>
    <li><a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bioinformatics actions</h6></li>
    <li><a class="dropdown-item" href="../articles/fastq_quality_check.html">Fastq quality check</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6></li>
    <li><a class="dropdown-item" href="../articles/tengeler.html">Tengeler dataset</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6></li>
    <li><a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>For developpers</h6></li>
    <li><a class="dropdown-item" href="../articles/Rules.html">Rules of code</a></li>
    <li><a class="dropdown-item" href="../articles/pkgnet-report.html">Report pkgnet</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Recluster sequences of an object of class <code>physeq</code> or a list of DNA sequences</h1>
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/R/dada_phyloseq.R" class="external-link"><code>R/dada_phyloseq.R</code></a></small>
      <div class="d-none name"><code>postcluster_pq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html#lifecycle">
<img src="https://img.shields.io/badge/lifecycle-maturing-blue" alt="lifecycle-maturing"></a></p>

<p>This function use the <code>merge_taxa_vec</code> function to  merge taxa into clusters.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">postcluster_pq</span><span class="op">(</span></span>
<span>  physeq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dna_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nproc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"clusterize"</span>,</span>
<span>  id <span class="op">=</span> <span class="fl">0.97</span>,</span>
<span>  vsearchpath <span class="op">=</span> <span class="st">"vsearch"</span>,</span>
<span>  tax_adjust <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  vsearch_cluster_method <span class="op">=</span> <span class="st">"--cluster_size"</span>,</span>
<span>  vsearch_args <span class="op">=</span> <span class="st">"--strand both"</span>,</span>
<span>  keep_temporary_files <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  swarmpath <span class="op">=</span> <span class="st">"swarm"</span>,</span>
<span>  d <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  swarm_args <span class="op">=</span> <span class="st">"--fastidious"</span>,</span>
<span>  method_clusterize <span class="op">=</span> <span class="st">"overlap"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">asv2otu</span><span class="op">(</span></span>
<span>  physeq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dna_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nproc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"clusterize"</span>,</span>
<span>  id <span class="op">=</span> <span class="fl">0.97</span>,</span>
<span>  vsearchpath <span class="op">=</span> <span class="st">"vsearch"</span>,</span>
<span>  tax_adjust <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  vsearch_cluster_method <span class="op">=</span> <span class="st">"--cluster_size"</span>,</span>
<span>  vsearch_args <span class="op">=</span> <span class="st">"--strand both"</span>,</span>
<span>  keep_temporary_files <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  swarmpath <span class="op">=</span> <span class="st">"swarm"</span>,</span>
<span>  d <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  swarm_args <span class="op">=</span> <span class="st">"--fastidious"</span>,</span>
<span>  method_clusterize <span class="op">=</span> <span class="st">"overlap"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>(required): a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object obtained
using the <code>phyloseq</code> package.</p></dd>


<dt id="arg-dna-seq">dna_seq<a class="anchor" aria-label="anchor" href="#arg-dna-seq"></a></dt>
<dd><p>You may directly use a character vector of DNA sequences
in place of physeq args. When physeq is set, dna sequences take the value of
<code>physeq@refseq</code></p></dd>


<dt id="arg-nproc">nproc<a class="anchor" aria-label="anchor" href="#arg-nproc"></a></dt>
<dd><p>(default: 1)
Set to number of cpus/processors to use for the clustering</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>(default: clusterize)
Set the clustering method.</p><ul><li><p><code>clusterize</code> use the <code><a href="https://rdrr.io/pkg/DECIPHER/man/Clusterize.html" class="external-link">DECIPHER::Clusterize()</a></code> fonction,</p></li>
<li><p><code>vsearch</code> use the vsearch software (https://github.com/torognes/vsearch)
with arguments <code>--cluster_size</code> by default (see args <code>vsearch_cluster_method</code>)
and <code>-strand both</code> (see args <code>vsearch_args</code>)</p></li>
<li><p><code>swarm</code> use the swarm</p></li>
</ul></dd>


<dt id="arg-id">id<a class="anchor" aria-label="anchor" href="#arg-id"></a></dt>
<dd><p>(default: 0.97) level of identity to cluster</p></dd>


<dt id="arg-vsearchpath">vsearchpath<a class="anchor" aria-label="anchor" href="#arg-vsearchpath"></a></dt>
<dd><p>(default: vsearch) path to vsearch</p></dd>


<dt id="arg-tax-adjust">tax_adjust<a class="anchor" aria-label="anchor" href="#arg-tax-adjust"></a></dt>
<dd><p>(Default 0) See the man page
of <code><a href="merge_taxa_vec.html">merge_taxa_vec()</a></code> for more details.
To conserved the taxonomic rank of the most abundant taxa (ASV, OTU,...),
set tax_adjust to 0 (default). For the moment only tax_adjust = 0 is
robust</p></dd>


<dt id="arg-vsearch-cluster-method">vsearch_cluster_method<a class="anchor" aria-label="anchor" href="#arg-vsearch-cluster-method"></a></dt>
<dd><p>(default: "–cluster_size) See other possible
methods in the <a href="https://github.com/torognes/vsearch/" class="external-link">vsearch manual</a> (e.g. <code>--cluster_size</code> or <code>--cluster_smallmem</code>)</p><ul><li><p><code>--cluster_fast</code> : Clusterize the fasta sequences in filename, automatically sort by decreasing sequence length beforehand.</p></li>
<li><p><code>--cluster_size</code> : Clusterize the fasta sequences in filename, automatically sort by decreasing sequence abundance beforehand.</p></li>
<li><p><code>--cluster_smallmem</code> : Clusterize the fasta sequences in filename without automatically modifying their order beforehand. Sequence are expected to be sorted by decreasing sequence length, unless <em>–usersort</em> is used.
In that case you may set <code>vsearch_args</code> to vsearch_args = "–strand both –usersort"</p></li>
</ul></dd>


<dt id="arg-vsearch-args">vsearch_args<a class="anchor" aria-label="anchor" href="#arg-vsearch-args"></a></dt>
<dd><p>(default : "–strand both") a one length character element defining other parameters to
passed on to vsearch.</p></dd>


<dt id="arg-keep-temporary-files">keep_temporary_files<a class="anchor" aria-label="anchor" href="#arg-keep-temporary-files"></a></dt>
<dd><p>(logical, default: FALSE) Do we keep temporary files</p><ul><li><p>temp.fasta (refseq in fasta or dna_seq sequences)</p></li>
<li><p>cluster.fasta (centroid if method = "vsearch")</p></li>
<li><p>temp.uc (clusters if method = "vsearch")</p></li>
</ul></dd>


<dt id="arg-swarmpath">swarmpath<a class="anchor" aria-label="anchor" href="#arg-swarmpath"></a></dt>
<dd><p>(default: swarm) path to swarm</p></dd>


<dt id="arg-d">d<a class="anchor" aria-label="anchor" href="#arg-d"></a></dt>
<dd><p>(default: 1) maximum number of differences allowed between two
amplicons, meaning that two amplicons will be grouped if they have <code>d</code>
(or less) differences</p></dd>


<dt id="arg-swarm-args">swarm_args<a class="anchor" aria-label="anchor" href="#arg-swarm-args"></a></dt>
<dd><p>(default : "–fastidious") a one length character
element defining other parameters to  passed on to swarm See other possible
methods in the <a href="https://github.com/torognes/swarm/blob/master/man/swarm_manual.pdf" class="external-link">SWARM pdf manual</a></p></dd>


<dt id="arg-method-clusterize">method_clusterize<a class="anchor" aria-label="anchor" href="#arg-method-clusterize"></a></dt>
<dd><p>(default "overlap") the method for the <code><a href="https://rdrr.io/pkg/DECIPHER/man/Clusterize.html" class="external-link">DECIPHER::Clusterize()</a></code> method</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed on to <code><a href="https://rdrr.io/pkg/DECIPHER/man/Clusterize.html" class="external-link">DECIPHER::Clusterize()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A new object of class <code>physeq</code> or a list of cluster if dna_seq
args was used.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function use the <code>merge_taxa_vec</code> function to
merge taxa into clusters. By default tax_adjust = 0. See the man page
of <code><a href="merge_taxa_vec.html">merge_taxa_vec()</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>VSEARCH can be downloaded from
<a href="https://github.com/torognes/vsearch" class="external-link">https://github.com/torognes/vsearch</a>.
More information in the associated publication
<a href="https://pubmed.ncbi.nlm.nih.gov/27781170" class="external-link">https://pubmed.ncbi.nlm.nih.gov/27781170</a>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="vsearch_clustering.html">vsearch_clustering()</a></code> and <code><a href="swarm_clustering.html">swarm_clustering()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Adrien Taudière</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"DECIPHER"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">postcluster_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Partitioning sequences by 3-mer similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by relatedness within 11 groups:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iteration 1 of up to 17 (100.0% stability) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.01 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clustering sequences by 9-mer similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.07 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clusters via relatedness sorting: 100% (0% exclusively)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clusters via rare 3-mers: 100% (0% exclusively)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Estimated clustering effectiveness: 100%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 32 taxa and 137 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 137 samples by 7 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 32 taxa by 12 taxonomic ranks ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> refseq()      DNAStringSet:      [ 32 reference sequences ]</span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"DECIPHER"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">postcluster_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, method_clusterize <span class="op">=</span> <span class="st">"longest"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="kw">if</span> <span class="op">(</span><span class="fu">MiscMetabar</span><span class="fu">::</span><span class="fu"><a href="is_swarm_installed.html">is_swarm_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">d_swarm</span> <span class="op">&lt;-</span> <span class="fu">postcluster_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, method <span class="op">=</span> <span class="st">"swarm"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">}</span></span></span>
<span class="r-in"><span>  <span class="kw">if</span> <span class="op">(</span><span class="fu">MiscMetabar</span><span class="fu">::</span><span class="fu"><a href="is_vsearch_installed.html">is_vsearch_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">d_vs</span> <span class="op">&lt;-</span> <span class="fu">postcluster_pq</span><span class="op">(</span><span class="va">data_fungi_mini</span>, method <span class="op">=</span> <span class="st">"vsearch"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">}</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Partitioning sequences by 3-mer similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by relatedness within 11 groups:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iteration 1 of up to 17 (100.0% stability) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.01 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clustering sequences by 9-mer similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.07 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clusters via relatedness sorting: 100% (0% exclusively)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clusters via rare 3-mers: 100% (0% exclusively)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Estimated clustering effectiveness: 100%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

