<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MiscMetabar">
<title>alpha-div â€¢ MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="alpha-div">
<meta property="og:description" content="MiscMetabar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.6.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Tutorials with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>alpha-div</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/alpha-div.Rmd" class="external-link"><code>vignettes/alpha-div.Rmd</code></a></small>
      <div class="d-none name"><code>alpha-div.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">MiscMetabar</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="alpha-diversity">Alpha diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h4>
<div class="section level5">
<h5 id="hill-number">Hill number<a class="anchor" aria-label="anchor" href="#hill-number"></a>
</h5>
<p>Numerous metrics of diversity exist. Hill numbers <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Hill MO. 1973. Diversity and evenness: a unifying
notation and its consequences. Ecology 54, 427-473.&lt;/p&gt;"><sup>1</sup></a> is a kind of general
framework for alpha diversity index.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">renyi_res</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/renyi.html" class="external-link">renyi</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">otu_table</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">renyi_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                       0     0.25      0.5         1         2</span></span>
<span><span class="co">#&gt; A10-005-B_S188_MERGED.fastq.gz 4.204693 3.615323 3.044244 2.0754183 1.1561862</span></span>
<span><span class="co">#&gt; A10-005-H_S189_MERGED.fastq.gz 4.248495 3.150337 2.063712 0.9938545 0.6464533</span></span>
<span><span class="co">#&gt; A10-005-M_S190_MERGED.fastq.gz 3.988984 3.607773 3.308361 2.9304156 2.5781603</span></span>
<span><span class="co">#&gt; A12-007_S191_MERGED.fastq.gz   5.036953 4.423283 3.751946 2.6309045 1.8468497</span></span>
<span><span class="co">#&gt; A12-007-B_S2_MERGED.fastq.gz   3.850148 3.349461 2.874269 2.1881145 1.6533262</span></span>
<span><span class="co">#&gt; A15-004_S3_MERGED.fastq.gz     4.025352 3.890614 3.747555 3.4533352 2.9618789</span></span>
<span><span class="co">#&gt;                                        4         8        16        32</span></span>
<span><span class="co">#&gt; A10-005-B_S188_MERGED.fastq.gz 0.8007816 0.6864775 0.6407124 0.6200442</span></span>
<span><span class="co">#&gt; A10-005-H_S189_MERGED.fastq.gz 0.5146170 0.4469617 0.4172228 0.4037640</span></span>
<span><span class="co">#&gt; A10-005-M_S190_MERGED.fastq.gz 2.3022639 2.1015018 1.9744485 1.9110311</span></span>
<span><span class="co">#&gt; A12-007_S191_MERGED.fastq.gz   1.5920328 1.4947479 1.4421368 1.4109287</span></span>
<span><span class="co">#&gt; A12-007-B_S2_MERGED.fastq.gz   1.3979355 1.2774402 1.2106906 1.1733049</span></span>
<span><span class="co">#&gt; A15-004_S3_MERGED.fastq.gz     2.4997916 2.2378349 2.1045646 2.0377518</span></span>
<span><span class="co">#&gt;                                       64       Inf</span></span>
<span><span class="co">#&gt; A10-005-B_S188_MERGED.fastq.gz 0.6102022 0.6006678</span></span>
<span><span class="co">#&gt; A10-005-H_S189_MERGED.fastq.gz 0.3973550 0.3911464</span></span>
<span><span class="co">#&gt; A10-005-M_S190_MERGED.fastq.gz 1.8806976 1.8513117</span></span>
<span><span class="co">#&gt; A12-007_S191_MERGED.fastq.gz   1.3916365 1.3700776</span></span>
<span><span class="co">#&gt; A12-007-B_S2_MERGED.fastq.gz   1.1547034 1.1366612</span></span>
<span><span class="co">#&gt; A15-004_S3_MERGED.fastq.gz     2.0054156 1.9740810</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="test-for-difference-in-diversity-hill-number">Test for difference in diversity (hill number)<a class="anchor" aria-label="anchor" href="#test-for-difference-in-diversity-hill-number"></a>
</h5>
<p>One way to keep into account for difference in the number of
sequences per samples is to use a Tukey test on a linear model with the
square roots of the number of sequence as the first explanatory variable
of the linear model <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;BÃ¡lint M et al.Â 2015. Relocation, high-latitude warming
and host genetic identity shape the foliar fungal microbiome of poplars.
Molecular Ecology 24, 235-248. &lt;a href="https://doi.org/10.1111/mec.13018" class="external-link uri"&gt;https://doi.org/10.1111/mec.13018&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">MiscMetabar</span><span class="fu">::</span><span class="fu"><a href="../reference/hill_pq.html">hill_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, variable <span class="op">=</span> <span class="st">"Height"</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">plot_Hill_0</span></span></code></pre></div>
<div class="figure">
<img src="alpha-div_files/figure-html/unnamed-chunk-3-1.png" alt="Hill number 1" width="700"><p class="caption">
Hill number 1
</p>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">plot_tuckey</span></span></code></pre></div>
<div class="figure">
<img src="alpha-div_files/figure-html/unnamed-chunk-4-1.png" alt="Result of the Tuckey post-hoc test" width="700"><p class="caption">
Result of the Tuckey post-hoc test
</p>
</div>
<p>See also the <a href="https://microbiome.github.io/tutorials/Alphadiversity.html" class="external-link">tutorial</a>
of the microbiome package for an alternative using the non-parametric <a href="https://www.rdocumentation.org/packages/dgof/versions/1.2/topics/ks.test" class="external-link">Kolmogorov-Smirnov
test</a> for two-group comparisons when there are no relevant
covariates.</p>
</div>
</div>
<div class="section level3">
<h3 id="alpha-diversity-using-package-microbiotaprocess">Alpha diversity using package <code>MicrobiotaProcess</code><a class="anchor" aria-label="anchor" href="#alpha-diversity-using-package-microbiotaprocess"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/YuLab-SMU/MicrobiotaProcess/" class="external-link">"MicrobiotaProcess"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/teunbrand/ggh4x" class="external-link">"ggh4x"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/clean_pq.html">clean_pq</a></span><span class="op">(</span><span class="fu"><a href="../reference/subset_samples_pq.html">subset_samples_pq</a></span><span class="op">(</span><span class="va">data_fungi</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">@</span><span class="va">sam_data</span><span class="op">$</span><span class="va">Height</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/as.MPSE.html" class="external-link">as.MPSE</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/mp_cal_alpha-methods.html" class="external-link">mp_cal_alpha</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/mp_plot_alpha-methods.html" class="external-link">mp_plot_alpha</a></span><span class="op">(</span>.group <span class="op">=</span> <span class="st">"Height"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(4, 4, 5, 5, 4, 4, 3, 6, 3, 4, 3, 2, 3, :</span></span>
<span><span class="co">#&gt; cannot compute exact p-value with ties</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(3, 4, 3, 5, 4, 6, 3, 5, 4, 4, 3, 3, 3, :</span></span>
<span><span class="co">#&gt; cannot compute exact p-value with ties</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(3, 4, 3, 5, 4, 6, 3, 5, 4, 4, 3, 3, 3, :</span></span>
<span><span class="co">#&gt; cannot compute exact p-value with ties</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(1.32966134885476, 1.242453324894,</span></span>
<span><span class="co">#&gt; 1.56071040904141, : cannot compute exact p-value with ties</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(0.867563228481461, 1.32966134885476,</span></span>
<span><span class="co">#&gt; 0.867563228481461, : cannot compute exact p-value with ties</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(0.867563228481461, 1.32966134885476,</span></span>
<span><span class="co">#&gt; 0.867563228481461, : cannot compute exact p-value with ties</span></span></code></pre></div>
<p><img src="alpha-div_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggh4x_0.2.7              MicrobiotaProcess_1.14.0 MiscMetabar_0.6.1       </span></span>
<span><span class="co">#&gt; [4] dplyr_1.1.4              dada2_1.30.0             Rcpp_1.0.12             </span></span>
<span><span class="co">#&gt; [7] ggplot2_3.4.4            phyloseq_1.46.0         </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] libcoin_1.0-10              RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [3] jsonlite_1.8.8              magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [5] TH.data_1.1-2               modeltools_0.2-23          </span></span>
<span><span class="co">#&gt;   [7] farver_2.1.1                rmarkdown_2.25             </span></span>
<span><span class="co">#&gt;   [9] fs_1.6.3                    zlibbioc_1.48.0            </span></span>
<span><span class="co">#&gt;  [11] ragg_1.2.7                  vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [13] multtest_2.58.0             memoise_2.0.1              </span></span>
<span><span class="co">#&gt;  [15] Rsamtools_2.18.0            RCurl_1.98-1.14            </span></span>
<span><span class="co">#&gt;  [17] ggtree_3.10.0               htmltools_0.5.7            </span></span>
<span><span class="co">#&gt;  [19] S4Arrays_1.2.0              Rhdf5lib_1.24.1            </span></span>
<span><span class="co">#&gt;  [21] gridGraphics_0.5-1          SparseArray_1.2.3          </span></span>
<span><span class="co">#&gt;  [23] rhdf5_2.46.1                sass_0.4.8                 </span></span>
<span><span class="co">#&gt;  [25] bslib_0.6.1                 desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [27] plyr_1.8.9                  sandwich_3.1-0             </span></span>
<span><span class="co">#&gt;  [29] zoo_1.8-12                  cachem_1.0.8               </span></span>
<span><span class="co">#&gt;  [31] GenomicAlignments_1.38.2    igraph_1.6.0               </span></span>
<span><span class="co">#&gt;  [33] lifecycle_1.0.4             iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [35] pkgconfig_2.0.3             Matrix_1.6-1.1             </span></span>
<span><span class="co">#&gt;  [37] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [39] GenomeInfoDbData_1.2.11     MatrixGenerics_1.14.0      </span></span>
<span><span class="co">#&gt;  [41] aplot_0.2.2                 digest_0.6.34              </span></span>
<span><span class="co">#&gt;  [43] ggnewscale_0.4.9            colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [45] ShortRead_1.60.0            patchwork_1.2.0            </span></span>
<span><span class="co">#&gt;  [47] S4Vectors_0.40.2            textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [49] GenomicRanges_1.54.1        hwriter_1.3.2.1            </span></span>
<span><span class="co">#&gt;  [51] vegan_2.6-4                 labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [53] fansi_1.0.6                 abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [55] mgcv_1.9-0                  compiler_4.3.2             </span></span>
<span><span class="co">#&gt;  [57] withr_3.0.0                 BiocParallel_1.36.0        </span></span>
<span><span class="co">#&gt;  [59] highr_0.10                  ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt;  [61] MASS_7.3-60                 DelayedArray_0.28.0        </span></span>
<span><span class="co">#&gt;  [63] biomformat_1.30.0           permute_0.9-7              </span></span>
<span><span class="co">#&gt;  [65] tools_4.3.2                 ape_5.7-1                  </span></span>
<span><span class="co">#&gt;  [67] glue_1.7.0                  nlme_3.1-163               </span></span>
<span><span class="co">#&gt;  [69] rhdf5filters_1.14.1         grid_4.3.2                 </span></span>
<span><span class="co">#&gt;  [71] cluster_2.1.4               reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [73] ade4_1.7-22                 generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [75] gtable_0.3.4                tidyr_1.3.0                </span></span>
<span><span class="co">#&gt;  [77] data.table_1.14.10          coin_1.4-3                 </span></span>
<span><span class="co">#&gt;  [79] utf8_1.2.4                  XVector_0.42.0             </span></span>
<span><span class="co">#&gt;  [81] BiocGenerics_0.48.1         ggrepel_0.9.5              </span></span>
<span><span class="co">#&gt;  [83] foreach_1.5.2               pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [85] stringr_1.5.1               yulab.utils_0.1.3          </span></span>
<span><span class="co">#&gt;  [87] splines_4.3.2               gghalves_0.1.4             </span></span>
<span><span class="co">#&gt;  [89] treeio_1.26.0               lattice_0.21-9             </span></span>
<span><span class="co">#&gt;  [91] survival_3.5-7              deldir_2.0-2               </span></span>
<span><span class="co">#&gt;  [93] tidyselect_1.2.0            Biostrings_2.70.1          </span></span>
<span><span class="co">#&gt;  [95] knitr_1.45                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [97] IRanges_2.36.0              SummarizedExperiment_1.32.0</span></span>
<span><span class="co">#&gt;  [99] ggtreeExtra_1.12.0          stats4_4.3.2               </span></span>
<span><span class="co">#&gt; [101] xfun_0.41                   Biobase_2.62.0             </span></span>
<span><span class="co">#&gt; [103] matrixStats_1.2.0           stringi_1.8.3              </span></span>
<span><span class="co">#&gt; [105] lazyeval_0.2.2              ggfun_0.1.4                </span></span>
<span><span class="co">#&gt; [107] yaml_2.3.8                  evaluate_0.23              </span></span>
<span><span class="co">#&gt; [109] codetools_0.2-19            interp_1.1-5               </span></span>
<span><span class="co">#&gt; [111] tibble_3.2.1                ggplotify_0.1.2            </span></span>
<span><span class="co">#&gt; [113] cli_3.6.2                   RcppParallel_5.1.7         </span></span>
<span><span class="co">#&gt; [115] systemfonts_1.0.5           munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [117] jquerylib_0.1.4             GenomeInfoDb_1.38.5        </span></span>
<span><span class="co">#&gt; [119] png_0.1-8                   parallel_4.3.2             </span></span>
<span><span class="co">#&gt; [121] pkgdown_2.0.7               latticeExtra_0.6-30        </span></span>
<span><span class="co">#&gt; [123] jpeg_0.1-10                 bitops_1.0-7               </span></span>
<span><span class="co">#&gt; [125] ggstar_1.0.4                mvtnorm_1.2-4              </span></span>
<span><span class="co">#&gt; [127] tidytree_0.4.6              scales_1.3.0               </span></span>
<span><span class="co">#&gt; [129] purrr_1.0.2                 crayon_1.5.2               </span></span>
<span><span class="co">#&gt; [131] rlang_1.1.3                 multcomp_1.4-25</span></span></code></pre></div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Adrien TaudiÃ¨re.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
