<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MiscMetabar">
<title>Introduction • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction">
<meta property="og:description" content="MiscMetabar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.7.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="MiscMetabar_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="MiscMetabar_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="MiscMetabar_files/datatables-binding-0.31/datatables.js"></script><link href="MiscMetabar_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="MiscMetabar_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="MiscMetabar_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="MiscMetabar_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="MiscMetabar_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><script src="MiscMetabar_files/d3-4.5.0/d3.min.js"></script><script src="MiscMetabar_files/sankey-1/sankey.js"></script><script src="MiscMetabar_files/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/MiscMetabar.Rmd" class="external-link"><code>vignettes/MiscMetabar.Rmd</code></a></small>
      <div class="d-none name"><code>MiscMetabar.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction-to-miscmetabar-an-r-package-to-facilitate-visualization-and-reproducibility-in-metabarcoding-analysis">Introduction to MiscMetabar: an R package to facilitate
visualization and reproducibility in metabarcoding analysis<a class="anchor" aria-label="anchor" href="#introduction-to-miscmetabar-an-r-package-to-facilitate-visualization-and-reproducibility-in-metabarcoding-analysis"></a>
</h2>
<div class="section level3">
<h3 id="raison-dêtre">Raison d’être<a class="anchor" aria-label="anchor" href="#raison-d%C3%AAtre"></a>
</h3>
<ul>
<li>Complete R packages dada2 and phyloseq</li>
<li>Useful visualizations (<code>biplot_pq</code>,
<code>circle_pq</code>, <code>upset_pq</code>,
<code>ggvenn_pq</code>)</li>
<li>Facilitate the use of targets package</li>
</ul>
</div>
<div class="section level3">
<h3 id="quick-overview">Quick overview<a class="anchor" aria-label="anchor" href="#quick-overview"></a>
</h3>
<p>For an introduction to metabarcoding in R, Please visite the <a href="https://adrientaudiere.github.io/MiscMetabar/articles/states_of_fields_in_R.html">state
of the field</a> vignettes. The <a href="https://adrientaudiere.github.io/MiscMetabar/articles/import_export_track.html">import,
export and track</a> vignette explains how import and export
<code>phyloseq</code> object. Its also show how to summarize useful
information (number of sequences, samples and clusters) accross
bioinformatic pipelines.</p>
<p>If you are interested in ecological metrics, see the vignettes
describing <a href="https://adrientaudiere.github.io/MiscMetabar/articles/alpha-div.html">alpha-diversity</a>
and <a href="https://adrientaudiere.github.io/MiscMetabar/articles/beta-div.html">beta-diversity</a>
analysis. The vignette <a href="https://adrientaudiere.github.io/MiscMetabar/articles/filter.html">filter
taxa and samples</a> describes some data-filtering processes using
MiscMetabar and the <a href="https://adrientaudiere.github.io/MiscMetabar/articles/Reclustering.html">reclustering</a>
tutorial introduces the different way of clustering already-clustered
OTU/ASV. The vignette <a href="https://adrientaudiere.github.io/MiscMetabar/articles/tengeler.html">tengeler</a>
explore the dataset from Tengeler et al. (2020) using some MiscMetabar
functions.</p>
<p>For developers, I also wrote a vignette describing som <a href="https://adrientaudiere.github.io/MiscMetabar/articles/Rules.html">rules
of codes</a>.</p>
<div class="section level4">
<h4 id="summarize-a-physeq-object">Summarize a physeq object<a class="anchor" aria-label="anchor" href="#summarize-a-physeq-object"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">"MiscMetabar"</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by</span></span>
<span><span class="co">#&gt; 'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">"phyloseq"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://magrittr.tidyverse.org" class="external-link">"magrittr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"data_fungi"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">data_fungi</span><span class="op">)</span></span></code></pre></div>
<p><img src="MiscMetabar_files/figure-html/example-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="create-an-interactive-table-of-the-tax_table">Create an interactive table of the tax_table<a class="anchor" aria-label="anchor" href="#create-an-interactive-table-of-the-tax_table"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"phyloseq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tax_datatable.html">tax_datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span></span>
<span>  <span class="va">GlobalPatterns</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">GlobalPatterns</span><span class="op">@</span><span class="va">otu_table</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">100000</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-40beb384b06145e18c64" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-40beb384b06145e18c64">{"x":{"filter":"none","vertical":false,"data":[["12812","317658","329744","471141","326977","64396","317182","549656","327536","340216","279599","235567","557211","263681","360229","536311","9514","94166","550960","323845","534609","339261","322235","469778","158660","331820","244304","471122","361496","248140","259569","298875","359171","192573","470973","316732","325407","171551","189047","180658","316587","98605"],["Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria","Bacteria"],["Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Fusobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Cyanobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Proteobacteria","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Bacteroidetes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes","Firmicutes"],["Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Actinobacteria","Fusobacteria","Chloroplast","Chloroplast","Chloroplast","Chloroplast","Nostocophycideae","Nostocophycideae","Synechococcophycideae","4C0d-2","Betaproteobacteria","Betaproteobacteria","Gammaproteobacteria","Gammaproteobacteria","Gammaproteobacteria","Gammaproteobacteria","Alphaproteobacteria","Flavobacteria","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Bacteroidia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Clostridia","Bacilli","Bacilli"],["koll13","Actinomycetales","Actinomycetales","Actinomycetales","Bifidobacteriales","Fusobacteriales","Stramenopiles","Stramenopiles","Stramenopiles","Stramenopiles","Nostocales","Nostocales","Synechococcales","YS2","Neisseriales","Neisseriales","Pasteurellales","Pasteurellales","Enterobacteriales","Oceanospirillales","Rhodobacterales",null,"Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Bacteroidales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Clostridiales","Lactobacillales","Lactobacillales"],[null,"ACK-M1","ACK-M1","Micrococcaceae","Bifidobacteriaceae","Fusobacteriaceae",null,null,null,null,"Nostocaceae","Nostocaceae","Synechococcaceae",null,"Neisseriaceae","Neisseriaceae","Pasteurellaceae","Pasteurellaceae","Enterobacteriaceae","Halomonadaceae","Rhodobacteraceae",null,"Bacteroidaceae","Bacteroidaceae","Bacteroidaceae","Bacteroidaceae","Bacteroidaceae","Prevotellaceae","Bacteroidaceae","Bacteroidaceae","Rikenellaceae","Clostridiaceae","ClostridialesFamilyXI.IncertaeSedis","Ruminococcaceae","Lachnospiraceae","Lachnospiraceae","Ruminococcaceae","Ruminococcaceae","Ruminococcaceae","Ruminococcaceae","Streptococcaceae","Streptococcaceae"],[null,null,null,"Rothia","Bifidobacterium","Fusobacterium",null,null,null,null,"Dolichospermum",null,"Prochlorococcus",null,"Neisseria","Kingella","Actinobacillus","Haemophilus","Providencia","CandidatusPortiera",null,null,"Bacteroides","Bacteroides","Bacteroides","Bacteroides","Bacteroides","Prevotella","Bacteroides","Bacteroides","Alistipes","Clostridium","Anaerococcus",null,"Ruminococcus","Lachnospira","Faecalibacterium","Faecalibacterium",null,null,"Streptococcus","Streptococcus"],[null,null,null,"Rothiamucilaginosa","Bifidobacteriumadolescentis",null,null,null,null,null,null,null,null,null,null,null,"Actinobacillusporcinus","Haemophilusparainfluenzae",null,null,null,null,"Bacteroidesuniformis","Bacteroidescoprophilus",null,null,null,"Prevotellamelaninogenica",null,"Bacteroidescaccae","Alistipesputredinis",null,"Anaerococcushydrogenalis",null,"Ruminococcustorques",null,null,"Faecalibacteriumprausnitzii",null,null,null,"Streptococcussanguinis"],[196550,234024,436539,104260,291577,155864,528629,2481214,143627,104199,927850,122697,169240,212089,556441,196355,158228,444142,384205,139264,177026,124393,169042,133795,443938,1001492,222365,146594,109007,149891,149352,177879,106772,179312,137132,101484,110044,215470,312161,152940,106399,203583]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Kingdom<\/th>\n      <th>Phylum<\/th>\n      <th>Class<\/th>\n      <th>Order<\/th>\n      <th>Family<\/th>\n      <th>Genus<\/th>\n      <th>Species<\/th>\n      <th>nb_seq<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":8},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Kingdom","targets":1},{"name":"Phylum","targets":2},{"name":"Class","targets":3},{"name":"Order","targets":4},{"name":"Family","targets":5},{"name":"Genus","targets":6},{"name":"Species","targets":7},{"name":"nb_seq","targets":8}],"order":[],"autoWidth":false,"orderClasses":false,"rowCallback":"function(row, data, displayNum, displayIndex, dataIndex) {\nvar value=data[8]; $(this.api().cell(row, 8).node()).css({'background':isNaN(parseFloat(value)) || value <= 101484.000000 ? '' : 'linear-gradient(90.000000deg, transparent ' + Math.max(2481214.000000 - value, 0)/2379730.000000 * 100 + '%, steelblue ' + Math.max(2481214.000000 - value, 0)/2379730.000000 * 100 + '%)','background-size':'100% 90%','background-repeat':'no-repeat','background-position':'center'});\n}"}},"evals":["options.rowCallback"],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="sankey-diagram-of-the-tax_table">Sankey diagram of the tax_table<a class="anchor" aria-label="anchor" href="#sankey-diagram-of-the-tax_table"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, <span class="va">GlobalPatterns</span><span class="op">@</span><span class="va">tax_table</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">==</span> <span class="st">"Archaea"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/sankey_pq.html">sankey_pq</a></span><span class="op">(</span><span class="va">gp</span>, taxa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sankeyNetwork html-widget html-fill-item" id="htmlwidget-ea100ea9851908182858" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-ea100ea9851908182858">{"x":{"links":{"source":[0,0,1,2,2,2,1,1,1,2,1,3,4,5,6,7,8,8,4,9,3,10,11,12,13,13,14,15,16,16,13,14,17,13,11,18,13,14],"target":[1,2,3,6,7,8,19,9,4,5,10,20,11,13,12,14,15,16,17,21,22,18,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38],"value":[106,102,27,7,27,24,8,8,57,44,3,7,37,44,7,27,12,12,20,5,18,1,34,7,10,1,25,2,5,6,3,1,20,16,3,1,12,1]},"nodes":{"name":["Archaea","Crenarchaeota","Euryarchaeota","C2","Thaumarchaeota","Thermoplasmata","Halobacteria","Methanobacteria","Methanomicrobia","SdNA","Thermoprotei","Cenarchaeales","Halobacteriales","E2","Methanobacteriales","Methanomicrobiales","Methanosarcinales","Nitrososphaerales","Sulfolobales","pMC2A209","B10","NRPJ","pGrfC26","Cenarchaeaceae","Halobacteriaceae","MarinegroupII","MarinegroupIII","Methanobacteriaceae","Methanocorpusculaceae","Methanosaetaceae","Methanosarcinaceae","MOB79","MSBL1","Nitrososphaeraceae","pMC2A33","SAGMAX","Sulfolobaceae","WCHD302","WSA2"],"group":["Archaea","Crenarchaeota","Euryarchaeota","C2","Thaumarchaeota","Thermoplasmata","Halobacteria","Methanobacteria","Methanomicrobia","SdNA","Thermoprotei","Cenarchaeales","Halobacteriales","E2","Methanobacteriales","Methanomicrobiales","Methanosarcinales","Nitrososphaerales","Sulfolobales","pMC2A209","B10","NRPJ","pGrfC26","Cenarchaeaceae","Halobacteriaceae","MarinegroupII","MarinegroupIII","Methanobacteriaceae","Methanocorpusculaceae","Methanosaetaceae","Methanosarcinaceae","MOB79","MSBL1","Nitrososphaeraceae","pMC2A33","SAGMAX","Sulfolobaceae","WCHD302","WSA2"]},"options":{"NodeID":"name","NodeGroup":"name","LinkGroup":null,"colourScale":"d3.scaleOrdinal(d3.schemeCategory20);","fontSize":7,"fontFamily":null,"nodeWidth":15,"nodePadding":10,"units":"OTUs","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"sinksRight":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="upset-plot-for-visualize-distribution-of-taxa-in-function-of-samples-variables">Upset plot for visualize distribution of taxa in function of samples
variables<a class="anchor" aria-label="anchor" href="#upset-plot-for-visualize-distribution-of-taxa-in-function-of-samples-variables"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/upset_pq.html">upset_pq</a></span><span class="op">(</span><span class="va">gp</span>, <span class="st">"SampleType"</span>, taxa <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span></span></code></pre></div>
<p><img src="MiscMetabar_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Tengeler, A.C., Dam, S.A., Wiesmann, M. et al. Gut microbiota from
persons with attention-deficit/hyperactivity disorder affects the brain
in mice. Microbiome 8, 44 (2020). <a href="https://doi.org/10.1186/s40168-020-00816-x" class="external-link uri">https://doi.org/10.1186/s40168-020-00816-x</a></p>
</div>
<div class="section level2">
<h2 id="session-inform">Session inform<a class="anchor" aria-label="anchor" href="#session-inform"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] magrittr_2.0.3    MiscMetabar_0.7.7 dplyr_1.1.4       dada2_1.30.0     </span></span>
<span><span class="co">#&gt; [5] Rcpp_1.0.12       ggplot2_3.4.4     phyloseq_1.46.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7                deldir_2.0-2               </span></span>
<span><span class="co">#&gt;   [3] permute_0.9-7               rlang_1.1.3                </span></span>
<span><span class="co">#&gt;   [5] ade4_1.7-22                 matrixStats_1.2.0          </span></span>
<span><span class="co">#&gt;   [7] compiler_4.3.2              mgcv_1.9-0                 </span></span>
<span><span class="co">#&gt;   [9] png_0.1-8                   systemfonts_1.0.5          </span></span>
<span><span class="co">#&gt;  [11] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [13] stringr_1.5.1               pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [15] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [17] ellipsis_0.3.2              XVector_0.42.0             </span></span>
<span><span class="co">#&gt;  [19] labeling_0.4.3              utf8_1.2.4                 </span></span>
<span><span class="co">#&gt;  [21] Rsamtools_2.18.0            rmarkdown_2.25             </span></span>
<span><span class="co">#&gt;  [23] ragg_1.2.7                  purrr_1.0.2                </span></span>
<span><span class="co">#&gt;  [25] xfun_0.42                   zlibbioc_1.48.0            </span></span>
<span><span class="co">#&gt;  [27] cachem_1.0.8                GenomeInfoDb_1.38.6        </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.8.8              biomformat_1.30.0          </span></span>
<span><span class="co">#&gt;  [31] highr_0.10                  rhdf5filters_1.14.1        </span></span>
<span><span class="co">#&gt;  [33] DelayedArray_0.28.0         Rhdf5lib_1.24.2            </span></span>
<span><span class="co">#&gt;  [35] BiocParallel_1.36.0         jpeg_0.1-10                </span></span>
<span><span class="co">#&gt;  [37] parallel_4.3.2              cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [39] R6_2.5.1                    bslib_0.6.1                </span></span>
<span><span class="co">#&gt;  [41] stringi_1.8.3               RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [43] ComplexUpset_1.3.3          GenomicRanges_1.54.1       </span></span>
<span><span class="co">#&gt;  [45] jquerylib_0.1.4             SummarizedExperiment_1.32.0</span></span>
<span><span class="co">#&gt;  [47] iterators_1.0.14            knitr_1.45                 </span></span>
<span><span class="co">#&gt;  [49] IRanges_2.36.0              Matrix_1.6-1.1             </span></span>
<span><span class="co">#&gt;  [51] splines_4.3.2               igraph_2.0.1.1             </span></span>
<span><span class="co">#&gt;  [53] tidyselect_1.2.0            abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [55] yaml_2.3.8                  vegan_2.6-4                </span></span>
<span><span class="co">#&gt;  [57] codetools_0.2-19            hwriter_1.3.2.1            </span></span>
<span><span class="co">#&gt;  [59] lattice_0.21-9              tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [61] plyr_1.8.9                  Biobase_2.62.0             </span></span>
<span><span class="co">#&gt;  [63] withr_3.0.0                 ShortRead_1.60.0           </span></span>
<span><span class="co">#&gt;  [65] evaluate_0.23               desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [67] survival_3.5-7              RcppParallel_5.1.7         </span></span>
<span><span class="co">#&gt;  [69] Biostrings_2.70.2           pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [71] MatrixGenerics_1.14.0       DT_0.31                    </span></span>
<span><span class="co">#&gt;  [73] foreach_1.5.2               stats4_4.3.2               </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3              RCurl_1.98-1.14            </span></span>
<span><span class="co">#&gt;  [77] S4Vectors_0.40.2            munsell_0.5.0              </span></span>
<span><span class="co">#&gt;  [79] scales_1.3.0                glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [81] tools_4.3.2                 interp_1.1-6               </span></span>
<span><span class="co">#&gt;  [83] data.table_1.15.0           GenomicAlignments_1.38.2   </span></span>
<span><span class="co">#&gt;  [85] fs_1.6.3                    rhdf5_2.46.1               </span></span>
<span><span class="co">#&gt;  [87] grid_4.3.2                  tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [89] ape_5.7-1                   crosstalk_1.2.1            </span></span>
<span><span class="co">#&gt;  [91] latticeExtra_0.6-30         colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [93] patchwork_1.2.0             networkD3_0.4              </span></span>
<span><span class="co">#&gt;  [95] nlme_3.1-163                GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">#&gt;  [97] cli_3.6.2                   textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [99] fansi_1.0.6                 S4Arrays_1.2.0             </span></span>
<span><span class="co">#&gt; [101] gtable_0.3.4                sass_0.4.8                 </span></span>
<span><span class="co">#&gt; [103] digest_0.6.34               BiocGenerics_0.48.1        </span></span>
<span><span class="co">#&gt; [105] SparseArray_1.2.4           htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt; [107] farver_2.1.1                memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [109] htmltools_0.5.7             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [111] multtest_2.58.0             lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [113] MASS_7.3-60</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
