<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MiscMetabar">
<title>Reclustering • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Reclustering">
<meta property="og:description" content="MiscMetabar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.6.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Tutorials with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Reclustering</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/Reclustering.Rmd" class="external-link"><code>vignettes/Reclustering.Rmd</code></a></small>
      <div class="d-none name"><code>Reclustering.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">MiscMetabar</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="re-clustering-asvs">Re-clustering ASVs<a class="anchor" aria-label="anchor" href="#re-clustering-asvs"></a>
</h2>
<p><strong>ASV</strong> (stands for <em>Amplicon Sequence Variant</em>;
also called <strong>ESV</strong> for Exact Amplicon Variant) is a DNA
sequence obtained from high-throughput analysis of marker genes.
<strong>OTU</strong> are a group of closely related individuals created
by clustering sequences based on a threshold of similarity. An ASV is a
special case of an OTU with a similarity threshold of 100%. A third
concept is the zero-radius OTU <strong>zOTU</strong> <span class="citation">(Edgar 2016)</span> which is the same concept than ASV
but compute with other softwares than <a href="https://github.com/benjjneb/dada2" class="external-link">dada</a> (e.g. <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a>).</p>
<p>The choice between ASV and OTU is important because they lead to
different results (<span class="citation">Joos et al. (2020)</span>, Box
2 in <span class="citation">Tedersoo et al. (2022)</span>, <span class="citation">Chiarello et al. (2022)</span>). Most articles
recommend making a choice depending on the question. For example, ASV
may be better than OTU for describing a group of very closely related
species. In addition, ASV are comparable across different datasets
(obtained using identical marker genes). On the other hand, <span class="citation">(Tedersoo et al. 2022)</span> report that ASV
approaches overestimate the richness of common fungal species (due to
haplotype variation), but underestimate the richness of rare species.
They therefore recommend the use of OTUs in metabarcoding analyses of
fungal communities. Finally, <span class="citation">(Kauserud
2023)</span> argues that the ASV term falls within the original OTU term
and recommends adopting only the OTU terms, but with a concise and clear
report on how the OTUs were generated.</p>
<p>Recent articles <span class="citation">(Forster et al. 2019; Antich
et al. 2021)</span> propose to use both approach together. They
recommend (i) using ASV to denoise the dataset and (ii) for some
questions, clustering the ASV sequences into OTUs. This is the goal of
the function <code><a href="../reference/asv2otu.html">asv2otu()</a></code>, using either the
<code><a href="https://rdrr.io/pkg/DECIPHER/man/Clusterize.html" class="external-link">DECIPHER::Clusterize</a></code> function from R or the <a href="https://github.com/torognes/vsearch" class="external-link">vsearch</a> software.</p>
<div class="section level3">
<h3 id="using-decipher-or-vsearch-algorithm">Using decipher or Vsearch algorithm<a class="anchor" aria-label="anchor" href="#using-decipher-or-vsearch-algorithm"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(data_fungi_sp_known)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>otu <span class="ot">&lt;-</span> <span class="fu">asv2otu</span>(data_fungi_sp_known, <span class="at">method =</span> <span class="st">"clusterize"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Partitioning sequences by 5-mer similarity:</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ================================================================================</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time difference of 0.15 secs</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Sorting by relatedness within 1 group:</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>iteration <span class="dv">110</span> of up to <span class="dv">325</span> (<span class="fl">100.0</span>% stability) </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time difference of 5.11 secs</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Clustering sequences by 8-mer similarity:</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ================================================================================</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time difference of 0.76 secs</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Clusters via relatedness sorting: 100% (0% exclusively)</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Clusters via rare 5-mers: 100% (0% exclusively)</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Estimated clustering effectiveness: 100%</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">otu_vs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/asv2otu.html">asv2otu</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span>, method <span class="op">=</span> <span class="st">"vsearch"</span><span class="op">)</span></span></code></pre></div>
<p>The vsearch method requires the installation of <a href="https://github.com/torognes/vsearch" class="external-link">Vsearch</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">otu</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="using-lulu-algorithm-link-to-lulu-article">Using lulu algorithm (<a href="https://www.nature.com/articles/s41467-017-01312-x" class="external-link">link to LULU
article</a>)<a class="anchor" aria-label="anchor" href="#using-lulu-algorithm-link-to-lulu-article"></a>
</h3>
<p>Another post-clustering transformation method is implemented in
<code><a href="../reference/lulu_pq.html">lulu_pq()</a></code>, which uses <span class="citation">Frøslev et al.
(2017)</span>’s method for curation of DNA amplicon data. The aim is
more to clean non-biological information than to make explicitly less
clusters.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span>
<span><span class="va">lulu_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lulu_pq.html">lulu_pq</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">data_fungi_sp_known</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">lulu_res</span><span class="op">$</span><span class="va">new_physeq</span><span class="op">)</span></span></code></pre></div>
<p><img src="Reclustering_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="tracking-number-of-samples-sequences-and-clusters">Tracking number of samples, sequences and clusters<a class="anchor" aria-label="anchor" href="#tracking-number-of-samples-sequences-and-clusters"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"Raw data"</span> <span class="op">=</span> <span class="va">data_fungi_sp_known</span>,</span>
<span>  <span class="st">"OTU"</span> <span class="op">=</span> <span class="va">otu</span>,</span>
<span>  <span class="st">"OTU_vsearch"</span> <span class="op">=</span> <span class="va">otu_vs</span>,</span>
<span>  <span class="st">"LULU"</span> <span class="op">=</span> <span class="va">lulu_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;             nb_sequences nb_clusters nb_samples</span></span>
<span><span class="co">#&gt; Raw data         1106581         651        185</span></span>
<span><span class="co">#&gt; OTU              1106581         360        185</span></span>
<span><span class="co">#&gt; OTU_vsearch      1106581         362        185</span></span>
<span><span class="co">#&gt; LULU             1106581         549        185</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] MiscMetabar_0.6.1 dplyr_1.1.4       dada2_1.30.0      Rcpp_1.0.12      </span></span>
<span><span class="co">#&gt; [5] ggplot2_3.4.4     phyloseq_1.46.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.1                   bitops_1.0-7               </span></span>
<span><span class="co">#&gt;   [3] pbapply_1.7-2               deldir_2.0-2               </span></span>
<span><span class="co">#&gt;   [5] permute_0.9-7               rlang_1.1.3                </span></span>
<span><span class="co">#&gt;   [7] magrittr_2.0.3              ade4_1.7-22                </span></span>
<span><span class="co">#&gt;   [9] RSQLite_2.3.5               matrixStats_1.2.0          </span></span>
<span><span class="co">#&gt;  [11] compiler_4.3.2              mgcv_1.9-0                 </span></span>
<span><span class="co">#&gt;  [13] png_0.1-8                   systemfonts_1.0.5          </span></span>
<span><span class="co">#&gt;  [15] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [17] stringr_1.5.1               pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [19] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [21] XVector_0.42.0              labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [23] utf8_1.2.4                  Rsamtools_2.18.0           </span></span>
<span><span class="co">#&gt;  [25] rmarkdown_2.25              ragg_1.2.7                 </span></span>
<span><span class="co">#&gt;  [27] bit_4.0.5                   purrr_1.0.2                </span></span>
<span><span class="co">#&gt;  [29] xfun_0.41                   zlibbioc_1.48.0            </span></span>
<span><span class="co">#&gt;  [31] cachem_1.0.8                GenomeInfoDb_1.38.5        </span></span>
<span><span class="co">#&gt;  [33] jsonlite_1.8.8              biomformat_1.30.0          </span></span>
<span><span class="co">#&gt;  [35] blob_1.2.4                  highr_0.10                 </span></span>
<span><span class="co">#&gt;  [37] rhdf5filters_1.14.1         DelayedArray_0.28.0        </span></span>
<span><span class="co">#&gt;  [39] Rhdf5lib_1.24.1             BiocParallel_1.36.0        </span></span>
<span><span class="co">#&gt;  [41] jpeg_0.1-10                 parallel_4.3.2             </span></span>
<span><span class="co">#&gt;  [43] cluster_2.1.4               R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [45] bslib_0.6.1                 stringi_1.8.3              </span></span>
<span><span class="co">#&gt;  [47] RColorBrewer_1.1-3          GenomicRanges_1.54.1       </span></span>
<span><span class="co">#&gt;  [49] jquerylib_0.1.4             SummarizedExperiment_1.32.0</span></span>
<span><span class="co">#&gt;  [51] iterators_1.0.14            knitr_1.45                 </span></span>
<span><span class="co">#&gt;  [53] DECIPHER_2.30.0             IRanges_2.36.0             </span></span>
<span><span class="co">#&gt;  [55] Matrix_1.6-1.1              splines_4.3.2              </span></span>
<span><span class="co">#&gt;  [57] igraph_1.6.0                tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;  [59] abind_1.4-5                 yaml_2.3.8                 </span></span>
<span><span class="co">#&gt;  [61] vegan_2.6-4                 codetools_0.2-19           </span></span>
<span><span class="co">#&gt;  [63] hwriter_1.3.2.1             lattice_0.21-9             </span></span>
<span><span class="co">#&gt;  [65] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [67] Biobase_2.62.0              withr_3.0.0                </span></span>
<span><span class="co">#&gt;  [69] ShortRead_1.60.0            evaluate_0.23              </span></span>
<span><span class="co">#&gt;  [71] desc_1.4.3                  survival_3.5-7             </span></span>
<span><span class="co">#&gt;  [73] RcppParallel_5.1.7          Biostrings_2.70.1          </span></span>
<span><span class="co">#&gt;  [75] pillar_1.9.0                MatrixGenerics_1.14.0      </span></span>
<span><span class="co">#&gt;  [77] foreach_1.5.2               stats4_4.3.2               </span></span>
<span><span class="co">#&gt;  [79] generics_0.1.3              RCurl_1.98-1.14            </span></span>
<span><span class="co">#&gt;  [81] S4Vectors_0.40.2            munsell_0.5.0              </span></span>
<span><span class="co">#&gt;  [83] scales_1.3.0                glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [85] tools_4.3.2                 interp_1.1-5               </span></span>
<span><span class="co">#&gt;  [87] data.table_1.14.10          GenomicAlignments_1.38.2   </span></span>
<span><span class="co">#&gt;  [89] fs_1.6.3                    rhdf5_2.46.1               </span></span>
<span><span class="co">#&gt;  [91] grid_4.3.2                  ape_5.7-1                  </span></span>
<span><span class="co">#&gt;  [93] latticeExtra_0.6-30         colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [95] nlme_3.1-163                GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">#&gt;  [97] cli_3.6.2                   textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [99] fansi_1.0.6                 S4Arrays_1.2.0             </span></span>
<span><span class="co">#&gt; [101] gtable_0.3.4                sass_0.4.8                 </span></span>
<span><span class="co">#&gt; [103] digest_0.6.34               BiocGenerics_0.48.1        </span></span>
<span><span class="co">#&gt; [105] SparseArray_1.2.3           farver_2.1.1               </span></span>
<span><span class="co">#&gt; [107] memoise_2.0.1               htmltools_0.5.7            </span></span>
<span><span class="co">#&gt; [109] pkgdown_2.0.7               multtest_2.58.0            </span></span>
<span><span class="co">#&gt; [111] lifecycle_1.0.4             bit64_4.0.5                </span></span>
<span><span class="co">#&gt; [113] MASS_7.3-60</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-antich2021" class="csl-entry">
Antich, Adrià, Creu Palacin, Owen S Wangensteen, and Xavier Turon. 2021.
<span>“To Denoise or to Cluster, That Is Not the Question: Optimizing
Pipelines for COI Metabarcoding and Metaphylogeography.”</span> <em>BMC
Bioinformatics</em> 22: 1–24. <a href="https://doi.org/10.1101/2021.01.08.425760" class="external-link">https://doi.org/10.1101/2021.01.08.425760</a>.
</div>
<div id="ref-chiarello2022" class="csl-entry">
Chiarello, Marlène, Mark McCauley, Sébastien Villéger, and Colin R
Jackson. 2022. <span>“Ranking the Biases: The Choice of OTUs Vs. ASVs in
16S rRNA Amplicon Data Analysis Has Stronger Effects on Diversity
Measures Than Rarefaction and OTU Identity Threshold.”</span> <em>PLoS
One</em> 17 (2): e0264443. <a href="https://doi.org/10.1371/journal.pone.0264443" class="external-link">https://doi.org/10.1371/journal.pone.0264443</a>.
</div>
<div id="ref-edgar2016" class="csl-entry">
Edgar, Robert C. 2016. <span>“UNOISE2: Improved Error-Correction for
Illumina 16S and ITS Amplicon Sequencing.”</span> <em>BioRxiv</em>,
081257.
</div>
<div id="ref-forster2019" class="csl-entry">
Forster, Dominik, Guillaume Lentendu, Sabine Filker, Elyssa Dubois,
Thomas A Wilding, and Thorsten Stoeck. 2019. <span>“Improving eDNA-Based
Protist Diversity Assessments Using Networks of Amplicon Sequence
Variants.”</span> <em>Environmental Microbiology</em> 21 (11): 4109–24.
<a href="https://doi.org/10.1111/1462-2920.14764" class="external-link">https://doi.org/10.1111/1462-2920.14764</a>.
</div>
<div id="ref-froslev2017" class="csl-entry">
Frøslev, Tobias Guldberg, Rasmus Kjøller, Hans Henrik Bruun, Rasmus
Ejrnæs, Ane Kirstine Brunbjerg, Carlotta Pietroni, and Anders Johannes
Hansen. 2017. <span>“Algorithm for Post-Clustering Curation of DNA
Amplicon Data Yields Reliable Biodiversity Estimates.”</span> <em>Nature
Communications</em> 8 (1): 1188. <a href="https://doi.org/10.1038/s41467-017-01312-x" class="external-link">https://doi.org/10.1038/s41467-017-01312-x</a>.
</div>
<div id="ref-joos2020" class="csl-entry">
Joos, Lisa, Stien Beirinckx, Annelies Haegeman, Jane Debode, Bart
Vandecasteele, Steve Baeyen, Sofie Goormachtig, Lieven Clement, and
Caroline De Tender. 2020. <span>“Daring to Be Differential:
Metabarcoding Analysis of Soil and Plant-Related Microbial Communities
Using Amplicon Sequence Variants and Operational Taxonomical
Units.”</span> <em>BMC Genomics</em> 21 (1): 1–17. <a href="https://doi.org/10.1186/s12864-020-07126-4" class="external-link">https://doi.org/10.1186/s12864-020-07126-4</a>.
</div>
<div id="ref-kauserud2023" class="csl-entry">
Kauserud, Håvard. 2023. <span>“ITS Alchemy: On the Use of ITS as a DNA
Marker in Fungal Ecology.”</span> <em>Fungal Ecology</em>, 101274. <a href="https://doi.org/10.1016/j.funeco.2023.101274" class="external-link">https://doi.org/10.1016/j.funeco.2023.101274</a>.
</div>
<div id="ref-tedersoo2022" class="csl-entry">
Tedersoo, Leho, Mohammad Bahram, Lucie Zinger, R Henrik Nilsson, Peter G
Kennedy, Teng Yang, Sten Anslan, and Vladimir Mikryukov. 2022.
<span>“Best Practices in Metabarcoding of Fungi: From Experimental
Design to Results.”</span> <em>Molecular Ecology</em> 31 (10): 2769–95.
<a href="https://doi.org/10.22541/au.163430390.04226544/v1" class="external-link">https://doi.org/10.22541/au.163430390.04226544/v1</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
