<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MiscMetabar">
<title>Metabarcoding with R • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Metabarcoding with R">
<meta property="og:description" content="MiscMetabar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.7.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Metabarcoding with R</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/articles/states_of_fields_in_R.Rmd" class="external-link"><code>vignettes/articles/states_of_fields_in_R.Rmd</code></a></small>
      <div class="d-none name"><code>states_of_fields_in_R.Rmd</code></div>
    </div>

    
    
<p>This is a short introduction to other R packages in the field of
metabarcoding analysis.</p>
<div class="section level2">
<h2 id="state-of-the-field-in-r">State of the Field in R<a class="anchor" aria-label="anchor" href="#state-of-the-field-in-r"></a>
</h2>
<p>The metabarcoding ecosystem in the R language is mature,
well-constructed, and relies on a very active community in both the <a href="https://www.bioconductor.org/" class="external-link">bioconductor</a> and <a href="https://cran.r-project.org/" class="external-link">cran</a> projects. The <a href="https://www.bioconductor.org/" class="external-link">bioconductor</a> even creates
specific task views in <a href="https://bioconductor.org/packages/release/BiocViews.html#___Metagenomics" class="external-link">Metagenomics</a>
and <a href="https://bioconductor.org/packages/release/BiocViews.html#___Microbiome" class="external-link">Microbiome</a>.</p>
<p>R package <a href="https://bioconductor.org/packages/release/bioc/html/dada2.html" class="external-link"><code>dada2</code></a>
<span class="citation">(Callahan et al. 2016)</span> provides a highly
cited and recommended clustering method <span class="citation">(Pauvert
et al. 2019)</span>. <code>dada2</code> also provides tools to complete
the metabarcoding analysis pipeline, including chimera detection and
taxonomic assignment. <code>phyloseq</code> <span class="citation">(McMurdie and Holmes 2013)</span> (<a href="https://bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link uri">https://bioconductor.org/packages/release/bioc/html/phyloseq.html</a>)
facilitate metagenomics analysis by providing a way to store data (the
<code>phyloseq</code> class) and both graphical and statistical
functions.</p>
<p>The phyloseq package introduces the S4 class object (class physeq),
which contains (i) an OTU sample matrix, (ii) a taxonomic table, (iii) a
sample metadata table, and two optional slots for (iv) a phylogenetic
tree and (v) reference sequences.</p>
<p>Some packages already extend the phyloseq packages. For example, the
<a href="https://microbiome.github.io/" class="external-link"><code>microbiome</code></a>
package collection <span class="citation">(Ernst et al. 2023)</span>
provides some scripts and functions for manipulating microbiome
datasets.The <code>speedyseq</code> package <span class="citation">(McLaren 2020)</span> provides faster versions of
phyloseq’s plotting and taxonomic merging functions, some of which
([merge_samples2()] and [merge_taxa_vec()]) are integrated in
<code>MiscMetabar</code> (thanks to Mike. R. McLaren). The <a href="https://schuyler-smith.github.io/phylosmith/" class="external-link">phylosmith</a> <a href="https://joss.theoj.org/papers/10.21105/joss.01442" class="external-link"><span class="citation">Smith (2023)</span></a> package already provides some
functions to extend and simplify the use of the phyloseq packages.</p>
<p>Other packages (<a href="https://github.com/microbiome/mia/" class="external-link"><code>mia</code></a> forming
the <a href="https://microbiome.github.io/" class="external-link"><code>microbiome</code></a>
package collection and <a href="https://github.com/YuLab-SMU/MicrobiotaProcess" class="external-link"><code>MicrobiotaProcess</code></a>
<span class="citation">(Xu et al. 2023)</span>) extend a new data
structure using the comprehensive Bioconductor ecosystem of the
<code>SummarizedExperiment</code> family.</p>
<p><code>MiscMetabar</code> enriches this R ecosystem by providing
functions to (i) <strong>describe</strong> your dataset visually, (ii)
<strong>transform</strong> your data, (iii) <strong>explore</strong>
biological diversity (alpha, beta, and taxonomic diversity), and (iv)
simplify <strong>reproducibility</strong>. <code>MiscMetabar</code> is
designed to complement and not compete with other R packages mentioned
above. For example. The <code>mia</code> package is recommended for
studies focusing on phylogenetic trees, and <code>phylosmith</code>
allows easy visualization of co-occurrence networks. Using the
<code><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/as.MPSE.html" class="external-link">MicrobiotaProcess::as.MPSE</a></code> function, most of the utilities
in the <code>MicrobiotaProcess</code> package are available with
functions from the <code>MiscMetabar</code>.</p>
<p>I do not try to reinvent the wheel and prefer to rely on existing
packages and classes rather than building a new framework.
<code>MiscMetabar</code> is based on the phyloseq class from phyloseq,
the most cited package in metagenomics <span class="citation">(Wen et
al. 2023)</span>. For a description and comparison of these integrated
packages competing with phyloseq (e.g. <a href="https://github.com/ChiLiubio/microeco" class="external-link">microeco</a> by <span class="citation">C. Liu et al. (2020)</span>, <a href="https://github.com/YongxinLiu/EasyAmplicon" class="external-link">EasyAmplicon</a> by
<span class="citation">Y.-X. Liu et al. (2023)</span> and <a href="https://www.microbiomeanalyst.ca" class="external-link">MicrobiomeAnalystR</a> by <span class="citation">Lu et al. (2023)</span>) see <span class="citation">Wen
et al. (2023)</span>. Note that some limitations of the phyloseq
packages are circumvented thanks to <a href="https://schuyler-smith.github.io/phylosmith/" class="external-link">phylosmith</a> <span class="citation">(Smith 2023)</span>, <a href="https://david-barnett.github.io/microViz/" class="external-link"><code>microViz</code></a>
(<span class="citation">(Barnett, Arts, and Penders 2021)</span>) and <a href="https://adrientaudiere.github.io/MiscMetabar/"><code>MiscMetabar</code></a>.</p>
<p>Some packages provide an interactive interface useful for rapid
exploration and for code-beginner biologists. <a href="https://github.com/wejlab/animalcules" class="external-link">Animalcules</a> <span class="citation">(Zhao et al. 2021)</span> and <a href="https://david-barnett.github.io/microViz/" class="external-link"><code>microViz</code></a>
<span class="citation">(Barnett, Arts, and Penders 2021)</span> provides
shiny interactive interface whereas <a href="https://www.microbiomeanalyst.ca" class="external-link">MicrobiomeAnalystR</a> <span class="citation">(Lu et al. 2023)</span> is a web-based platform.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] vctrs_0.6.5       cli_3.6.2         knitr_1.45        rlang_1.1.3      </span></span>
<span><span class="co">##  [5] xfun_0.41         stringi_1.8.3     purrr_1.0.2       textshaping_0.3.7</span></span>
<span><span class="co">##  [9] jsonlite_1.8.8    glue_1.7.0        htmltools_0.5.7   ragg_1.2.7       </span></span>
<span><span class="co">## [13] sass_0.4.8        rmarkdown_2.25    evaluate_0.23     jquerylib_0.1.4  </span></span>
<span><span class="co">## [17] fastmap_1.1.1     yaml_2.3.8        lifecycle_1.0.4   memoise_2.0.1    </span></span>
<span><span class="co">## [21] stringr_1.5.1     compiler_4.3.2    fs_1.6.3          systemfonts_1.0.5</span></span>
<span><span class="co">## [25] digest_0.6.34     R6_2.5.1          magrittr_2.0.3    bslib_0.6.1      </span></span>
<span><span class="co">## [29] tools_4.3.2       pkgdown_2.0.7     cachem_1.0.8      desc_1.4.3</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Barnett2021" class="csl-entry">
Barnett, David J. M., Ilja C. W. Arts, and John Penders. 2021.
<span>“microViz: An r Package for Microbiome Data Visualization and
Statistics.”</span> <em>Journal of Open Source Software</em> 6 (63):
3201. <a href="https://doi.org/10.21105/joss.03201" class="external-link">https://doi.org/10.21105/joss.03201</a>.
</div>
<div id="ref-callahan2016" class="csl-entry">
Callahan, Benjamin J, Paul J McMurdie, Michael J Rosen, Andrew W Han,
Amy Jo A Johnson, and Susan P Holmes. 2016. <span>“DADA2:
High-Resolution Sample Inference from Illumina Amplicon Data.”</span>
<em>Nature Methods</em> 13 (7): 581–83. <a href="https://doi.org/10.1038/nmeth.3869" class="external-link">https://doi.org/10.1038/nmeth.3869</a>.
</div>
<div id="ref-ernst2023" class="csl-entry">
Ernst, Felix G. M., Sudarshan A. Shetty, Tuomas Borman, and Leo Lahti.
2023. <em>Mia: Microbiome Analysis</em>. <a href="https://doi.org/10.18129/B9.bioc.mia" class="external-link">https://doi.org/10.18129/B9.bioc.mia</a>.
</div>
<div id="ref-liu2020" class="csl-entry">
Liu, Chi, Yaoming Cui, Xiangzhen Li, and Minjie Yao. 2020. <span>“<span class="nocase">microeco: an R package for data mining in microbial
community ecology</span>.”</span> <em>FEMS Microbiology Ecology</em> 97
(2): fiaa255. <a href="https://doi.org/10.1093/femsec/fiaa255" class="external-link">https://doi.org/10.1093/femsec/fiaa255</a>.
</div>
<div id="ref-liu2023" class="csl-entry">
Liu, Yong-Xin, Lei Chen, Tengfei Ma, Xiaofang Li, Maosheng Zheng, Xin
Zhou, Liang Chen, et al. 2023. <span>“EasyAmplicon: An Easy-to-Use,
Open-Source, Reproducible, and Community-Based Pipeline for Amplicon
Data Analysis in Microbiome Research.”</span> <em>iMeta</em> 2 (1): e83.
</div>
<div id="ref-lu2023" class="csl-entry">
Lu, Yao, Guangyan Zhou, Jessica Ewald, Zhiqiang Pang, Tanisha Shiri, and
Jianguo Xia. 2023. <span>“<span class="nocase">MicrobiomeAnalyst 2.0:
comprehensive statistical, functional and integrative analysis of
microbiome data</span>.”</span> <em>Nucleic Acids Research</em> 51 (W1):
W310–18. <a href="https://doi.org/10.1093/nar/gkad407" class="external-link">https://doi.org/10.1093/nar/gkad407</a>.
</div>
<div id="ref-mclaren2020" class="csl-entry">
McLaren, Michael. 2020. <em>Mikemc/Speedyseq: Speedyseq V0.2.0</em>
(version v0.2.0). Zenodo. <a href="https://doi.org/10.5281/zenodo.3923184" class="external-link">https://doi.org/10.5281/zenodo.3923184</a>.
</div>
<div id="ref-mcmurdie2013" class="csl-entry">
McMurdie, Paul J., and Susan Holmes. 2013. <span>“Phyloseq: An r Package
for Reproducible Interactive Analysis and Graphics of Microbiome Census
Data.”</span> <em>PLoS ONE</em> 8 (4): e61217. <a href="https://doi.org/10.1371/journal.pone.0061217" class="external-link">https://doi.org/10.1371/journal.pone.0061217</a>.
</div>
<div id="ref-pauvert2019" class="csl-entry">
Pauvert, Charlie, Marc Buée, Valérie Laval, Véronique Edel-Hermann,
Laure Fauchery, Angélique Gautier, Isabelle Lesur, Jessica Vallance, and
Corinne Vacher. 2019. <span>“Bioinformatics Matters: The Accuracy of
Plant and Soil Fungal Community Data Is Highly Dependent on the
Metabarcoding Pipeline.”</span> <em>Fungal Ecology</em> 41: 23–33. <a href="https://doi.org/10.1016/j.funeco.2019.03.005" class="external-link">https://doi.org/10.1016/j.funeco.2019.03.005</a>.
</div>
<div id="ref-smith2023" class="csl-entry">
Smith, Schuyler. 2023. <em>Phylosmith: Functions to Help Analyze Data as
Phyloseq Objects</em>. <a href="https://schuyler-smith.github.io/phylosmith/" class="external-link">https://schuyler-smith.github.io/phylosmith/</a>.
</div>
<div id="ref-wen2023" class="csl-entry">
Wen, Tao, Guoqing Niu, Tong Chen, Qirong Shen, Jun Yuan, and Yong-Xin
Liu. 2023. <span>“The Best Practice for Microbiome Analysis Using
r.”</span> <em>Protein &amp; Cell</em>, pwad024.
</div>
<div id="ref-xu2023" class="csl-entry">
Xu, Shuangbin, Li Zhan, Wenli Tang, Qianwen Wang, Zehan Dai, Lang Zhou,
Tingze Feng, et al. 2023. <span>“MicrobiotaProcess: A Comprehensive r
Package for Deep Mining Microbiome.”</span> <em>The Innovation</em> 4
(2). <a href="https://doi.org/10.1016/j.xinn.2023.100388" class="external-link">https://doi.org/10.1016/j.xinn.2023.100388</a>.
</div>
<div id="ref-zhao2021" class="csl-entry">
Zhao, Yue, Anthony Federico, Tyler Faits, Solaiappan Manimaran, Daniel
Segrè, Stefano Monti, and W Evan Johnson. 2021. <span>“Animalcules:
Interactive Microbiome Analytics and Visualization in r.”</span>
<em>Microbiome</em> 9 (1): 1–16. <a href="https://doi.org/10.21203/rs.3.rs-29649/v2" class="external-link">https://doi.org/10.21203/rs.3.rs-29649/v2</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
