<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MiscMetabar">
<title>tengeler • MiscMetabar</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="tengeler">
<meta property="og:description" content="MiscMetabar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MiscMetabar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/MiscMetabar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/import_export_track.html">Import/export and track</a>
    <a class="dropdown-item" href="../articles/explore_data.html">Explore phyloseq data</a>
    <a class="dropdown-item" href="../articles/filter.html">Filter taxa and samples</a>
    <a class="dropdown-item" href="../articles/Reclustering.html">Reclustering</a>
    <a class="dropdown-item" href="../articles/tree_visualization.html">Tree visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Diversity analysis</h6>
    <a class="dropdown-item" href="../articles/alpha-div.html">Alpha-diversity</a>
    <a class="dropdown-item" href="../articles/beta-div.html">Beta-diversity</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Examples with published dataset</h6>
    <a class="dropdown-item" href="../articles/tengeler.html">Tengeler</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>R ecosystem for metabarcoding</h6>
    <a class="dropdown-item" href="../articles/states_of_fields_in_R.html">Metabarcoding with R</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>For developpers</h6>
    <a class="dropdown-item" href="../articles/Rules.html">Rules of code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/adrientaudiere/MiscMetabar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>tengeler</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/adrientaudiere/MiscMetabar/blob/HEAD/vignettes/articles/tengeler.Rmd" class="external-link"><code>vignettes/articles/tengeler.Rmd</code></a></small>
      <div class="d-none name"><code>tengeler.Rmd</code></div>
    </div>

    
    
<p>This tutorial explore a phyloseq version of the dataset from Tengeler
et al. (2020) available in the <code>mia</code> package.</p>
<div class="section level2">
<h2 id="load-library">Load library<a class="anchor" aria-label="anchor" href="#load-library"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/YuLab-SMU/MicrobiotaProcess/" class="external-link">"MicrobiotaProcess"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/adrientaudiere/MiscMetabar" class="external-link">"MiscMetabar"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://chao.stat.nthu.edu.tw/wordpress/software_download/" class="external-link">"iNEXT"</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">Tengeler2020</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="import-dataset-in-phyloseq-format">Import dataset in phyloseq format<a class="anchor" aria-label="anchor" href="#import-dataset-in-phyloseq-format"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Tengeler2020_pq</span><span class="op">)</span></span>
<span><span class="va">ten</span> <span class="op">&lt;-</span> <span class="va">Tengeler2020_pq</span></span>
<span><span class="fu"><a href="../reference/summary_plot_pq.html">summary_plot_pq</a></span><span class="op">(</span><span class="va">ten</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-4-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="alpha-diversity-analysis">Alpha-diversity analysis<a class="anchor" aria-label="anchor" href="#alpha-diversity-analysis"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hill_pq.html">hill_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"patient_status"</span>, one_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-5-1.png" width="960"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_inext</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/iNEXT_pq.html">iNEXT_pq</a></span><span class="op">(</span><span class="va">ten</span>,</span>
<span>    datatype <span class="op">=</span> <span class="st">"abundance"</span>,</span>
<span>    merge_sample_by <span class="op">=</span> <span class="st">"patient_status_vs_cohort"</span>,</span>
<span>    nboot <span class="op">=</span> <span class="fl">5</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/iNEXT/man/ggiNEXT.html" class="external-link">ggiNEXT</a></span><span class="op">(</span><span class="va">res_inext</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-6-1.png" width="960"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/accu_plot.html">accu_plot</a></span><span class="op">(</span></span>
<span>  <span class="va">ten</span>,</span>
<span>  fact <span class="op">=</span> <span class="st">"sample_name"</span>,</span>
<span>  add_nb_seq <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  by.fact <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.8</span>, <span class="fl">.6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 9</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 17</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 25</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 7</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 8</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 6</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 14</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 28</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 7</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 2</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 15</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 5</span></span>
<span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 5</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 6</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 4</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 6</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 3</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 21</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 5</span></span>
<span><span class="co">#&gt; Warning in vegan::rarefy(as.matrix(unclass(x[i, ])), n, se = TRUE): most</span></span>
<span><span class="co">#&gt; observed count data have counts 1, but smallest count is 13</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values (`geom_line()`).</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-7-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="explore-taxonomy">Explore taxonomy<a class="anchor" aria-label="anchor" href="#explore-taxonomy"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://grunwaldlab.github.io/metacoder_documentation/" class="external-link">metacoder</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/heat_tree_pq.html">heat_tree_pq</a></span><span class="op">(</span><span class="va">ten</span>,</span>
<span>  node_size <span class="op">=</span> <span class="va">n_obs</span>,</span>
<span>  node_color <span class="op">=</span> <span class="va">nb_sequences</span>,</span>
<span>  node_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>  tree_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>  node_size_trans <span class="op">=</span> <span class="st">"log10 area"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-8-1.png" width="960"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/treemap_pq.html">treemap_pq</a></span><span class="op">(</span><span class="va">ten</span>, lvl1 <span class="op">=</span> <span class="st">"Order"</span>, lvl2 <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-9-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="beta-diversity-analysis-effect-of-patient-status-and-cohort">Beta-diversity analysis : effect of patient status and cohort<a class="anchor" aria-label="anchor" href="#beta-diversity-analysis-effect-of-patient-status-and-cohort"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/circle_pq.html">circle_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"patient_status"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-10-1.png" width="960"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/upset_pq.html">upset_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"patient_status_vs_cohort"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-11-1.png" width="960"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggvenn_pq.html">ggvenn_pq</a></span><span class="op">(</span><span class="fu"><a href="../reference/clean_pq.html">clean_pq</a></span><span class="op">(</span><span class="va">ten</span>, force_taxa_as_columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  <span class="st">"cohort"</span>,</span>
<span>  rarefy_before_merging <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ten_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clean_pq.html">clean_pq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">ten</span>, <span class="va">patient_status</span> <span class="op">==</span> <span class="st">"Control"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/heat_tree_pq.html">heat_tree_pq</a></span><span class="op">(</span><span class="va">ten_control</span>,</span>
<span>  node_size <span class="op">=</span> <span class="va">n_obs</span>,</span>
<span>  node_color <span class="op">=</span> <span class="va">nb_sequences</span>,</span>
<span>  node_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>  tree_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>  node_size_trans <span class="op">=</span> <span class="st">"log10 area"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ten_ADHD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clean_pq.html">clean_pq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">ten</span>, <span class="va">patient_status</span> <span class="op">==</span> <span class="st">"ADHD"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_ADHD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/heat_tree_pq.html">heat_tree_pq</a></span><span class="op">(</span><span class="va">ten_ADHD</span>,</span>
<span>  node_size <span class="op">=</span> <span class="va">n_obs</span>,</span>
<span>  node_color <span class="op">=</span> <span class="va">nb_sequences</span>,</span>
<span>  node_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>  tree_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>  node_size_trans <span class="op">=</span> <span class="st">"log10 area"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_control</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Control"</span><span class="op">)</span> <span class="op">+</span> <span class="va">p_ADHD</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"ADHD"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-13-1.png" width="960"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="../reference/track_wkflow.html">track_wkflow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"All samples"</span> <span class="op">=</span> <span class="va">ten</span>,</span>
<span>  <span class="st">"Control samples"</span> <span class="op">=</span> <span class="va">ten_control</span>,</span>
<span>  <span class="st">"ADHD samples"</span> <span class="op">=</span> <span class="va">ten_ADHD</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">nb_sequences</th>
<th align="right">nb_clusters</th>
<th align="right">nb_samples</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">All samples</td>
<td align="right">485932</td>
<td align="right">151</td>
<td align="right">27</td>
</tr>
<tr class="even">
<td align="left">Control samples</td>
<td align="right">239329</td>
<td align="right">130</td>
<td align="right">14</td>
</tr>
<tr class="odd">
<td align="left">ADHD samples</td>
<td align="right">246603</td>
<td align="right">142</td>
<td align="right">13</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/adonis_pq.html">adonis_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"cohort + patient_status"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Permutation test for adonis under reduced model</span></span>
<span><span class="co">#&gt; Terms added sequentially (first to last)</span></span>
<span><span class="co">#&gt; Permutation: free</span></span>
<span><span class="co">#&gt; Number of permutations: 999</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; vegan::adonis2(formula = .formula, data = metadata)</span></span>
<span><span class="co">#&gt;                Df SumOfSqs      R2      F Pr(&gt;F)  </span></span>
<span><span class="co">#&gt; cohort          2   0.7922 0.11785 1.6626  0.076 .</span></span>
<span><span class="co">#&gt; patient_status  1   0.4503 0.06698 1.8898  0.077 .</span></span>
<span><span class="co">#&gt; Residual       23   5.4799 0.81517                </span></span>
<span><span class="co">#&gt; Total          26   6.7223 1.00000                </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ten</span><span class="op">@</span><span class="va">tax_table</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="va">ten</span><span class="op">@</span><span class="va">tax_table</span>,</span>
<span>  <span class="st">"Species"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">ten</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/biplot_pq.html">biplot_pq</a></span><span class="op">(</span><span class="fu"><a href="../reference/subset_taxa_pq.html">subset_taxa_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="va">ten</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3000</span><span class="op">)</span>,</span>
<span>  merge_sample_by <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>  fact <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>  nudge_y <span class="op">=</span> <span class="fl">0.4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-16-1.png" width="960"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/multitax_bar_pq.html">multitax_bar_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"patient_status"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-17-1.png" width="960"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/multitax_bar_pq.html">multitax_bar_pq</a></span><span class="op">(</span><span class="va">ten</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"patient_status"</span>,</span>
<span>  nb_seq <span class="op">=</span> <span class="cn">FALSE</span>, log10trans <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_deseq2_pq.html">plot_deseq2_pq</a></span><span class="op">(</span><span class="va">ten</span>,</span>
<span>  contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"patient_status"</span>, <span class="st">"ADHD"</span>, <span class="st">"Control"</span><span class="op">)</span>,</span>
<span>  taxolev <span class="op">=</span> <span class="st">"Genus"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in</span></span>
<span><span class="co">#&gt; design formula are characters, converting to factors</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-19-1.png" width="960"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LEfSe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/diff_analysis.html" class="external-link">diff_analysis</a></span><span class="op">(</span></span>
<span>  <span class="va">ten</span>,</span>
<span>  classgroup <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>  mlfun <span class="op">=</span> <span class="st">"lda"</span>,</span>
<span>  ldascore <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  p.adjust.methods <span class="op">=</span> <span class="st">"bh"</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MicrobiotaProcess/man/ggeffectsize.html" class="external-link">ggeffectsize</a></span><span class="op">(</span><span class="va">LEfSe</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"#00AED7"</span>,</span>
<span>    <span class="st">"#FD9347"</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tengeler_files/figure-html/unnamed-chunk-20-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] metacoder_0.3.6          iNEXT_3.0.0              patchwork_1.2.0         </span></span>
<span><span class="co">#&gt;  [4] MiscMetabar_0.7.1        dplyr_1.1.4              dada2_1.30.0            </span></span>
<span><span class="co">#&gt;  [7] Rcpp_1.0.12              ggplot2_3.4.4            phyloseq_1.46.0         </span></span>
<span><span class="co">#&gt; [10] MicrobiotaProcess_1.14.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] libcoin_1.0-10              RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [3] shape_1.4.6                 jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;   [5] magrittr_2.0.3              TH.data_1.1-2              </span></span>
<span><span class="co">#&gt;   [7] modeltools_0.2-23           farver_2.1.1               </span></span>
<span><span class="co">#&gt;   [9] rmarkdown_2.25              GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt;  [11] fs_1.6.3                    zlibbioc_1.48.0            </span></span>
<span><span class="co">#&gt;  [13] ragg_1.2.7                  vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [15] multtest_2.58.0             memoise_2.0.1              </span></span>
<span><span class="co">#&gt;  [17] Rsamtools_2.18.0            RCurl_1.98-1.14            </span></span>
<span><span class="co">#&gt;  [19] ggtree_3.10.0               htmltools_0.5.7            </span></span>
<span><span class="co">#&gt;  [21] S4Arrays_1.2.0              ComplexUpset_1.3.3         </span></span>
<span><span class="co">#&gt;  [23] Rhdf5lib_1.24.1             SparseArray_1.2.3          </span></span>
<span><span class="co">#&gt;  [25] rhdf5_2.46.1                gridGraphics_0.5-1         </span></span>
<span><span class="co">#&gt;  [27] sass_0.4.8                  bslib_0.6.1                </span></span>
<span><span class="co">#&gt;  [29] desc_1.4.3                  plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [31] sandwich_3.1-0              zoo_1.8-12                 </span></span>
<span><span class="co">#&gt;  [33] cachem_1.0.8                ggfittext_0.10.2           </span></span>
<span><span class="co">#&gt;  [35] GenomicAlignments_1.38.2    igraph_2.0.1.1             </span></span>
<span><span class="co">#&gt;  [37] lifecycle_1.0.4             iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [39] pkgconfig_2.0.3             Matrix_1.6-1.1             </span></span>
<span><span class="co">#&gt;  [41] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [43] GenomeInfoDbData_1.2.11     MatrixGenerics_1.14.0      </span></span>
<span><span class="co">#&gt;  [45] digest_0.6.34               aplot_0.2.2                </span></span>
<span><span class="co">#&gt;  [47] colorspace_2.1-0            ggnewscale_0.4.9           </span></span>
<span><span class="co">#&gt;  [49] ShortRead_1.60.0            S4Vectors_0.40.2           </span></span>
<span><span class="co">#&gt;  [51] DESeq2_1.42.0               textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [53] GenomicRanges_1.54.1        hwriter_1.3.2.1            </span></span>
<span><span class="co">#&gt;  [55] vegan_2.6-4                 labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [57] fansi_1.0.6                 abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [59] mgcv_1.9-0                  compiler_4.3.2             </span></span>
<span><span class="co">#&gt;  [61] withr_3.0.0                 BiocParallel_1.36.0        </span></span>
<span><span class="co">#&gt;  [63] highr_0.10                  ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt;  [65] MASS_7.3-60                 DelayedArray_0.28.0        </span></span>
<span><span class="co">#&gt;  [67] biomformat_1.30.0           permute_0.9-7              </span></span>
<span><span class="co">#&gt;  [69] tools_4.3.2                 ape_5.7-1                  </span></span>
<span><span class="co">#&gt;  [71] glue_1.7.0                  treemapify_2.5.6           </span></span>
<span><span class="co">#&gt;  [73] nlme_3.1-163                rhdf5filters_1.14.1        </span></span>
<span><span class="co">#&gt;  [75] grid_4.3.2                  cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [77] reshape2_1.4.4              ade4_1.7-22                </span></span>
<span><span class="co">#&gt;  [79] generics_0.1.3              gtable_0.3.4               </span></span>
<span><span class="co">#&gt;  [81] tidyr_1.3.1                 ggVennDiagram_1.5.0        </span></span>
<span><span class="co">#&gt;  [83] data.table_1.15.0           coin_1.4-3                 </span></span>
<span><span class="co">#&gt;  [85] utf8_1.2.4                  XVector_0.42.0             </span></span>
<span><span class="co">#&gt;  [87] BiocGenerics_0.48.1         ggrepel_0.9.5              </span></span>
<span><span class="co">#&gt;  [89] foreach_1.5.2               pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [91] stringr_1.5.1               yulab.utils_0.1.4          </span></span>
<span><span class="co">#&gt;  [93] circlize_0.4.15             splines_4.3.2              </span></span>
<span><span class="co">#&gt;  [95] treeio_1.26.0               lattice_0.21-9             </span></span>
<span><span class="co">#&gt;  [97] survival_3.5-7              deldir_2.0-2               </span></span>
<span><span class="co">#&gt;  [99] tidyselect_1.2.0            locfit_1.5-9.8             </span></span>
<span><span class="co">#&gt; [101] pbapply_1.7-2               Biostrings_2.70.2          </span></span>
<span><span class="co">#&gt; [103] knitr_1.45                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [105] IRanges_2.36.0              SummarizedExperiment_1.32.0</span></span>
<span><span class="co">#&gt; [107] ggtreeExtra_1.12.0          stats4_4.3.2               </span></span>
<span><span class="co">#&gt; [109] xfun_0.41                   Biobase_2.62.0             </span></span>
<span><span class="co">#&gt; [111] matrixStats_1.2.0           stringi_1.8.3              </span></span>
<span><span class="co">#&gt; [113] lazyeval_0.2.2              ggfun_0.1.4                </span></span>
<span><span class="co">#&gt; [115] yaml_2.3.8                  evaluate_0.23              </span></span>
<span><span class="co">#&gt; [117] codetools_0.2-19            interp_1.1-6               </span></span>
<span><span class="co">#&gt; [119] tibble_3.2.1                ggplotify_0.1.2            </span></span>
<span><span class="co">#&gt; [121] cli_3.6.2                   RcppParallel_5.1.7         </span></span>
<span><span class="co">#&gt; [123] systemfonts_1.0.5           munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [125] jquerylib_0.1.4             GenomeInfoDb_1.38.5        </span></span>
<span><span class="co">#&gt; [127] png_0.1-8                   parallel_4.3.2             </span></span>
<span><span class="co">#&gt; [129] ggh4x_0.2.8                 pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [131] latticeExtra_0.6-30         jpeg_0.1-10                </span></span>
<span><span class="co">#&gt; [133] bitops_1.0-7                ggstar_1.0.4               </span></span>
<span><span class="co">#&gt; [135] mvtnorm_1.2-4               tidytree_0.4.6             </span></span>
<span><span class="co">#&gt; [137] GA_3.2.4                    scales_1.3.0               </span></span>
<span><span class="co">#&gt; [139] purrr_1.0.2                 crayon_1.5.2               </span></span>
<span><span class="co">#&gt; [141] rlang_1.1.3                 multcomp_1.4-25</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Tengeler, A.C., Dam, S.A., Wiesmann, M. et al. Gut microbiota from
persons with attention-deficit/hyperactivity disorder affects the brain
in mice. Microbiome 8, 44 (2020). <a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-020-00816-x" class="external-link uri">https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-020-00816-x</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Adrien Taudière.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
